{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "97f9ac63-1c15-481c-af9a-ba032429149d",
   "metadata": {},
   "source": [
    "# Pediatric DS AML vs TAM\n",
    "\n",
    "DESeq2 Analysis with Kallisto Quantitation Input\n",
    "\n",
    "## Following the Instructions from BioConductor DESeq2 \n",
    "[Transcript abundance to DESeq2 Analysis](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#quick-start)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ac92837-8970-4820-8557-342a4993df84",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Transcript abundance files and tximport / tximeta\n",
    "\n",
    "Our recommended pipeline for DESeq2 is to use fast transcript abundance quantifiers upstream of DESeq2, and then to create gene-level count matrices for use with DESeq2 by importing the quantification data using tximport (Soneson, Love, and Robinson 2015). This workflow allows users to import transcript abundance estimates from a variety of external software, including the following methods:\n",
    "\n",
    "* [Salmon](http://combine-lab.github.io/salmon/) (Patro et al. 2017)\n",
    "* [Sailfish](http://www.cs.cmu.edu/~ckingsf/software/sailfish/) (Patro, Mount, and Kingsford 2014)\n",
    "* [kallisto](https://pachterlab.github.io/kallisto/about.html) (Bray et al. 2016)\n",
    "* [RSEM](http://deweylab.github.io/RSEM/) (Li and Dewey 2011)\n",
    "\n",
    "Some advantages of using the above methods for transcript abundance estimation are: \n",
    "* (i) this approach corrects for potential changes in gene length across samples (e.g. from differential isoform usage) (Trapnell et al. 2013), \n",
    "* (ii) some of these methods (Salmon, Sailfish, kallisto) are substantially faster and require less memory and disk usage compared to alignment-based methods that require creation and storage of BAM files, and \n",
    "* (iii) it is possible to avoid discarding those fragments that can align to multiple genes with homologous sequence, thus increasing sensitivity (Robert and Watson 2015).\n",
    "\n",
    "Full details on the motivation and methods for importing transcript level abundance and count estimates, summarizing to gene-level count matrices and producing an offset which corrects for potential changes in average transcript length across samples are described in (Soneson, Love, and Robinson 2015). Note that the tximport-to-DESeq2 approach uses estimated gene counts from the transcript abundance quantifiers, but not normalized counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f50203a6-0b7e-4e70-8bd1-39ca8cad55c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"readr\")\n",
    "library(\"readr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6a8330eb-745a-41f7-8924-843fb9bf8ec2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version '3.14' is out-of-date; the current release version '3.15'\n",
      "  is available with R version '4.2'; see https://bioconductor.org/install\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.18), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Old packages: 'backports', 'blob', 'brew', 'brio', 'broom', 'bslib', 'callr',\n",
      "  'caret', 'class', 'clipr', 'colorspace', 'commonmark', 'conflicted', 'covr',\n",
      "  'cpp11', 'crayon', 'credentials', 'crosstalk', 'curl', 'data.table', 'DBI',\n",
      "  'dbplyr', 'desc', 'devtools', 'dials', 'diffobj', 'digest', 'dplyr', 'DT',\n",
      "  'dtplyr', 'e1071', 'evaluate', 'fansi', 'farver', 'forcats', 'foreach',\n",
      "  'forecast', 'fs', 'furrr', 'future', 'future.apply', 'gargle', 'generics',\n",
      "  'gert', 'ggplot2', 'gh', 'git2r', 'gitcreds', 'globals', 'glue',\n",
      "  'googlesheets4', 'gower', 'gtable', 'hardhat', 'haven', 'hms', 'htmltools',\n",
      "  'httpuv', 'httr', 'infer', 'ipred', 'IRdisplay', 'IRkernel', 'isoband',\n",
      "  'iterators', 'jsonlite', 'knitr', 'later', 'lattice', 'lhs', 'lifecycle',\n",
      "  'lmtest', 'lubridate', 'magrittr', 'markdown', 'MASS', 'Matrix', 'memoise',\n",
      "  'mgcv', 'mime', 'modeldata', 'modelr', 'nlme', 'nnet', 'objectProperties',\n",
      "  'objectSignals', 'openssl', 'parallelly', 'parsnip', 'patchwork', 'pbdZMQ',\n",
      "  'pillar', 'pkgbuild', 'pkgload', 'plyr', 'processx', 'progressr', 'proxy',\n",
      "  'ps', 'purrr', 'quantmod', 'randomForest', 'rcmdcheck', 'RColorBrewer',\n",
      "  'Rcpp', 'RcppArmadillo', 'RCurl', 'readxl', 'recipes', 'remotes', 'repr',\n",
      "  'reprex', 'rex', 'rlang', 'rmarkdown', 'RODBC', 'roxygen2', 'rpart',\n",
      "  'rprojroot', 'rsample', 'RSQLite', 'rstudioapi', 'rversions', 'rvest',\n",
      "  'S4Vectors', 'sass', 'scales', 'sessioninfo', 'sevenbridges', 'shiny',\n",
      "  'stringi', 'stringr', 'survival', 'sys', 'testthat', 'tibble', 'tidymodels',\n",
      "  'tidyr', 'tidyselect', 'tidyverse', 'timeDate', 'tinytex', 'tseries', 'TTR',\n",
      "  'tune', 'tzdb', 'urca', 'usethis', 'uuid', 'vctrs', 'viridisLite', 'waldo',\n",
      "  'withr', 'workflows', 'workflowsets', 'xfun', 'xml2', 'xts', 'yaml',\n",
      "  'yardstick', 'zip', 'zoo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!require(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "BiocManager::install(version = \"3.14\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5d288d00-c591-4fab-8261-37fe14b883c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.18), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as current; use `force = TRUE` to\n",
      "  re-install: 'tximport'”\n",
      "Old packages: 'backports', 'blob', 'brew', 'brio', 'broom', 'bslib', 'callr',\n",
      "  'caret', 'class', 'clipr', 'colorspace', 'commonmark', 'conflicted', 'covr',\n",
      "  'cpp11', 'crayon', 'credentials', 'crosstalk', 'curl', 'data.table', 'DBI',\n",
      "  'dbplyr', 'desc', 'devtools', 'dials', 'diffobj', 'digest', 'dplyr', 'DT',\n",
      "  'dtplyr', 'e1071', 'evaluate', 'fansi', 'farver', 'forcats', 'foreach',\n",
      "  'forecast', 'fs', 'furrr', 'future', 'future.apply', 'gargle', 'generics',\n",
      "  'gert', 'ggplot2', 'gh', 'git2r', 'gitcreds', 'globals', 'glue',\n",
      "  'googlesheets4', 'gower', 'gtable', 'hardhat', 'haven', 'hms', 'htmltools',\n",
      "  'httpuv', 'httr', 'infer', 'ipred', 'IRdisplay', 'IRkernel', 'isoband',\n",
      "  'iterators', 'jsonlite', 'knitr', 'later', 'lattice', 'lhs', 'lifecycle',\n",
      "  'lmtest', 'lubridate', 'magrittr', 'markdown', 'MASS', 'Matrix', 'memoise',\n",
      "  'mgcv', 'mime', 'modeldata', 'modelr', 'nlme', 'nnet', 'objectProperties',\n",
      "  'objectSignals', 'openssl', 'parallelly', 'parsnip', 'patchwork', 'pbdZMQ',\n",
      "  'pillar', 'pkgbuild', 'pkgload', 'plyr', 'processx', 'progressr', 'proxy',\n",
      "  'ps', 'purrr', 'quantmod', 'randomForest', 'rcmdcheck', 'RColorBrewer',\n",
      "  'Rcpp', 'RcppArmadillo', 'RCurl', 'readxl', 'recipes', 'remotes', 'repr',\n",
      "  'reprex', 'rex', 'rlang', 'rmarkdown', 'RODBC', 'roxygen2', 'rpart',\n",
      "  'rprojroot', 'rsample', 'RSQLite', 'rstudioapi', 'rversions', 'rvest',\n",
      "  'S4Vectors', 'sass', 'scales', 'sessioninfo', 'sevenbridges', 'shiny',\n",
      "  'stringi', 'stringr', 'survival', 'sys', 'testthat', 'tibble', 'tidymodels',\n",
      "  'tidyr', 'tidyselect', 'tidyverse', 'timeDate', 'tinytex', 'tseries', 'TTR',\n",
      "  'tune', 'tzdb', 'urca', 'usethis', 'uuid', 'vctrs', 'viridisLite', 'waldo',\n",
      "  'withr', 'workflows', 'workflowsets', 'xfun', 'xml2', 'xts', 'yaml',\n",
      "  'yardstick', 'zip', 'zoo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(\"tximport\")\n",
    "library(tximport)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "558b915a-1a77-450b-ac5c-f607a3c31dd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.18), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as current; use `force = TRUE` to\n",
      "  re-install: 'biomaRt'”\n",
      "Old packages: 'backports', 'blob', 'brew', 'brio', 'broom', 'bslib', 'callr',\n",
      "  'caret', 'class', 'clipr', 'colorspace', 'commonmark', 'conflicted', 'covr',\n",
      "  'cpp11', 'crayon', 'credentials', 'crosstalk', 'curl', 'data.table', 'DBI',\n",
      "  'dbplyr', 'desc', 'devtools', 'dials', 'diffobj', 'digest', 'dplyr', 'DT',\n",
      "  'dtplyr', 'e1071', 'evaluate', 'fansi', 'farver', 'forcats', 'foreach',\n",
      "  'forecast', 'fs', 'furrr', 'future', 'future.apply', 'gargle', 'generics',\n",
      "  'gert', 'ggplot2', 'gh', 'git2r', 'gitcreds', 'globals', 'glue',\n",
      "  'googlesheets4', 'gower', 'gtable', 'hardhat', 'haven', 'hms', 'htmltools',\n",
      "  'httpuv', 'httr', 'infer', 'ipred', 'IRdisplay', 'IRkernel', 'isoband',\n",
      "  'iterators', 'jsonlite', 'knitr', 'later', 'lattice', 'lhs', 'lifecycle',\n",
      "  'lmtest', 'lubridate', 'magrittr', 'markdown', 'MASS', 'Matrix', 'memoise',\n",
      "  'mgcv', 'mime', 'modeldata', 'modelr', 'nlme', 'nnet', 'objectProperties',\n",
      "  'objectSignals', 'openssl', 'parallelly', 'parsnip', 'patchwork', 'pbdZMQ',\n",
      "  'pillar', 'pkgbuild', 'pkgload', 'plyr', 'processx', 'progressr', 'proxy',\n",
      "  'ps', 'purrr', 'quantmod', 'randomForest', 'rcmdcheck', 'RColorBrewer',\n",
      "  'Rcpp', 'RcppArmadillo', 'RCurl', 'readxl', 'recipes', 'remotes', 'repr',\n",
      "  'reprex', 'rex', 'rlang', 'rmarkdown', 'RODBC', 'roxygen2', 'rpart',\n",
      "  'rprojroot', 'rsample', 'RSQLite', 'rstudioapi', 'rversions', 'rvest',\n",
      "  'S4Vectors', 'sass', 'scales', 'sessioninfo', 'sevenbridges', 'shiny',\n",
      "  'stringi', 'stringr', 'survival', 'sys', 'testthat', 'tibble', 'tidymodels',\n",
      "  'tidyr', 'tidyselect', 'tidyverse', 'timeDate', 'tinytex', 'tseries', 'TTR',\n",
      "  'tune', 'tzdb', 'urca', 'usethis', 'uuid', 'vctrs', 'viridisLite', 'waldo',\n",
      "  'withr', 'workflows', 'workflowsets', 'xfun', 'xml2', 'xts', 'yaml',\n",
      "  'yardstick', 'zip', 'zoo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(\"biomaRt\")\n",
    "library(biomaRt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "29aad1f8-fc8e-41ca-b0be-13eeb8e96865",
   "metadata": {},
   "outputs": [],
   "source": [
    "mart <- biomaRt::useMart(biomart=\"ensembl\", \n",
    "                     dataset = \"hsapiens_gene_ensembl\",\n",
    "                        host = \"https://useast.ensembl.org\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e546c36b-55e7-4a88-bf71-8eca88b0d25e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttg <- biomaRt::getBM(\n",
    "  attributes = c(\"ensembl_transcript_id\", \"transcript_version\",\n",
    "  \"ensembl_gene_id\", \"external_gene_name\", \"description\",\n",
    "  \"transcript_biotype\"),\n",
    "  mart = mart)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6f6cef45-9fac-4c8d-849a-7a969742de11",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttg <- dplyr::rename(ttg, target_id = ensembl_transcript_id,\n",
    "  ens_gene = ensembl_gene_id, ext_gene = external_gene_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b024adbd-6e9a-4aeb-9f14-295402818f25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>target_id</th><th scope=col>ens_gene</th><th scope=col>ext_gene</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENST00000387314</td><td>ENSG00000210049</td><td>MT-TF  </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENST00000389680</td><td>ENSG00000211459</td><td>MT-RNR1</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENST00000387342</td><td>ENSG00000210077</td><td>MT-TV  </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENST00000387347</td><td>ENSG00000210082</td><td>MT-RNR2</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENST00000386347</td><td>ENSG00000209082</td><td>MT-TL1 </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENST00000361390</td><td>ENSG00000198888</td><td>MT-ND1 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & target\\_id & ens\\_gene & ext\\_gene\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENST00000387314 & ENSG00000210049 & MT-TF  \\\\\n",
       "\t2 & ENST00000389680 & ENSG00000211459 & MT-RNR1\\\\\n",
       "\t3 & ENST00000387342 & ENSG00000210077 & MT-TV  \\\\\n",
       "\t4 & ENST00000387347 & ENSG00000210082 & MT-RNR2\\\\\n",
       "\t5 & ENST00000386347 & ENSG00000209082 & MT-TL1 \\\\\n",
       "\t6 & ENST00000361390 & ENSG00000198888 & MT-ND1 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | target_id &lt;chr&gt; | ens_gene &lt;chr&gt; | ext_gene &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | ENST00000387314 | ENSG00000210049 | MT-TF   |\n",
       "| 2 | ENST00000389680 | ENSG00000211459 | MT-RNR1 |\n",
       "| 3 | ENST00000387342 | ENSG00000210077 | MT-TV   |\n",
       "| 4 | ENST00000387347 | ENSG00000210082 | MT-RNR2 |\n",
       "| 5 | ENST00000386347 | ENSG00000209082 | MT-TL1  |\n",
       "| 6 | ENST00000361390 | ENSG00000198888 | MT-ND1  |\n",
       "\n"
      ],
      "text/plain": [
       "  target_id       ens_gene        ext_gene\n",
       "1 ENST00000387314 ENSG00000210049 MT-TF   \n",
       "2 ENST00000389680 ENSG00000211459 MT-RNR1 \n",
       "3 ENST00000387342 ENSG00000210077 MT-TV   \n",
       "4 ENST00000387347 ENSG00000210082 MT-RNR2 \n",
       "5 ENST00000386347 ENSG00000209082 MT-TL1  \n",
       "6 ENST00000361390 ENSG00000198888 MT-ND1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>target_id</th><th scope=col>ext_gene</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENST00000387314</td><td>MT-TF  </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENST00000389680</td><td>MT-RNR1</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENST00000387342</td><td>MT-TV  </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENST00000387347</td><td>MT-RNR2</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENST00000386347</td><td>MT-TL1 </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENST00000361390</td><td>MT-ND1 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & target\\_id & ext\\_gene\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENST00000387314 & MT-TF  \\\\\n",
       "\t2 & ENST00000389680 & MT-RNR1\\\\\n",
       "\t3 & ENST00000387342 & MT-TV  \\\\\n",
       "\t4 & ENST00000387347 & MT-RNR2\\\\\n",
       "\t5 & ENST00000386347 & MT-TL1 \\\\\n",
       "\t6 & ENST00000361390 & MT-ND1 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | target_id &lt;chr&gt; | ext_gene &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | ENST00000387314 | MT-TF   |\n",
       "| 2 | ENST00000389680 | MT-RNR1 |\n",
       "| 3 | ENST00000387342 | MT-TV   |\n",
       "| 4 | ENST00000387347 | MT-RNR2 |\n",
       "| 5 | ENST00000386347 | MT-TL1  |\n",
       "| 6 | ENST00000361390 | MT-ND1  |\n",
       "\n"
      ],
      "text/plain": [
       "  target_id       ext_gene\n",
       "1 ENST00000387314 MT-TF   \n",
       "2 ENST00000389680 MT-RNR1 \n",
       "3 ENST00000387342 MT-TV   \n",
       "4 ENST00000387347 MT-RNR2 \n",
       "5 ENST00000386347 MT-TL1  \n",
       "6 ENST00000361390 MT-ND1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ttg <- dplyr::select(ttg, c('target_id', 'ens_gene', 'ext_gene'))\n",
    "head(ttg)\n",
    "tx2gene <- dplyr::select(ttg, c('target_id','ext_gene'))\n",
    "head(tx2gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e61cba9-f4c4-494e-9135-f3833af79e6c",
   "metadata": {},
   "source": [
    "### Read in Metadata\n",
    "Copied from the project-files directory a custom `metadata_ten_samples_only.csv` file was created"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3099f471-55ff-45a9-9c1c-a06e156bda28",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- read.table('/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/metadata_ten_samples_only.csv', sep=\",\", header=TRUE, stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "963f85b5-0a83-4b01-b680-b99b5062a5ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>File.name</th><th scope=col>Case.ID</th><th scope=col>subject</th><th scope=col>Sample.ID</th><th scope=col>sample</th><th scope=col>Gender</th><th scope=col>Disease.type</th><th scope=col>Paired.end</th><th scope=col>Abundance</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAXSBH</td><td>PAXSBH</td><td>PAXSBH-03A-01R</td><td>PAXSBH-03A-01R</td><td>Female</td><td>TAM</td><td>1</td><td>PAXSBH-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAXSBH</td><td>PAXSBH</td><td>PAXSBH-03A-01R</td><td>PAXSBH-03A-01R</td><td>Female</td><td>TAM</td><td>2</td><td>PAXSBH-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PAXWGW</td><td>PAXWGW</td><td>PAXWGW-03A-01R</td><td>PAXWGW-03A-01R</td><td>Female</td><td>TAM</td><td>1</td><td>PAXWGW-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAXWGW</td><td>PAXWGW</td><td>PAXWGW-03A-01R</td><td>PAXWGW-03A-01R</td><td>Female</td><td>TAM</td><td>2</td><td>PAXWGW-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PASNSP</td><td>PASNSP</td><td>PASNSP-03A-01R</td><td>PASNSP-03A-01R</td><td>Male  </td><td>TAM</td><td>1</td><td>PASNSP-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PASNSP</td><td>PASNSP</td><td>PASNSP-03A-01R</td><td>PASNSP-03A-01R</td><td>Male  </td><td>TAM</td><td>2</td><td>PASNSP-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & File.name & Case.ID & subject & Sample.ID & sample & Gender & Disease.type & Paired.end & Abundance\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & PAXSBH-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAXSBH & PAXSBH & PAXSBH-03A-01R & PAXSBH-03A-01R & Female & TAM & 1 & PAXSBH-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t2 & PAXSBH-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAXSBH & PAXSBH & PAXSBH-03A-01R & PAXSBH-03A-01R & Female & TAM & 2 & PAXSBH-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t3 & PAXWGW-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PAXWGW & PAXWGW & PAXWGW-03A-01R & PAXWGW-03A-01R & Female & TAM & 1 & PAXWGW-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t4 & PAXWGW-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAXWGW & PAXWGW & PAXWGW-03A-01R & PAXWGW-03A-01R & Female & TAM & 2 & PAXWGW-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t5 & PASNSP-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PASNSP & PASNSP & PASNSP-03A-01R & PASNSP-03A-01R & Male   & TAM & 1 & PASNSP-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t6 & PASNSP-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PASNSP & PASNSP & PASNSP-03A-01R & PASNSP-03A-01R & Male   & TAM & 2 & PASNSP-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | File.name &lt;chr&gt; | Case.ID &lt;chr&gt; | subject &lt;chr&gt; | Sample.ID &lt;chr&gt; | sample &lt;chr&gt; | Gender &lt;chr&gt; | Disease.type &lt;chr&gt; | Paired.end &lt;int&gt; | Abundance &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAXSBH | PAXSBH | PAXSBH-03A-01R | PAXSBH-03A-01R | Female | TAM | 1 | PAXSBH-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 2 | PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAXSBH | PAXSBH | PAXSBH-03A-01R | PAXSBH-03A-01R | Female | TAM | 2 | PAXSBH-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 3 | PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PAXWGW | PAXWGW | PAXWGW-03A-01R | PAXWGW-03A-01R | Female | TAM | 1 | PAXWGW-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 4 | PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAXWGW | PAXWGW | PAXWGW-03A-01R | PAXWGW-03A-01R | Female | TAM | 2 | PAXWGW-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 5 | PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PASNSP | PASNSP | PASNSP-03A-01R | PASNSP-03A-01R | Male   | TAM | 1 | PASNSP-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 6 | PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PASNSP | PASNSP | PASNSP-03A-01R | PASNSP-03A-01R | Male   | TAM | 2 | PASNSP-03A-01R.kallisto_quant.abundance.tsv |\n",
       "\n"
      ],
      "text/plain": [
       "  File.name                                                     Case.ID subject\n",
       "1 PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAXSBH  PAXSBH \n",
       "2 PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAXSBH  PAXSBH \n",
       "3 PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PAXWGW  PAXWGW \n",
       "4 PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAXWGW  PAXWGW \n",
       "5 PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PASNSP  PASNSP \n",
       "6 PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PASNSP  PASNSP \n",
       "  Sample.ID      sample         Gender Disease.type Paired.end\n",
       "1 PAXSBH-03A-01R PAXSBH-03A-01R Female TAM          1         \n",
       "2 PAXSBH-03A-01R PAXSBH-03A-01R Female TAM          2         \n",
       "3 PAXWGW-03A-01R PAXWGW-03A-01R Female TAM          1         \n",
       "4 PAXWGW-03A-01R PAXWGW-03A-01R Female TAM          2         \n",
       "5 PASNSP-03A-01R PASNSP-03A-01R Male   TAM          1         \n",
       "6 PASNSP-03A-01R PASNSP-03A-01R Male   TAM          2         \n",
       "  Abundance                                  \n",
       "1 PAXSBH-03A-01R.kallisto_quant.abundance.tsv\n",
       "2 PAXSBH-03A-01R.kallisto_quant.abundance.tsv\n",
       "3 PAXWGW-03A-01R.kallisto_quant.abundance.tsv\n",
       "4 PAXWGW-03A-01R.kallisto_quant.abundance.tsv\n",
       "5 PASNSP-03A-01R.kallisto_quant.abundance.tsv\n",
       "6 PASNSP-03A-01R.kallisto_quant.abundance.tsv"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9349a2c7-26b0-4d03-a8a5-fe52502f9820",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- dplyr::select(metadata, c('Case.ID', 'Sample.ID', 'Gender', 'Disease.type', 'Abundance'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8b9d1daa-2123-4c95-a932-ce75c0b6b9d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- dplyr::distinct(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "906c6185-9147-4a16-b483-1da766627a52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Case.ID</th><th scope=col>Sample.ID</th><th scope=col>Gender</th><th scope=col>Disease.type</th><th scope=col>Abundance</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>PAXSBH</td><td>PAXSBH-03A-01R</td><td>Female</td><td>TAM   </td><td>PAXSBH-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>PAXWGW</td><td>PAXWGW-03A-01R</td><td>Female</td><td>TAM   </td><td>PAXWGW-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>PASNSP</td><td>PASNSP-03A-01R</td><td>Male  </td><td>TAM   </td><td>PASNSP-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>PASWXF</td><td>PASWXF-03A-01R</td><td>Male  </td><td>TAM   </td><td>PASWXF-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>PASXCL</td><td>PASXCL-03A-01R</td><td>Male  </td><td>TAM   </td><td>PASXCL-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>PAVZTK</td><td>PAVZTK-09A-01R</td><td>Female</td><td>DS-AML</td><td>PAVZTK-09A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & Case.ID & Sample.ID & Gender & Disease.type & Abundance\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & PAXSBH & PAXSBH-03A-01R & Female & TAM    & PAXSBH-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t2 & PAXWGW & PAXWGW-03A-01R & Female & TAM    & PAXWGW-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t3 & PASNSP & PASNSP-03A-01R & Male   & TAM    & PASNSP-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t4 & PASWXF & PASWXF-03A-01R & Male   & TAM    & PASWXF-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t5 & PASXCL & PASXCL-03A-01R & Male   & TAM    & PASXCL-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t6 & PAVZTK & PAVZTK-09A-01R & Female & DS-AML & PAVZTK-09A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | Case.ID &lt;chr&gt; | Sample.ID &lt;chr&gt; | Gender &lt;chr&gt; | Disease.type &lt;chr&gt; | Abundance &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | PAXSBH | PAXSBH-03A-01R | Female | TAM    | PAXSBH-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 2 | PAXWGW | PAXWGW-03A-01R | Female | TAM    | PAXWGW-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 3 | PASNSP | PASNSP-03A-01R | Male   | TAM    | PASNSP-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 4 | PASWXF | PASWXF-03A-01R | Male   | TAM    | PASWXF-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 5 | PASXCL | PASXCL-03A-01R | Male   | TAM    | PASXCL-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 6 | PAVZTK | PAVZTK-09A-01R | Female | DS-AML | PAVZTK-09A-01R.kallisto_quant.abundance.tsv |\n",
       "\n"
      ],
      "text/plain": [
       "  Case.ID Sample.ID      Gender Disease.type\n",
       "1 PAXSBH  PAXSBH-03A-01R Female TAM         \n",
       "2 PAXWGW  PAXWGW-03A-01R Female TAM         \n",
       "3 PASNSP  PASNSP-03A-01R Male   TAM         \n",
       "4 PASWXF  PASWXF-03A-01R Male   TAM         \n",
       "5 PASXCL  PASXCL-03A-01R Male   TAM         \n",
       "6 PAVZTK  PAVZTK-09A-01R Female DS-AML      \n",
       "  Abundance                                  \n",
       "1 PAXSBH-03A-01R.kallisto_quant.abundance.tsv\n",
       "2 PAXWGW-03A-01R.kallisto_quant.abundance.tsv\n",
       "3 PASNSP-03A-01R.kallisto_quant.abundance.tsv\n",
       "4 PASWXF-03A-01R.kallisto_quant.abundance.tsv\n",
       "5 PASXCL-03A-01R.kallisto_quant.abundance.tsv\n",
       "6 PAVZTK-09A-01R.kallisto_quant.abundance.tsv"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "8c35aecf-4ef3-4a3d-a075-64b1ec513292",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- dplyr::rename(metadata, sample = Sample.ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "1e0f9eca-0467-4500-be6e-b397da13f469",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- dplyr::rename(metadata, path = Abundance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0070eea0-f641-496c-badb-33aa6d72a34a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Case.ID</th><th scope=col>sample</th><th scope=col>Gender</th><th scope=col>Disease.type</th><th scope=col>path</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>PAXSBH</td><td>PAXSBH-03A-01R</td><td>Female</td><td>TAM   </td><td>PAXSBH-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>PAXWGW</td><td>PAXWGW-03A-01R</td><td>Female</td><td>TAM   </td><td>PAXWGW-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>PASNSP</td><td>PASNSP-03A-01R</td><td>Male  </td><td>TAM   </td><td>PASNSP-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>PASWXF</td><td>PASWXF-03A-01R</td><td>Male  </td><td>TAM   </td><td>PASWXF-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>PASXCL</td><td>PASXCL-03A-01R</td><td>Male  </td><td>TAM   </td><td>PASXCL-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>PAVZTK</td><td>PAVZTK-09A-01R</td><td>Female</td><td>DS-AML</td><td>PAVZTK-09A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & Case.ID & sample & Gender & Disease.type & path\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & PAXSBH & PAXSBH-03A-01R & Female & TAM    & PAXSBH-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t2 & PAXWGW & PAXWGW-03A-01R & Female & TAM    & PAXWGW-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t3 & PASNSP & PASNSP-03A-01R & Male   & TAM    & PASNSP-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t4 & PASWXF & PASWXF-03A-01R & Male   & TAM    & PASWXF-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t5 & PASXCL & PASXCL-03A-01R & Male   & TAM    & PASXCL-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t6 & PAVZTK & PAVZTK-09A-01R & Female & DS-AML & PAVZTK-09A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | Case.ID &lt;chr&gt; | sample &lt;chr&gt; | Gender &lt;chr&gt; | Disease.type &lt;chr&gt; | path &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | PAXSBH | PAXSBH-03A-01R | Female | TAM    | PAXSBH-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 2 | PAXWGW | PAXWGW-03A-01R | Female | TAM    | PAXWGW-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 3 | PASNSP | PASNSP-03A-01R | Male   | TAM    | PASNSP-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 4 | PASWXF | PASWXF-03A-01R | Male   | TAM    | PASWXF-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 5 | PASXCL | PASXCL-03A-01R | Male   | TAM    | PASXCL-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 6 | PAVZTK | PAVZTK-09A-01R | Female | DS-AML | PAVZTK-09A-01R.kallisto_quant.abundance.tsv |\n",
       "\n"
      ],
      "text/plain": [
       "  Case.ID sample         Gender Disease.type\n",
       "1 PAXSBH  PAXSBH-03A-01R Female TAM         \n",
       "2 PAXWGW  PAXWGW-03A-01R Female TAM         \n",
       "3 PASNSP  PASNSP-03A-01R Male   TAM         \n",
       "4 PASWXF  PASWXF-03A-01R Male   TAM         \n",
       "5 PASXCL  PASXCL-03A-01R Male   TAM         \n",
       "6 PAVZTK  PAVZTK-09A-01R Female DS-AML      \n",
       "  path                                       \n",
       "1 PAXSBH-03A-01R.kallisto_quant.abundance.tsv\n",
       "2 PAXWGW-03A-01R.kallisto_quant.abundance.tsv\n",
       "3 PASNSP-03A-01R.kallisto_quant.abundance.tsv\n",
       "4 PASWXF-03A-01R.kallisto_quant.abundance.tsv\n",
       "5 PASXCL-03A-01R.kallisto_quant.abundance.tsv\n",
       "6 PAVZTK-09A-01R.kallisto_quant.abundance.tsv"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "dadf7544-afb6-4df7-95de-d1f155adb242",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.18), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as current; use `force = TRUE` to\n",
      "  re-install: 'tximport'”\n",
      "Old packages: 'backports', 'blob', 'brew', 'brio', 'broom', 'bslib', 'callr',\n",
      "  'caret', 'class', 'clipr', 'colorspace', 'commonmark', 'conflicted', 'covr',\n",
      "  'cpp11', 'crayon', 'credentials', 'crosstalk', 'curl', 'data.table', 'DBI',\n",
      "  'dbplyr', 'desc', 'devtools', 'dials', 'diffobj', 'digest', 'dplyr', 'DT',\n",
      "  'dtplyr', 'e1071', 'evaluate', 'fansi', 'farver', 'forcats', 'foreach',\n",
      "  'forecast', 'fs', 'furrr', 'future', 'future.apply', 'gargle', 'generics',\n",
      "  'gert', 'ggplot2', 'gh', 'git2r', 'gitcreds', 'globals', 'glue',\n",
      "  'googlesheets4', 'gower', 'gtable', 'hardhat', 'haven', 'hms', 'htmltools',\n",
      "  'httpuv', 'httr', 'infer', 'ipred', 'IRdisplay', 'IRkernel', 'isoband',\n",
      "  'iterators', 'jsonlite', 'knitr', 'later', 'lattice', 'lhs', 'lifecycle',\n",
      "  'lmtest', 'lubridate', 'magrittr', 'markdown', 'MASS', 'Matrix', 'memoise',\n",
      "  'mgcv', 'mime', 'modeldata', 'modelr', 'nlme', 'nnet', 'objectProperties',\n",
      "  'objectSignals', 'openssl', 'parallelly', 'parsnip', 'patchwork', 'pbdZMQ',\n",
      "  'pillar', 'pkgbuild', 'pkgload', 'plyr', 'processx', 'progressr', 'proxy',\n",
      "  'ps', 'purrr', 'quantmod', 'randomForest', 'rcmdcheck', 'RColorBrewer',\n",
      "  'Rcpp', 'RcppArmadillo', 'RCurl', 'readxl', 'recipes', 'remotes', 'repr',\n",
      "  'reprex', 'rex', 'rlang', 'rmarkdown', 'RODBC', 'roxygen2', 'rpart',\n",
      "  'rprojroot', 'rsample', 'RSQLite', 'rstudioapi', 'rversions', 'rvest',\n",
      "  'S4Vectors', 'sass', 'scales', 'sessioninfo', 'sevenbridges', 'shiny',\n",
      "  'stringi', 'stringr', 'survival', 'sys', 'testthat', 'tibble', 'tidymodels',\n",
      "  'tidyr', 'tidyselect', 'tidyverse', 'timeDate', 'tinytex', 'tseries', 'TTR',\n",
      "  'tune', 'tzdb', 'urca', 'usethis', 'uuid', 'vctrs', 'viridisLite', 'waldo',\n",
      "  'withr', 'workflows', 'workflowsets', 'xfun', 'xml2', 'xts', 'yaml',\n",
      "  'yardstick', 'zip', 'zoo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(\"tximport\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ae942cde-2f80-4bb4-abf5-2b222c429646",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(tximport)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d86cfb29-cd51-4b3d-b6fb-954afc8ab2bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.18), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Installing package(s) 'rhdf5'\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'backports', 'blob', 'brew', 'brio', 'broom', 'bslib', 'callr',\n",
      "  'caret', 'class', 'clipr', 'colorspace', 'commonmark', 'conflicted', 'covr',\n",
      "  'cpp11', 'crayon', 'credentials', 'crosstalk', 'curl', 'data.table', 'DBI',\n",
      "  'dbplyr', 'desc', 'devtools', 'dials', 'diffobj', 'digest', 'dplyr', 'DT',\n",
      "  'dtplyr', 'e1071', 'evaluate', 'fansi', 'farver', 'forcats', 'foreach',\n",
      "  'forecast', 'fs', 'furrr', 'future', 'future.apply', 'gargle', 'generics',\n",
      "  'gert', 'ggplot2', 'gh', 'git2r', 'gitcreds', 'globals', 'glue',\n",
      "  'googlesheets4', 'gower', 'gtable', 'hardhat', 'haven', 'hms', 'htmltools',\n",
      "  'httpuv', 'httr', 'infer', 'ipred', 'IRdisplay', 'IRkernel', 'isoband',\n",
      "  'iterators', 'jsonlite', 'knitr', 'later', 'lattice', 'lhs', 'lifecycle',\n",
      "  'lmtest', 'lubridate', 'magrittr', 'markdown', 'MASS', 'Matrix', 'memoise',\n",
      "  'mgcv', 'mime', 'modeldata', 'modelr', 'nlme', 'nnet', 'objectProperties',\n",
      "  'objectSignals', 'openssl', 'parallelly', 'parsnip', 'patchwork', 'pbdZMQ',\n",
      "  'pillar', 'pkgbuild', 'pkgload', 'plyr', 'processx', 'progressr', 'proxy',\n",
      "  'ps', 'purrr', 'quantmod', 'randomForest', 'rcmdcheck', 'RColorBrewer',\n",
      "  'Rcpp', 'RcppArmadillo', 'RCurl', 'readxl', 'recipes', 'remotes', 'repr',\n",
      "  'reprex', 'rex', 'rlang', 'rmarkdown', 'RODBC', 'roxygen2', 'rpart',\n",
      "  'rprojroot', 'rsample', 'RSQLite', 'rstudioapi', 'rversions', 'rvest',\n",
      "  'S4Vectors', 'sass', 'scales', 'sessioninfo', 'sevenbridges', 'shiny',\n",
      "  'stringi', 'stringr', 'survival', 'sys', 'testthat', 'tibble', 'tidymodels',\n",
      "  'tidyr', 'tidyselect', 'tidyverse', 'timeDate', 'tinytex', 'tseries', 'TTR',\n",
      "  'tune', 'tzdb', 'urca', 'usethis', 'uuid', 'vctrs', 'viridisLite', 'waldo',\n",
      "  'withr', 'workflows', 'workflowsets', 'xfun', 'xml2', 'xts', 'yaml',\n",
      "  'yardstick', 'zip', 'zoo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(\"rhdf5\", force=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "210580bd-4c29-4f56-af8f-b2ddcd12819c",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(rhdf5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a62d5619-006f-4cc0-a39d-e76dbaf81803",
   "metadata": {},
   "source": [
    "Made a local copy of the file in the projects file directory for that moment"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14e5ad7a-00a4-4b51-9220-bbede42c8b55",
   "metadata": {},
   "source": [
    "### Prepping abundance data from Kallisto using the BioConductor tximport Manual options\n",
    "\n",
    "Build for our manual was on October 16, 2022\n",
    "\n",
    "From the manual on tximport we see the following options:\n",
    "\n",
    "* **files**:  a character vector of filenames for the transcript-level abundances\n",
    "* **type**:   character string, the type of software used to generate the abundances. Options are \"salmon\", \"sailfish\", \"alevin\", \"kallisto\", \"rsem\", \"stringtie\", or \"none\". This argument is used to autofill the arguments below (geneIdCol, etc.) \"none\" means that the user will specify these columns. Be aware that specifying type other than \"none\" will ignore the arguments below (geneIdCol, etc.)\n",
    "* **txIn**:  a logical (TRUE/FALSE), whether the incoming files are transcript level (default TRUE)\n",
    "* **txOut**: a logical (TRUE/FALSE), whether the function should just output transcript-level (default FALSE) \n",
    "* **countsFromAbundance** a character, either \"no\" (default), \"scaledTPM\", \"lengthScaledTPM\", or \"dtuS- caledTPM\". Whether to generate estimated counts using abundance estimates:\n",
    "* • scaled up to library size (scaledTPM),\n",
    "* • scaled using the average transcript length over samples and then the library size (lengthScaledTPM), or\n",
    "* • scaled using the median transcript length among isoforms of a gene, and then the library size (dtuScaledTPM).\n",
    "dtuScaledTPM is designed for DTU analysis in combination with txOut=TRUE, and it requires specifing a tx2gene data.frame. dtuScaledTPM works such that within a gene, values from all samples and all transcripts get scaled by the same fixed median transcript length. If using scaledTPM, lengthScaledTPM, or gene- LengthScaledTPM, the counts are no longer correlated across samples with tran- script length, and so the length offset matrix should not be used.\n",
    "* **tx2gene** a two-column data.frame linking transcript id (column 1) to gene id (column 2). the column names are not relevant, but this column order must be used. this argument is required for gene-level summarization, and the tximport vignette describes how to construct this data.frame (see Details below). An automated solution to avoid having to create tx2gene if one has quantified with Salmon or alevin with human or mouse transcriptomes is to use the tximeta function from the tximeta Bioconductor package.\n",
    "* **ignoreTxVersion** a logical (TRUE/FALSE), whether to split the tx id on the ’.’ character to remove version informa- tion to facilitate matching with the tx id in tx2gene (default FALSE)\n",
    "* **ignoreAfterBar** a logical (TRUE/FALSE), whether to split the tx id on the ’|’ character to facilitate matching with the tx id in tx2gene (default FALSE). if txOut=TRUE it will strip the text after ’|’ on the rownames of the matrices\n",
    "* **geneIdCol** name of column with gene id. if missing, the tx2gene argument can be used. Note that this argument and the other four \"...Col\" arguments below are ignored unless type=\"none\"\n",
    "* **txIdCol** name of column with tx id\n",
    "* **abundanceCol** name of column with abundances (e.g. TPM or FPKM) name of column with estimated counts\n",
    "* **countsCol** name of column with feature length information\n",
    "* **lengthCol** name of column with feature length information\n",
    "* **importer** a function used to read in the files\n",
    "\n",
    "We are going to use the following:\n",
    "\n",
    "* files           - will be specified from the metadata file - column renamed to `path`\n",
    "* type            - \"kallisto\"\n",
    "* txIn            - TRUE\n",
    "* txOut           - FALSE\n",
    "* tx2gene         - \"tx2gene\" (we used biomaRt to create the tx2gene file)\n",
    "* ignoreTxVersion - TRUE (we are going to ignore the version info after the \".\"\n",
    "* ignoreAfterBar  - TRUE (we will ignore the other info after the \"|\"\n",
    "* geneIdCol       - \"target_id\" (from the ttg)\n",
    "* txIdCol         - \"ext_gene\"\n",
    "* abundanceCol    - \"tpm\"\n",
    "* countsCol       - \"est_counts\"\n",
    "* lengthCol       - \"eff_length\"\n",
    "* importer        - \"readr\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "449985d1-5f33-48e3-98bb-735a23c9b1ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(readr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "135089e7-8ec4-4866-b013-202bdb579adf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'PAXSBH-03A-01R.kallisto_quant.abundance.tsv'</li><li>'PAXWGW-03A-01R.kallisto_quant.abundance.tsv'</li><li>'PASNSP-03A-01R.kallisto_quant.abundance.tsv'</li><li>'PASWXF-03A-01R.kallisto_quant.abundance.tsv'</li><li>'PASXCL-03A-01R.kallisto_quant.abundance.tsv'</li><li>'PAVZTK-09A-01R.kallisto_quant.abundance.tsv'</li><li>'PAWIYJ-09A-01R.kallisto_quant.abundance.tsv'</li><li>'PAXCGV-09A-01R.kallisto_quant.abundance.tsv'</li><li>'PAXLGI-09A-01R.kallisto_quant.abundance.tsv'</li><li>'PAXPEY-09A-01R.kallisto_quant.abundance.tsv'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'PAXSBH-03A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item 'PAXWGW-03A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item 'PASNSP-03A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item 'PASWXF-03A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item 'PASXCL-03A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item 'PAVZTK-09A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item 'PAWIYJ-09A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item 'PAXCGV-09A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item 'PAXLGI-09A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item 'PAXPEY-09A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'PAXSBH-03A-01R.kallisto_quant.abundance.tsv'\n",
       "2. 'PAXWGW-03A-01R.kallisto_quant.abundance.tsv'\n",
       "3. 'PASNSP-03A-01R.kallisto_quant.abundance.tsv'\n",
       "4. 'PASWXF-03A-01R.kallisto_quant.abundance.tsv'\n",
       "5. 'PASXCL-03A-01R.kallisto_quant.abundance.tsv'\n",
       "6. 'PAVZTK-09A-01R.kallisto_quant.abundance.tsv'\n",
       "7. 'PAWIYJ-09A-01R.kallisto_quant.abundance.tsv'\n",
       "8. 'PAXCGV-09A-01R.kallisto_quant.abundance.tsv'\n",
       "9. 'PAXLGI-09A-01R.kallisto_quant.abundance.tsv'\n",
       "10. 'PAXPEY-09A-01R.kallisto_quant.abundance.tsv'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"PAXSBH-03A-01R.kallisto_quant.abundance.tsv\"\n",
       " [2] \"PAXWGW-03A-01R.kallisto_quant.abundance.tsv\"\n",
       " [3] \"PASNSP-03A-01R.kallisto_quant.abundance.tsv\"\n",
       " [4] \"PASWXF-03A-01R.kallisto_quant.abundance.tsv\"\n",
       " [5] \"PASXCL-03A-01R.kallisto_quant.abundance.tsv\"\n",
       " [6] \"PAVZTK-09A-01R.kallisto_quant.abundance.tsv\"\n",
       " [7] \"PAWIYJ-09A-01R.kallisto_quant.abundance.tsv\"\n",
       " [8] \"PAXCGV-09A-01R.kallisto_quant.abundance.tsv\"\n",
       " [9] \"PAXLGI-09A-01R.kallisto_quant.abundance.tsv\"\n",
       "[10] \"PAXPEY-09A-01R.kallisto_quant.abundance.tsv\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata$path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e3dd445b-efa1-492b-8c78-0f082504b107",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "dfd371e1-6d53-4119-9f51-9686463f2f87",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Note: importing `abundance.h5` is typically faster than `abundance.tsv`\n",
      "\n",
      "reading in files with read_tsv\n",
      "\n",
      "1 \n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error: 'metadata$path' does not exist in current working directory ('/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data').\n",
     "output_type": "error",
     "traceback": [
      "Error: 'metadata$path' does not exist in current working directory ('/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data').\nTraceback:\n",
      "1. tximport(files = \"metadata$path\", existenceOptional = TRUE, type = \"kallisto\", \n .     txIn = TRUE, txOut = FALSE, tx2gene = \"tx2gene\", ignoreTxVersion = TRUE, \n .     ignoreAfterBar = TRUE, geneIdCol = \"ext_gene\", txIdCol = \"target_id\", \n .     abundanceCol = \"tpm\", countsCol = \"est_counts\", lengthCol = \"eff_length\")",
      "2. as.data.frame(importer(files[i]))",
      "3. importer(files[i])",
      "4. readr::read_tsv(x, progress = FALSE, col_types = col.types)",
      "5. vroom::vroom(file, delim = \"\\t\", col_names = col_names, col_types = col_types, \n .     col_select = {\n .         {\n .             col_select\n .         }\n .     }, id = id, .name_repair = name_repair, skip = skip, n_max = n_max, \n .     na = na, quote = quote, comment = comment, skip_empty_rows = skip_empty_rows, \n .     trim_ws = trim_ws, escape_double = TRUE, escape_backslash = FALSE, \n .     locale = locale, guess_max = guess_max, show_col_types = show_col_types, \n .     progress = progress, altrep = lazy, num_threads = num_threads)",
      "6. vroom_(file, delim = delim %||% col_types$delim, col_names = col_names, \n .     col_types = col_types, id = id, skip = skip, col_select = col_select, \n .     name_repair = .name_repair, na = na, quote = quote, trim_ws = trim_ws, \n .     escape_double = escape_double, escape_backslash = escape_backslash, \n .     comment = comment, skip_empty_rows = skip_empty_rows, locale = locale, \n .     guess_max = guess_max, n_max = n_max, altrep = vroom_altrep(altrep), \n .     num_threads = num_threads, progress = progress)",
      "7. (function (path, write = FALSE) \n . {\n .     if (is.raw(path)) {\n .         return(rawConnection(path, \"rb\"))\n .     }\n .     if (!is.character(path)) {\n .         return(path)\n .     }\n .     if (is_url(path)) {\n .         if (requireNamespace(\"curl\", quietly = TRUE)) {\n .             con <- curl::curl(path)\n .         }\n .         else {\n .             rlang::inform(\"`curl` package not installed, falling back to using `url()`\")\n .             con <- url(path)\n .         }\n .         ext <- tolower(tools::file_ext(path))\n .         return(switch(ext, zip = , bz2 = , xz = {\n .             close(con)\n .             stop(\"Reading from remote `\", ext, \"` compressed files is not supported,\\n\", \n .                 \"  download the files locally first.\", call. = FALSE)\n .         }, gz = gzcon(con), con))\n .     }\n .     path <- enc2utf8(path)\n .     p <- split_path_ext(basename_utf8(path))\n .     if (write) {\n .         path <- normalizePath_utf8(path, mustWork = FALSE)\n .     }\n .     else {\n .         path <- check_path(path)\n .     }\n .     if (rlang::is_installed(\"archive\")) {\n .         formats <- archive_formats(p$extension)\n .         extension <- p$extension\n .         while (is.null(formats) && nzchar(extension)) {\n .             extension <- split_path_ext(extension)$extension\n .             formats <- archive_formats(extension)\n .         }\n .         if (!is.null(formats)) {\n .             p$extension <- extension\n .             if (write) {\n .                 if (is.null(formats[[1]])) {\n .                   return(archive::file_write(path, filter = formats[[2]]))\n .                 }\n .                 return(archive::archive_write(path, p$path, format = formats[[1]], \n .                   filter = formats[[2]]))\n .             }\n .             if (is.null(formats[[1]])) {\n .                 return(archive::file_read(path, filter = formats[[2]]))\n .             }\n .             return(archive::archive_read(path, format = formats[[1]], \n .                 filter = formats[[2]]))\n .         }\n .     }\n .     if (!write) {\n .         compression <- detect_compression(path)\n .     }\n .     else {\n .         compression <- NA\n .     }\n .     if (is.na(compression)) {\n .         compression <- tools::file_ext(path)\n .     }\n .     if (write && compression == \"zip\") {\n .         stop(\"Can only read from, not write to, .zip\", call. = FALSE)\n .     }\n .     switch(compression, gz = gzfile(path, \"\"), bz2 = bzfile(path, \n .         \"\"), xz = xzfile(path, \"\"), zip = zipfile(path, \"\"), \n .         if (!has_trailing_newline(path)) {\n .             file(path)\n .         } else {\n .             path\n .         })\n . })(\"metadata$path\")",
      "8. check_path(path)",
      "9. stop(\"'\", path, \"' does not exist\", if (!is_absolute_path(path)) {\n .     paste0(\" in current working directory ('\", getwd(), \"')\")\n . }, \".\", call. = FALSE)"
     ]
    }
   ],
   "source": [
    "txi.kallisto <- tximport(files           = \"metadata$path\", \n",
    "                         existenceOptional = TRUE,\n",
    "                         type            = \"kallisto\",\n",
    "                         txIn            = TRUE,\n",
    "                         txOut           = FALSE,\n",
    "                         tx2gene         = \"tx2gene\",\n",
    "                         ignoreTxVersion = TRUE,\n",
    "                         ignoreAfterBar  = TRUE,\n",
    "                         geneIdCol       = \"ext_gene\",\n",
    "                         txIdCol         = \"target_id\",\n",
    "                         abundanceCol    = \"tpm\",\n",
    "                         countsCol       = \"est_counts\",\n",
    "                         lengthCol       = \"eff_length\",)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0e092bb6-2350-4b9c-87e0-8559cc6c15ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all(file.exists(metadata$path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e18b844d-693c-4c23-99f4-d714dabbe4af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv'</li><li>'/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv'</li><li>'/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv'</li><li>'/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASWXF-03A-01R.kallisto_quant.abundance.tsv'</li><li>'/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASXCL-03A-01R.kallisto_quant.abundance.tsv'</li><li>'/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAVZTK-09A-01R.kallisto_quant.abundance.tsv'</li><li>'/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAWIYJ-09A-01R.kallisto_quant.abundance.tsv'</li><li>'/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXCGV-09A-01R.kallisto_quant.abundance.tsv'</li><li>'/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXLGI-09A-01R.kallisto_quant.abundance.tsv'</li><li>'/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXPEY-09A-01R.kallisto_quant.abundance.tsv'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASWXF-03A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASXCL-03A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAVZTK-09A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAWIYJ-09A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXCGV-09A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXLGI-09A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\item '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXPEY-09A-01R.kallisto\\_quant.abundance.tsv'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv'\n",
       "2. '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv'\n",
       "3. '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv'\n",
       "4. '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASWXF-03A-01R.kallisto_quant.abundance.tsv'\n",
       "5. '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASXCL-03A-01R.kallisto_quant.abundance.tsv'\n",
       "6. '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAVZTK-09A-01R.kallisto_quant.abundance.tsv'\n",
       "7. '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAWIYJ-09A-01R.kallisto_quant.abundance.tsv'\n",
       "8. '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXCGV-09A-01R.kallisto_quant.abundance.tsv'\n",
       "9. '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXLGI-09A-01R.kallisto_quant.abundance.tsv'\n",
       "10. '/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXPEY-09A-01R.kallisto_quant.abundance.tsv'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv\"\n",
       " [2] \"/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv\"\n",
       " [3] \"/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv\"\n",
       " [4] \"/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASWXF-03A-01R.kallisto_quant.abundance.tsv\"\n",
       " [5] \"/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASXCL-03A-01R.kallisto_quant.abundance.tsv\"\n",
       " [6] \"/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAVZTK-09A-01R.kallisto_quant.abundance.tsv\"\n",
       " [7] \"/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAWIYJ-09A-01R.kallisto_quant.abundance.tsv\"\n",
       " [8] \"/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXCGV-09A-01R.kallisto_quant.abundance.tsv\"\n",
       " [9] \"/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXLGI-09A-01R.kallisto_quant.abundance.tsv\"\n",
       "[10] \"/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXPEY-09A-01R.kallisto_quant.abundance.tsv\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata$path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "bfb4dcf5-247a-4b6e-958a-07faeb14518e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (files, type = c(\"none\", \"salmon\", \"sailfish\", \"alevin\", \n",
       "<span style=white-space:pre-wrap>    \"kallisto\", \"rsem\", \"stringtie\"), txIn = TRUE, txOut = FALSE, </span>\n",
       "<span style=white-space:pre-wrap>    countsFromAbundance = c(\"no\", \"scaledTPM\", \"lengthScaledTPM\", </span>\n",
       "<span style=white-space:pre-wrap>        \"dtuScaledTPM\"), tx2gene = NULL, varReduce = FALSE, dropInfReps = FALSE, </span>\n",
       "<span style=white-space:pre-wrap>    infRepStat = NULL, ignoreTxVersion = FALSE, ignoreAfterBar = FALSE, </span>\n",
       "<span style=white-space:pre-wrap>    geneIdCol, txIdCol, abundanceCol, countsCol, lengthCol, importer = NULL, </span>\n",
       "<span style=white-space:pre-wrap>    existenceOptional = FALSE, sparse = FALSE, sparseThreshold = 1, </span>\n",
       "<span style=white-space:pre-wrap>    readLength = 75, alevinArgs = NULL) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    infRepImporter &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>    type &lt;- match.arg(type)</span>\n",
       "<span style=white-space:pre-wrap>    countsFromAbundance &lt;- match.arg(countsFromAbundance)</span>\n",
       "<span style=white-space:pre-wrap>    if (countsFromAbundance == \"dtuScaledTPM\") {</span>\n",
       "<span style=white-space:pre-wrap>        stopifnot(txOut)</span>\n",
       "<span style=white-space:pre-wrap>        if (is.null(tx2gene)) </span>\n",
       "<span style=white-space:pre-wrap>            stop(\"'dtuScaledTPM' requires 'tx2gene' input\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!existenceOptional) </span>\n",
       "<span style=white-space:pre-wrap>        stopifnot(all(file.exists(files)))</span>\n",
       "<span style=white-space:pre-wrap>    if (!txIn &amp; txOut) </span>\n",
       "<span style=white-space:pre-wrap>        stop(\"txOut only an option when transcript-level data is read in (txIn=TRUE)\")</span>\n",
       "<span style=white-space:pre-wrap>    stopifnot(length(files) &gt; 0)</span>\n",
       "<span style=white-space:pre-wrap>    kallisto.h5 &lt;- basename(files[1]) == \"abundance.h5\"</span>\n",
       "<span style=white-space:pre-wrap>    if (type == \"kallisto\" &amp; !kallisto.h5) {</span>\n",
       "<span style=white-space:pre-wrap>        message(\"Note: importing `abundance.h5` is typically faster than `abundance.tsv`\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (type == \"rsem\" &amp; txIn &amp; grepl(\"genes\", files[1])) {</span>\n",
       "<span style=white-space:pre-wrap>        message(\"It looks like you are importing RSEM genes.results files, setting txIn=FALSE\")</span>\n",
       "<span style=white-space:pre-wrap>        txIn &lt;- FALSE</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(infRepStat)) {</span>\n",
       "<span style=white-space:pre-wrap>        if (dropInfReps) </span>\n",
       "<span style=white-space:pre-wrap>            stop(\"infRepStat requires infReps\")</span>\n",
       "<span style=white-space:pre-wrap>        if (type == \"alevin\") </span>\n",
       "<span style=white-space:pre-wrap>            stop(\"infRepStat does not currently work with alevin input\")</span>\n",
       "<span style=white-space:pre-wrap>        if (sparse) </span>\n",
       "<span style=white-space:pre-wrap>            stop(\"infRepStat does not currently work with sparse output\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (type == \"alevin\") {</span>\n",
       "<span style=white-space:pre-wrap>        if (is.null(alevinArgs)) {</span>\n",
       "<span style=white-space:pre-wrap>            alevinArgs &lt;- list(filterBarcodes = FALSE, tierImport = FALSE, </span>\n",
       "<span style=white-space:pre-wrap>                forceSlow = FALSE, dropMeanVar = FALSE)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        stopifnot(is(alevinArgs, \"list\"))</span>\n",
       "<span style=white-space:pre-wrap>        stopifnot(all(sapply(alevinArgs, is.logical)))</span>\n",
       "<span style=white-space:pre-wrap>        alevinArgNms &lt;- c(\"filterBarcodes\", \"tierImport\", \"forceSlow\", </span>\n",
       "<span style=white-space:pre-wrap>            \"dropMeanVar\")</span>\n",
       "<span style=white-space:pre-wrap>        stopifnot(all(names(alevinArgs) %in% alevinArgNms))</span>\n",
       "<span style=white-space:pre-wrap>        filterBarcodes &lt;- if (is.null(alevinArgs$filterBarcodes)) </span>\n",
       "<span style=white-space:pre-wrap>            FALSE</span>\n",
       "<span style=white-space:pre-wrap>        else alevinArgs$filterBarcodes</span>\n",
       "<span style=white-space:pre-wrap>        tierImport &lt;- if (is.null(alevinArgs$tierImport)) </span>\n",
       "<span style=white-space:pre-wrap>            FALSE</span>\n",
       "<span style=white-space:pre-wrap>        else alevinArgs$tierImport</span>\n",
       "<span style=white-space:pre-wrap>        forceSlow &lt;- if (is.null(alevinArgs$forceSlow)) </span>\n",
       "<span style=white-space:pre-wrap>            FALSE</span>\n",
       "<span style=white-space:pre-wrap>        else alevinArgs$forceSlow</span>\n",
       "<span style=white-space:pre-wrap>        dropMeanVar &lt;- if (is.null(alevinArgs$dropMeanVar)) </span>\n",
       "<span style=white-space:pre-wrap>            FALSE</span>\n",
       "<span style=white-space:pre-wrap>        else alevinArgs$dropMeanVar</span>\n",
       "<span style=white-space:pre-wrap>        if (length(files) &gt; 1) </span>\n",
       "<span style=white-space:pre-wrap>            stop(\"alevin import currently only supports a single experiment\")</span>\n",
       "<span style=white-space:pre-wrap>        if (compareVersion(getAlevinVersion(files), \"0.14.0\") == </span>\n",
       "<span style=white-space:pre-wrap>            -1) {</span>\n",
       "<span style=white-space:pre-wrap>            stop(\"use of tximport version &gt;= 1.18 requires alevin version &gt;= 0.14\")</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        mat &lt;- readAlevin(files, dropInfReps, filterBarcodes, </span>\n",
       "<span style=white-space:pre-wrap>            tierImport, forceSlow, dropMeanVar)</span>\n",
       "<span style=white-space:pre-wrap>        if (!is.list(mat)) {</span>\n",
       "<span style=white-space:pre-wrap>            txi &lt;- list(abundance = NULL, counts = mat)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            txi &lt;- list(abundance = NULL, counts = mat$counts, </span>\n",
       "<span style=white-space:pre-wrap>                mean = mat$mean, variance = mat$variance)</span>\n",
       "<span style=white-space:pre-wrap>            if (tierImport) {</span>\n",
       "<span style=white-space:pre-wrap>                txi$tier &lt;- mat$tier</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            if (\"infReps\" %in% names(mat)) {</span>\n",
       "<span style=white-space:pre-wrap>                txi$infReps &lt;- mat$infReps</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        txi$length &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>        txi$countsFromAbundance = \"no\"</span>\n",
       "<span style=white-space:pre-wrap>        return(txi)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    readrStatus &lt;- FALSE</span>\n",
       "<span style=white-space:pre-wrap>    if (is.null(importer) &amp; !kallisto.h5) {</span>\n",
       "<span style=white-space:pre-wrap>        if (!requireNamespace(\"readr\", quietly = TRUE)) {</span>\n",
       "<span style=white-space:pre-wrap>            message(\"reading in files with read.delim (install 'readr' package for speed up)\")</span>\n",
       "<span style=white-space:pre-wrap>            importer &lt;- read.delim</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            timeZoneMissing &lt;- length(suppressWarnings(OlsonNames())) == </span>\n",
       "<span style=white-space:pre-wrap>                0 | is.na(Sys.timezone())</span>\n",
       "<span style=white-space:pre-wrap>            if (timeZoneMissing) {</span>\n",
       "<span style=white-space:pre-wrap>                message(\"reading in files with read.delim ('readr' installed but won't work w/o timezones)\")</span>\n",
       "<span style=white-space:pre-wrap>                importer &lt;- read.delim</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            else {</span>\n",
       "<span style=white-space:pre-wrap>                message(\"reading in files with read_tsv\")</span>\n",
       "<span style=white-space:pre-wrap>                readrStatus &lt;- TRUE</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (type %in% c(\"salmon\", \"sailfish\")) {</span>\n",
       "<span style=white-space:pre-wrap>        txIdCol &lt;- \"Name\"</span>\n",
       "<span style=white-space:pre-wrap>        abundanceCol &lt;- \"TPM\"</span>\n",
       "<span style=white-space:pre-wrap>        countsCol &lt;- \"NumReads\"</span>\n",
       "<span style=white-space:pre-wrap>        lengthCol &lt;- \"EffectiveLength\"</span>\n",
       "<span style=white-space:pre-wrap>        if (readrStatus &amp; is.null(importer)) {</span>\n",
       "<span style=white-space:pre-wrap>            col.types &lt;- readr::cols(readr::col_character(), </span>\n",
       "<span style=white-space:pre-wrap>                readr::col_integer(), readr::col_double(), readr::col_double(), </span>\n",
       "<span style=white-space:pre-wrap>                readr::col_double())</span>\n",
       "<span style=white-space:pre-wrap>            importer &lt;- function(x) readr::read_tsv(x, progress = FALSE, </span>\n",
       "<span style=white-space:pre-wrap>                col_types = col.types)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        infRepImporter &lt;- if (dropInfReps) {</span>\n",
       "<span style=white-space:pre-wrap>            NULL</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            function(x) readInfRepFish(x, type)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (type == \"kallisto\") {</span>\n",
       "<span style=white-space:pre-wrap>        txIdCol &lt;- \"target_id\"</span>\n",
       "<span style=white-space:pre-wrap>        abundanceCol &lt;- \"tpm\"</span>\n",
       "<span style=white-space:pre-wrap>        countsCol &lt;- \"est_counts\"</span>\n",
       "<span style=white-space:pre-wrap>        lengthCol &lt;- \"eff_length\"</span>\n",
       "<span style=white-space:pre-wrap>        if (kallisto.h5) {</span>\n",
       "<span style=white-space:pre-wrap>            importer &lt;- read_kallisto_h5</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else if (readrStatus &amp; is.null(importer)) {</span>\n",
       "<span style=white-space:pre-wrap>            col.types &lt;- readr::cols(readr::col_character(), </span>\n",
       "<span style=white-space:pre-wrap>                readr::col_integer(), readr::col_double(), readr::col_double(), </span>\n",
       "<span style=white-space:pre-wrap>                readr::col_double())</span>\n",
       "<span style=white-space:pre-wrap>            importer &lt;- function(x) readr::read_tsv(x, progress = FALSE, </span>\n",
       "<span style=white-space:pre-wrap>                col_types = col.types)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        infRepImporter &lt;- if (dropInfReps) {</span>\n",
       "<span style=white-space:pre-wrap>            NULL</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            readInfRepKallisto</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (type == \"rsem\") {</span>\n",
       "<span style=white-space:pre-wrap>        if (txIn) {</span>\n",
       "<span style=white-space:pre-wrap>            txIdCol &lt;- \"transcript_id\"</span>\n",
       "<span style=white-space:pre-wrap>            abundanceCol &lt;- \"TPM\"</span>\n",
       "<span style=white-space:pre-wrap>            countsCol &lt;- \"expected_count\"</span>\n",
       "<span style=white-space:pre-wrap>            lengthCol &lt;- \"effective_length\"</span>\n",
       "<span style=white-space:pre-wrap>            if (readrStatus &amp; is.null(importer)) {</span>\n",
       "<span style=white-space:pre-wrap>                col.types &lt;- readr::cols(readr::col_character(), </span>\n",
       "<span style=white-space:pre-wrap>                  readr::col_character(), readr::col_integer(), </span>\n",
       "<span style=white-space:pre-wrap>                  readr::col_double(), readr::col_double(), readr::col_double(), </span>\n",
       "<span style=white-space:pre-wrap>                  readr::col_double(), readr::col_double())</span>\n",
       "<span style=white-space:pre-wrap>                importer &lt;- function(x) readr::read_tsv(x, progress = FALSE, </span>\n",
       "<span style=white-space:pre-wrap>                  col_types = col.types)</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            geneIdCol &lt;- \"gene_id\"</span>\n",
       "<span style=white-space:pre-wrap>            abundanceCol &lt;- \"TPM\"</span>\n",
       "<span style=white-space:pre-wrap>            countsCol &lt;- \"expected_count\"</span>\n",
       "<span style=white-space:pre-wrap>            lengthCol &lt;- \"effective_length\"</span>\n",
       "<span style=white-space:pre-wrap>            if (readrStatus &amp; is.null(importer)) {</span>\n",
       "<span style=white-space:pre-wrap>                col.types &lt;- readr::cols(readr::col_character(), </span>\n",
       "<span style=white-space:pre-wrap>                  readr::col_character(), readr::col_double(), </span>\n",
       "<span style=white-space:pre-wrap>                  readr::col_double(), readr::col_double(), readr::col_double(), </span>\n",
       "<span style=white-space:pre-wrap>                  readr::col_double())</span>\n",
       "<span style=white-space:pre-wrap>                importer &lt;- function(x) readr::read_tsv(x, progress = FALSE, </span>\n",
       "<span style=white-space:pre-wrap>                  col_types = col.types)</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (type == c(\"stringtie\")) {</span>\n",
       "<span style=white-space:pre-wrap>        txIdCol &lt;- \"t_name\"</span>\n",
       "<span style=white-space:pre-wrap>        geneIdCol &lt;- \"gene_name\"</span>\n",
       "<span style=white-space:pre-wrap>        abundanceCol &lt;- \"FPKM\"</span>\n",
       "<span style=white-space:pre-wrap>        countsCol &lt;- \"cov\"</span>\n",
       "<span style=white-space:pre-wrap>        lengthCol &lt;- \"length\"</span>\n",
       "<span style=white-space:pre-wrap>        if (readrStatus &amp; is.null(importer)) {</span>\n",
       "<span style=white-space:pre-wrap>            col.types &lt;- readr::cols(readr::col_character(), </span>\n",
       "<span style=white-space:pre-wrap>                readr::col_character(), readr::col_character(), </span>\n",
       "<span style=white-space:pre-wrap>                readr::col_integer(), readr::col_integer(), readr::col_character(), </span>\n",
       "<span style=white-space:pre-wrap>                readr::col_integer(), readr::col_integer(), readr::col_character(), </span>\n",
       "<span style=white-space:pre-wrap>                readr::col_character(), readr::col_double(), </span>\n",
       "<span style=white-space:pre-wrap>                readr::col_double())</span>\n",
       "<span style=white-space:pre-wrap>            importer &lt;- function(x) readr::read_tsv(x, progress = FALSE, </span>\n",
       "<span style=white-space:pre-wrap>                col_types = col.types)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    infRepType &lt;- \"none\"</span>\n",
       "<span style=white-space:pre-wrap>    if (type %in% c(\"salmon\", \"sailfish\", \"kallisto\") &amp; !dropInfReps) {</span>\n",
       "<span style=white-space:pre-wrap>        infRepType &lt;- if (varReduce &amp; txOut) {</span>\n",
       "<span style=white-space:pre-wrap>            \"var\"</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            \"full\"</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!txIn) {</span>\n",
       "<span style=white-space:pre-wrap>        txi &lt;- computeRsemGeneLevel(files, importer, geneIdCol, </span>\n",
       "<span style=white-space:pre-wrap>            abundanceCol, countsCol, lengthCol, countsFromAbundance)</span>\n",
       "<span style=white-space:pre-wrap>        message(\"\")</span>\n",
       "<span style=white-space:pre-wrap>        return(txi)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (is.null(tx2gene) &amp; !txOut) {</span>\n",
       "<span style=white-space:pre-wrap>        summarizeFail()</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    repInfo &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>    if (infRepType != \"none\") {</span>\n",
       "<span style=white-space:pre-wrap>        repInfo &lt;- infRepImporter(dirname(files[1]))</span>\n",
       "<span style=white-space:pre-wrap>        if (is.null(repInfo)) {</span>\n",
       "<span style=white-space:pre-wrap>            infRepType &lt;- \"none\"</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (sparse) {</span>\n",
       "<span style=white-space:pre-wrap>        if (!requireNamespace(\"Matrix\", quietly = TRUE)) {</span>\n",
       "<span style=white-space:pre-wrap>            stop(\"sparse import requires core R package `Matrix`\")</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        message(\"importing sparsely, only counts and abundances returned, support limited to\\ntxOut=TRUE, CFA either 'no' or 'scaledTPM', and no inferential replicates\")</span>\n",
       "<span style=white-space:pre-wrap>        stopifnot(txOut)</span>\n",
       "<span style=white-space:pre-wrap>        stopifnot(infRepType == \"none\")</span>\n",
       "<span style=white-space:pre-wrap>        stopifnot(countsFromAbundance %in% c(\"no\", \"scaledTPM\"))</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    for (i in seq_along(files)) {</span>\n",
       "<span style=white-space:pre-wrap>        message(i, \" \", appendLF = FALSE)</span>\n",
       "<span style=white-space:pre-wrap>        raw &lt;- as.data.frame(importer(files[i]))</span>\n",
       "<span style=white-space:pre-wrap>        if (infRepType != \"none\") {</span>\n",
       "<span style=white-space:pre-wrap>            repInfo &lt;- infRepImporter(dirname(files[i]))</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            repInfo &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        stopifnot(all(c(abundanceCol, countsCol, lengthCol) %in% </span>\n",
       "<span style=white-space:pre-wrap>            names(raw)))</span>\n",
       "<span style=white-space:pre-wrap>        if (i == 1) {</span>\n",
       "<span style=white-space:pre-wrap>            txId &lt;- raw[[txIdCol]]</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            stopifnot(all(txId == raw[[txIdCol]]))</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        if (!sparse) {</span>\n",
       "<span style=white-space:pre-wrap>            if (i == 1) {</span>\n",
       "<span style=white-space:pre-wrap>                mat &lt;- matrix(nrow = nrow(raw), ncol = length(files))</span>\n",
       "<span style=white-space:pre-wrap>                rownames(mat) &lt;- raw[[txIdCol]]</span>\n",
       "<span style=white-space:pre-wrap>                colnames(mat) &lt;- names(files)</span>\n",
       "<span style=white-space:pre-wrap>                abundanceMatTx &lt;- mat</span>\n",
       "<span style=white-space:pre-wrap>                countsMatTx &lt;- mat</span>\n",
       "<span style=white-space:pre-wrap>                lengthMatTx &lt;- mat</span>\n",
       "<span style=white-space:pre-wrap>                if (infRepType == \"var\") {</span>\n",
       "<span style=white-space:pre-wrap>                  varMatTx &lt;- mat</span>\n",
       "<span style=white-space:pre-wrap>                }</span>\n",
       "<span style=white-space:pre-wrap>                else if (infRepType == \"full\") {</span>\n",
       "<span style=white-space:pre-wrap>                  infRepMatTx &lt;- list()</span>\n",
       "<span style=white-space:pre-wrap>                }</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            abundanceMatTx[, i] &lt;- raw[[abundanceCol]]</span>\n",
       "<span style=white-space:pre-wrap>            countsMatTx[, i] &lt;- raw[[countsCol]]</span>\n",
       "<span style=white-space:pre-wrap>            lengthMatTx[, i] &lt;- raw[[lengthCol]]</span>\n",
       "<span style=white-space:pre-wrap>            if (infRepType == \"var\") {</span>\n",
       "<span style=white-space:pre-wrap>                varMatTx[, i] &lt;- repInfo$vars</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            else if (infRepType == \"full\") {</span>\n",
       "<span style=white-space:pre-wrap>                infRepMatTx[[i]] &lt;- repInfo$reps</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            if (!is.null(infRepStat)) {</span>\n",
       "<span style=white-space:pre-wrap>                countsMatTx[, i] &lt;- infRepStat(repInfo$reps)</span>\n",
       "<span style=white-space:pre-wrap>                tpm &lt;- countsMatTx[, i]/lengthMatTx[, i]</span>\n",
       "<span style=white-space:pre-wrap>                abundanceMatTx[, i] &lt;- tpm * 1e+06/sum(tpm)</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            if (i == 1) {</span>\n",
       "<span style=white-space:pre-wrap>                txId &lt;- raw[[txIdCol]]</span>\n",
       "<span style=white-space:pre-wrap>                countsListI &lt;- list()</span>\n",
       "<span style=white-space:pre-wrap>                countsListX &lt;- list()</span>\n",
       "<span style=white-space:pre-wrap>                abundanceListX &lt;- list()</span>\n",
       "<span style=white-space:pre-wrap>                numNonzero &lt;- c()</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            stopifnot(all(txId == raw[[txIdCol]]))</span>\n",
       "<span style=white-space:pre-wrap>            sparse.idx &lt;- which(raw[[countsCol]] &gt;= sparseThreshold)</span>\n",
       "<span style=white-space:pre-wrap>            countsListI &lt;- c(countsListI, sparse.idx)</span>\n",
       "<span style=white-space:pre-wrap>            countsListX &lt;- c(countsListX, raw[[countsCol]][sparse.idx])</span>\n",
       "<span style=white-space:pre-wrap>            numNonzero &lt;- c(numNonzero, length(sparse.idx))</span>\n",
       "<span style=white-space:pre-wrap>            if (countsFromAbundance == \"scaledTPM\") {</span>\n",
       "<span style=white-space:pre-wrap>                abundanceListX &lt;- c(abundanceListX, raw[[abundanceCol]][sparse.idx])</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (sparse) {</span>\n",
       "<span style=white-space:pre-wrap>        countsMatTx &lt;- Matrix::sparseMatrix(i = unlist(countsListI), </span>\n",
       "<span style=white-space:pre-wrap>            j = rep(seq_along(numNonzero), numNonzero), x = unlist(countsListX), </span>\n",
       "<span style=white-space:pre-wrap>            dims = c(length(txId), length(files)), dimnames = list(txId, </span>\n",
       "<span style=white-space:pre-wrap>                names(files)))</span>\n",
       "<span style=white-space:pre-wrap>        if (countsFromAbundance == \"scaledTPM\") {</span>\n",
       "<span style=white-space:pre-wrap>            abundanceMatTx &lt;- Matrix::sparseMatrix(i = unlist(countsListI), </span>\n",
       "<span style=white-space:pre-wrap>                j = rep(seq_along(numNonzero), numNonzero), x = unlist(abundanceListX), </span>\n",
       "<span style=white-space:pre-wrap>                dims = c(length(txId), length(files)), dimnames = list(txId, </span>\n",
       "<span style=white-space:pre-wrap>                  names(files)))</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            abundanceMatTx &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        lengthMatTx &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (infRepType == \"full\") {</span>\n",
       "<span style=white-space:pre-wrap>        if (length(infRepMatTx) != length(files)) {</span>\n",
       "<span style=white-space:pre-wrap>            stop(\"Note: not all samples contain inferential replicates.\\n  tximport can only import data when either all or no samples\\n  contain inferential replicates. Instead first subset to the\\n  set of samples that all contain inferential replicates.\")</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        names(infRepMatTx) &lt;- names(files)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    message(\"\")</span>\n",
       "<span style=white-space:pre-wrap>    if (infRepType == \"none\") {</span>\n",
       "<span style=white-space:pre-wrap>        txi &lt;- list(abundance = abundanceMatTx, counts = countsMatTx, </span>\n",
       "<span style=white-space:pre-wrap>            length = lengthMatTx, countsFromAbundance = countsFromAbundance)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else if (infRepType == \"var\") {</span>\n",
       "<span style=white-space:pre-wrap>        txi &lt;- list(abundance = abundanceMatTx, counts = countsMatTx, </span>\n",
       "<span style=white-space:pre-wrap>            variance = varMatTx, length = lengthMatTx, countsFromAbundance = countsFromAbundance)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else if (infRepType == \"full\") {</span>\n",
       "<span style=white-space:pre-wrap>        txi &lt;- list(abundance = abundanceMatTx, counts = countsMatTx, </span>\n",
       "<span style=white-space:pre-wrap>            infReps = infRepMatTx, length = lengthMatTx, countsFromAbundance = countsFromAbundance)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (type == \"stringtie\") {</span>\n",
       "<span style=white-space:pre-wrap>        txi$counts &lt;- txi$counts * txi$length/readLength</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (type == \"rsem\") {</span>\n",
       "<span style=white-space:pre-wrap>        txi$length[txi$length &lt; 1] &lt;- 1</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (txOut) {</span>\n",
       "<span style=white-space:pre-wrap>        if (countsFromAbundance != \"no\") {</span>\n",
       "<span style=white-space:pre-wrap>            length4CFA &lt;- txi$length</span>\n",
       "<span style=white-space:pre-wrap>            if (countsFromAbundance == \"dtuScaledTPM\") {</span>\n",
       "<span style=white-space:pre-wrap>                length4CFA &lt;- medianLengthOverIsoform(length4CFA, </span>\n",
       "<span style=white-space:pre-wrap>                  tx2gene, ignoreTxVersion, ignoreAfterBar)</span>\n",
       "<span style=white-space:pre-wrap>                countsFromAbundance &lt;- \"lengthScaledTPM\"</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            txi$counts &lt;- makeCountsFromAbundance(countsMat = txi$counts, </span>\n",
       "<span style=white-space:pre-wrap>                abundanceMat = txi$abundance, lengthMat = length4CFA, </span>\n",
       "<span style=white-space:pre-wrap>                countsFromAbundance = countsFromAbundance)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        if (ignoreAfterBar) {</span>\n",
       "<span style=white-space:pre-wrap>            for (matNm in c(\"counts\", \"abundance\", \"length\")) {</span>\n",
       "<span style=white-space:pre-wrap>                rowNms &lt;- rownames(txi[[matNm]])</span>\n",
       "<span style=white-space:pre-wrap>                rownames(txi[[matNm]]) &lt;- sub(\"\\\\|.*\", \"\", rowNms)</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        return(txi)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    txi[[\"countsFromAbundance\"]] &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>    txiGene &lt;- summarizeToGene(txi, tx2gene, varReduce, ignoreTxVersion, </span>\n",
       "<span style=white-space:pre-wrap>        ignoreAfterBar, countsFromAbundance)</span>\n",
       "<span style=white-space:pre-wrap>    return(txiGene)</span>\n",
       "}</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (files, type = c(\"none\", \"salmon\", \"sailfish\", \"alevin\", \n",
       "    \"kallisto\", \"rsem\", \"stringtie\"), txIn = TRUE, txOut = FALSE, \n",
       "    countsFromAbundance = c(\"no\", \"scaledTPM\", \"lengthScaledTPM\", \n",
       "        \"dtuScaledTPM\"), tx2gene = NULL, varReduce = FALSE, dropInfReps = FALSE, \n",
       "    infRepStat = NULL, ignoreTxVersion = FALSE, ignoreAfterBar = FALSE, \n",
       "    geneIdCol, txIdCol, abundanceCol, countsCol, lengthCol, importer = NULL, \n",
       "    existenceOptional = FALSE, sparse = FALSE, sparseThreshold = 1, \n",
       "    readLength = 75, alevinArgs = NULL) \n",
       "\\{\n",
       "    infRepImporter <- NULL\n",
       "    type <- match.arg(type)\n",
       "    countsFromAbundance <- match.arg(countsFromAbundance)\n",
       "    if (countsFromAbundance == \"dtuScaledTPM\") \\{\n",
       "        stopifnot(txOut)\n",
       "        if (is.null(tx2gene)) \n",
       "            stop(\"'dtuScaledTPM' requires 'tx2gene' input\")\n",
       "    \\}\n",
       "    if (!existenceOptional) \n",
       "        stopifnot(all(file.exists(files)))\n",
       "    if (!txIn \\& txOut) \n",
       "        stop(\"txOut only an option when transcript-level data is read in (txIn=TRUE)\")\n",
       "    stopifnot(length(files) > 0)\n",
       "    kallisto.h5 <- basename(files{[}1{]}) == \"abundance.h5\"\n",
       "    if (type == \"kallisto\" \\& !kallisto.h5) \\{\n",
       "        message(\"Note: importing `abundance.h5` is typically faster than `abundance.tsv`\")\n",
       "    \\}\n",
       "    if (type == \"rsem\" \\& txIn \\& grepl(\"genes\", files{[}1{]})) \\{\n",
       "        message(\"It looks like you are importing RSEM genes.results files, setting txIn=FALSE\")\n",
       "        txIn <- FALSE\n",
       "    \\}\n",
       "    if (!is.null(infRepStat)) \\{\n",
       "        if (dropInfReps) \n",
       "            stop(\"infRepStat requires infReps\")\n",
       "        if (type == \"alevin\") \n",
       "            stop(\"infRepStat does not currently work with alevin input\")\n",
       "        if (sparse) \n",
       "            stop(\"infRepStat does not currently work with sparse output\")\n",
       "    \\}\n",
       "    if (type == \"alevin\") \\{\n",
       "        if (is.null(alevinArgs)) \\{\n",
       "            alevinArgs <- list(filterBarcodes = FALSE, tierImport = FALSE, \n",
       "                forceSlow = FALSE, dropMeanVar = FALSE)\n",
       "        \\}\n",
       "        stopifnot(is(alevinArgs, \"list\"))\n",
       "        stopifnot(all(sapply(alevinArgs, is.logical)))\n",
       "        alevinArgNms <- c(\"filterBarcodes\", \"tierImport\", \"forceSlow\", \n",
       "            \"dropMeanVar\")\n",
       "        stopifnot(all(names(alevinArgs) \\%in\\% alevinArgNms))\n",
       "        filterBarcodes <- if (is.null(alevinArgs\\$filterBarcodes)) \n",
       "            FALSE\n",
       "        else alevinArgs\\$filterBarcodes\n",
       "        tierImport <- if (is.null(alevinArgs\\$tierImport)) \n",
       "            FALSE\n",
       "        else alevinArgs\\$tierImport\n",
       "        forceSlow <- if (is.null(alevinArgs\\$forceSlow)) \n",
       "            FALSE\n",
       "        else alevinArgs\\$forceSlow\n",
       "        dropMeanVar <- if (is.null(alevinArgs\\$dropMeanVar)) \n",
       "            FALSE\n",
       "        else alevinArgs\\$dropMeanVar\n",
       "        if (length(files) > 1) \n",
       "            stop(\"alevin import currently only supports a single experiment\")\n",
       "        if (compareVersion(getAlevinVersion(files), \"0.14.0\") == \n",
       "            -1) \\{\n",
       "            stop(\"use of tximport version >= 1.18 requires alevin version >= 0.14\")\n",
       "        \\}\n",
       "        mat <- readAlevin(files, dropInfReps, filterBarcodes, \n",
       "            tierImport, forceSlow, dropMeanVar)\n",
       "        if (!is.list(mat)) \\{\n",
       "            txi <- list(abundance = NULL, counts = mat)\n",
       "        \\}\n",
       "        else \\{\n",
       "            txi <- list(abundance = NULL, counts = mat\\$counts, \n",
       "                mean = mat\\$mean, variance = mat\\$variance)\n",
       "            if (tierImport) \\{\n",
       "                txi\\$tier <- mat\\$tier\n",
       "            \\}\n",
       "            if (\"infReps\" \\%in\\% names(mat)) \\{\n",
       "                txi\\$infReps <- mat\\$infReps\n",
       "            \\}\n",
       "        \\}\n",
       "        txi\\$length <- NULL\n",
       "        txi\\$countsFromAbundance = \"no\"\n",
       "        return(txi)\n",
       "    \\}\n",
       "    readrStatus <- FALSE\n",
       "    if (is.null(importer) \\& !kallisto.h5) \\{\n",
       "        if (!requireNamespace(\"readr\", quietly = TRUE)) \\{\n",
       "            message(\"reading in files with read.delim (install 'readr' package for speed up)\")\n",
       "            importer <- read.delim\n",
       "        \\}\n",
       "        else \\{\n",
       "            timeZoneMissing <- length(suppressWarnings(OlsonNames())) == \n",
       "                0 \\textbar{} is.na(Sys.timezone())\n",
       "            if (timeZoneMissing) \\{\n",
       "                message(\"reading in files with read.delim ('readr' installed but won't work w/o timezones)\")\n",
       "                importer <- read.delim\n",
       "            \\}\n",
       "            else \\{\n",
       "                message(\"reading in files with read\\_tsv\")\n",
       "                readrStatus <- TRUE\n",
       "            \\}\n",
       "        \\}\n",
       "    \\}\n",
       "    if (type \\%in\\% c(\"salmon\", \"sailfish\")) \\{\n",
       "        txIdCol <- \"Name\"\n",
       "        abundanceCol <- \"TPM\"\n",
       "        countsCol <- \"NumReads\"\n",
       "        lengthCol <- \"EffectiveLength\"\n",
       "        if (readrStatus \\& is.null(importer)) \\{\n",
       "            col.types <- readr::cols(readr::col\\_character(), \n",
       "                readr::col\\_integer(), readr::col\\_double(), readr::col\\_double(), \n",
       "                readr::col\\_double())\n",
       "            importer <- function(x) readr::read\\_tsv(x, progress = FALSE, \n",
       "                col\\_types = col.types)\n",
       "        \\}\n",
       "        infRepImporter <- if (dropInfReps) \\{\n",
       "            NULL\n",
       "        \\}\n",
       "        else \\{\n",
       "            function(x) readInfRepFish(x, type)\n",
       "        \\}\n",
       "    \\}\n",
       "    if (type == \"kallisto\") \\{\n",
       "        txIdCol <- \"target\\_id\"\n",
       "        abundanceCol <- \"tpm\"\n",
       "        countsCol <- \"est\\_counts\"\n",
       "        lengthCol <- \"eff\\_length\"\n",
       "        if (kallisto.h5) \\{\n",
       "            importer <- read\\_kallisto\\_h5\n",
       "        \\}\n",
       "        else if (readrStatus \\& is.null(importer)) \\{\n",
       "            col.types <- readr::cols(readr::col\\_character(), \n",
       "                readr::col\\_integer(), readr::col\\_double(), readr::col\\_double(), \n",
       "                readr::col\\_double())\n",
       "            importer <- function(x) readr::read\\_tsv(x, progress = FALSE, \n",
       "                col\\_types = col.types)\n",
       "        \\}\n",
       "        infRepImporter <- if (dropInfReps) \\{\n",
       "            NULL\n",
       "        \\}\n",
       "        else \\{\n",
       "            readInfRepKallisto\n",
       "        \\}\n",
       "    \\}\n",
       "    if (type == \"rsem\") \\{\n",
       "        if (txIn) \\{\n",
       "            txIdCol <- \"transcript\\_id\"\n",
       "            abundanceCol <- \"TPM\"\n",
       "            countsCol <- \"expected\\_count\"\n",
       "            lengthCol <- \"effective\\_length\"\n",
       "            if (readrStatus \\& is.null(importer)) \\{\n",
       "                col.types <- readr::cols(readr::col\\_character(), \n",
       "                  readr::col\\_character(), readr::col\\_integer(), \n",
       "                  readr::col\\_double(), readr::col\\_double(), readr::col\\_double(), \n",
       "                  readr::col\\_double(), readr::col\\_double())\n",
       "                importer <- function(x) readr::read\\_tsv(x, progress = FALSE, \n",
       "                  col\\_types = col.types)\n",
       "            \\}\n",
       "        \\}\n",
       "        else \\{\n",
       "            geneIdCol <- \"gene\\_id\"\n",
       "            abundanceCol <- \"TPM\"\n",
       "            countsCol <- \"expected\\_count\"\n",
       "            lengthCol <- \"effective\\_length\"\n",
       "            if (readrStatus \\& is.null(importer)) \\{\n",
       "                col.types <- readr::cols(readr::col\\_character(), \n",
       "                  readr::col\\_character(), readr::col\\_double(), \n",
       "                  readr::col\\_double(), readr::col\\_double(), readr::col\\_double(), \n",
       "                  readr::col\\_double())\n",
       "                importer <- function(x) readr::read\\_tsv(x, progress = FALSE, \n",
       "                  col\\_types = col.types)\n",
       "            \\}\n",
       "        \\}\n",
       "    \\}\n",
       "    if (type == c(\"stringtie\")) \\{\n",
       "        txIdCol <- \"t\\_name\"\n",
       "        geneIdCol <- \"gene\\_name\"\n",
       "        abundanceCol <- \"FPKM\"\n",
       "        countsCol <- \"cov\"\n",
       "        lengthCol <- \"length\"\n",
       "        if (readrStatus \\& is.null(importer)) \\{\n",
       "            col.types <- readr::cols(readr::col\\_character(), \n",
       "                readr::col\\_character(), readr::col\\_character(), \n",
       "                readr::col\\_integer(), readr::col\\_integer(), readr::col\\_character(), \n",
       "                readr::col\\_integer(), readr::col\\_integer(), readr::col\\_character(), \n",
       "                readr::col\\_character(), readr::col\\_double(), \n",
       "                readr::col\\_double())\n",
       "            importer <- function(x) readr::read\\_tsv(x, progress = FALSE, \n",
       "                col\\_types = col.types)\n",
       "        \\}\n",
       "    \\}\n",
       "    infRepType <- \"none\"\n",
       "    if (type \\%in\\% c(\"salmon\", \"sailfish\", \"kallisto\") \\& !dropInfReps) \\{\n",
       "        infRepType <- if (varReduce \\& txOut) \\{\n",
       "            \"var\"\n",
       "        \\}\n",
       "        else \\{\n",
       "            \"full\"\n",
       "        \\}\n",
       "    \\}\n",
       "    if (!txIn) \\{\n",
       "        txi <- computeRsemGeneLevel(files, importer, geneIdCol, \n",
       "            abundanceCol, countsCol, lengthCol, countsFromAbundance)\n",
       "        message(\"\")\n",
       "        return(txi)\n",
       "    \\}\n",
       "    if (is.null(tx2gene) \\& !txOut) \\{\n",
       "        summarizeFail()\n",
       "    \\}\n",
       "    repInfo <- NULL\n",
       "    if (infRepType != \"none\") \\{\n",
       "        repInfo <- infRepImporter(dirname(files{[}1{]}))\n",
       "        if (is.null(repInfo)) \\{\n",
       "            infRepType <- \"none\"\n",
       "        \\}\n",
       "    \\}\n",
       "    if (sparse) \\{\n",
       "        if (!requireNamespace(\"Matrix\", quietly = TRUE)) \\{\n",
       "            stop(\"sparse import requires core R package `Matrix`\")\n",
       "        \\}\n",
       "        message(\"importing sparsely, only counts and abundances returned, support limited to\\textbackslash{}ntxOut=TRUE, CFA either 'no' or 'scaledTPM', and no inferential replicates\")\n",
       "        stopifnot(txOut)\n",
       "        stopifnot(infRepType == \"none\")\n",
       "        stopifnot(countsFromAbundance \\%in\\% c(\"no\", \"scaledTPM\"))\n",
       "    \\}\n",
       "    for (i in seq\\_along(files)) \\{\n",
       "        message(i, \" \", appendLF = FALSE)\n",
       "        raw <- as.data.frame(importer(files{[}i{]}))\n",
       "        if (infRepType != \"none\") \\{\n",
       "            repInfo <- infRepImporter(dirname(files{[}i{]}))\n",
       "        \\}\n",
       "        else \\{\n",
       "            repInfo <- NULL\n",
       "        \\}\n",
       "        stopifnot(all(c(abundanceCol, countsCol, lengthCol) \\%in\\% \n",
       "            names(raw)))\n",
       "        if (i == 1) \\{\n",
       "            txId <- raw{[}{[}txIdCol{]}{]}\n",
       "        \\}\n",
       "        else \\{\n",
       "            stopifnot(all(txId == raw{[}{[}txIdCol{]}{]}))\n",
       "        \\}\n",
       "        if (!sparse) \\{\n",
       "            if (i == 1) \\{\n",
       "                mat <- matrix(nrow = nrow(raw), ncol = length(files))\n",
       "                rownames(mat) <- raw{[}{[}txIdCol{]}{]}\n",
       "                colnames(mat) <- names(files)\n",
       "                abundanceMatTx <- mat\n",
       "                countsMatTx <- mat\n",
       "                lengthMatTx <- mat\n",
       "                if (infRepType == \"var\") \\{\n",
       "                  varMatTx <- mat\n",
       "                \\}\n",
       "                else if (infRepType == \"full\") \\{\n",
       "                  infRepMatTx <- list()\n",
       "                \\}\n",
       "            \\}\n",
       "            abundanceMatTx{[}, i{]} <- raw{[}{[}abundanceCol{]}{]}\n",
       "            countsMatTx{[}, i{]} <- raw{[}{[}countsCol{]}{]}\n",
       "            lengthMatTx{[}, i{]} <- raw{[}{[}lengthCol{]}{]}\n",
       "            if (infRepType == \"var\") \\{\n",
       "                varMatTx{[}, i{]} <- repInfo\\$vars\n",
       "            \\}\n",
       "            else if (infRepType == \"full\") \\{\n",
       "                infRepMatTx{[}{[}i{]}{]} <- repInfo\\$reps\n",
       "            \\}\n",
       "            if (!is.null(infRepStat)) \\{\n",
       "                countsMatTx{[}, i{]} <- infRepStat(repInfo\\$reps)\n",
       "                tpm <- countsMatTx{[}, i{]}/lengthMatTx{[}, i{]}\n",
       "                abundanceMatTx{[}, i{]} <- tpm * 1e+06/sum(tpm)\n",
       "            \\}\n",
       "        \\}\n",
       "        else \\{\n",
       "            if (i == 1) \\{\n",
       "                txId <- raw{[}{[}txIdCol{]}{]}\n",
       "                countsListI <- list()\n",
       "                countsListX <- list()\n",
       "                abundanceListX <- list()\n",
       "                numNonzero <- c()\n",
       "            \\}\n",
       "            stopifnot(all(txId == raw{[}{[}txIdCol{]}{]}))\n",
       "            sparse.idx <- which(raw{[}{[}countsCol{]}{]} >= sparseThreshold)\n",
       "            countsListI <- c(countsListI, sparse.idx)\n",
       "            countsListX <- c(countsListX, raw{[}{[}countsCol{]}{]}{[}sparse.idx{]})\n",
       "            numNonzero <- c(numNonzero, length(sparse.idx))\n",
       "            if (countsFromAbundance == \"scaledTPM\") \\{\n",
       "                abundanceListX <- c(abundanceListX, raw{[}{[}abundanceCol{]}{]}{[}sparse.idx{]})\n",
       "            \\}\n",
       "        \\}\n",
       "    \\}\n",
       "    if (sparse) \\{\n",
       "        countsMatTx <- Matrix::sparseMatrix(i = unlist(countsListI), \n",
       "            j = rep(seq\\_along(numNonzero), numNonzero), x = unlist(countsListX), \n",
       "            dims = c(length(txId), length(files)), dimnames = list(txId, \n",
       "                names(files)))\n",
       "        if (countsFromAbundance == \"scaledTPM\") \\{\n",
       "            abundanceMatTx <- Matrix::sparseMatrix(i = unlist(countsListI), \n",
       "                j = rep(seq\\_along(numNonzero), numNonzero), x = unlist(abundanceListX), \n",
       "                dims = c(length(txId), length(files)), dimnames = list(txId, \n",
       "                  names(files)))\n",
       "        \\}\n",
       "        else \\{\n",
       "            abundanceMatTx <- NULL\n",
       "        \\}\n",
       "        lengthMatTx <- NULL\n",
       "    \\}\n",
       "    if (infRepType == \"full\") \\{\n",
       "        if (length(infRepMatTx) != length(files)) \\{\n",
       "            stop(\"Note: not all samples contain inferential replicates.\\textbackslash{}n  tximport can only import data when either all or no samples\\textbackslash{}n  contain inferential replicates. Instead first subset to the\\textbackslash{}n  set of samples that all contain inferential replicates.\")\n",
       "        \\}\n",
       "        names(infRepMatTx) <- names(files)\n",
       "    \\}\n",
       "    message(\"\")\n",
       "    if (infRepType == \"none\") \\{\n",
       "        txi <- list(abundance = abundanceMatTx, counts = countsMatTx, \n",
       "            length = lengthMatTx, countsFromAbundance = countsFromAbundance)\n",
       "    \\}\n",
       "    else if (infRepType == \"var\") \\{\n",
       "        txi <- list(abundance = abundanceMatTx, counts = countsMatTx, \n",
       "            variance = varMatTx, length = lengthMatTx, countsFromAbundance = countsFromAbundance)\n",
       "    \\}\n",
       "    else if (infRepType == \"full\") \\{\n",
       "        txi <- list(abundance = abundanceMatTx, counts = countsMatTx, \n",
       "            infReps = infRepMatTx, length = lengthMatTx, countsFromAbundance = countsFromAbundance)\n",
       "    \\}\n",
       "    if (type == \"stringtie\") \\{\n",
       "        txi\\$counts <- txi\\$counts * txi\\$length/readLength\n",
       "    \\}\n",
       "    if (type == \"rsem\") \\{\n",
       "        txi\\$length{[}txi\\$length < 1{]} <- 1\n",
       "    \\}\n",
       "    if (txOut) \\{\n",
       "        if (countsFromAbundance != \"no\") \\{\n",
       "            length4CFA <- txi\\$length\n",
       "            if (countsFromAbundance == \"dtuScaledTPM\") \\{\n",
       "                length4CFA <- medianLengthOverIsoform(length4CFA, \n",
       "                  tx2gene, ignoreTxVersion, ignoreAfterBar)\n",
       "                countsFromAbundance <- \"lengthScaledTPM\"\n",
       "            \\}\n",
       "            txi\\$counts <- makeCountsFromAbundance(countsMat = txi\\$counts, \n",
       "                abundanceMat = txi\\$abundance, lengthMat = length4CFA, \n",
       "                countsFromAbundance = countsFromAbundance)\n",
       "        \\}\n",
       "        if (ignoreAfterBar) \\{\n",
       "            for (matNm in c(\"counts\", \"abundance\", \"length\")) \\{\n",
       "                rowNms <- rownames(txi{[}{[}matNm{]}{]})\n",
       "                rownames(txi{[}{[}matNm{]}{]}) <- sub(\"\\textbackslash{}\\textbackslash{}\\textbar{}.*\", \"\", rowNms)\n",
       "            \\}\n",
       "        \\}\n",
       "        return(txi)\n",
       "    \\}\n",
       "    txi{[}{[}\"countsFromAbundance\"{]}{]} <- NULL\n",
       "    txiGene <- summarizeToGene(txi, tx2gene, varReduce, ignoreTxVersion, \n",
       "        ignoreAfterBar, countsFromAbundance)\n",
       "    return(txiGene)\n",
       "\\}\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (files, type = c(\"none\", \"salmon\", \"sailfish\", \"alevin\", \n",
       "    \"kallisto\", \"rsem\", \"stringtie\"), txIn = TRUE, txOut = FALSE, \n",
       "    countsFromAbundance = c(\"no\", \"scaledTPM\", \"lengthScaledTPM\", \n",
       "        \"dtuScaledTPM\"), tx2gene = NULL, varReduce = FALSE, dropInfReps = FALSE, \n",
       "    infRepStat = NULL, ignoreTxVersion = FALSE, ignoreAfterBar = FALSE, \n",
       "    geneIdCol, txIdCol, abundanceCol, countsCol, lengthCol, importer = NULL, \n",
       "    existenceOptional = FALSE, sparse = FALSE, sparseThreshold = 1, \n",
       "    readLength = 75, alevinArgs = NULL) \n",
       "{\n",
       "    infRepImporter <- NULL\n",
       "    type <- match.arg(type)\n",
       "    countsFromAbundance <- match.arg(countsFromAbundance)\n",
       "    if (countsFromAbundance == \"dtuScaledTPM\") {\n",
       "        stopifnot(txOut)\n",
       "        if (is.null(tx2gene)) \n",
       "            stop(\"'dtuScaledTPM' requires 'tx2gene' input\")\n",
       "    }\n",
       "    if (!existenceOptional) \n",
       "        stopifnot(all(file.exists(files)))\n",
       "    if (!txIn & txOut) \n",
       "        stop(\"txOut only an option when transcript-level data is read in (txIn=TRUE)\")\n",
       "    stopifnot(length(files) > 0)\n",
       "    kallisto.h5 <- basename(files[1]) == \"abundance.h5\"\n",
       "    if (type == \"kallisto\" & !kallisto.h5) {\n",
       "        message(\"Note: importing `abundance.h5` is typically faster than `abundance.tsv`\")\n",
       "    }\n",
       "    if (type == \"rsem\" & txIn & grepl(\"genes\", files[1])) {\n",
       "        message(\"It looks like you are importing RSEM genes.results files, setting txIn=FALSE\")\n",
       "        txIn <- FALSE\n",
       "    }\n",
       "    if (!is.null(infRepStat)) {\n",
       "        if (dropInfReps) \n",
       "            stop(\"infRepStat requires infReps\")\n",
       "        if (type == \"alevin\") \n",
       "            stop(\"infRepStat does not currently work with alevin input\")\n",
       "        if (sparse) \n",
       "            stop(\"infRepStat does not currently work with sparse output\")\n",
       "    }\n",
       "    if (type == \"alevin\") {\n",
       "        if (is.null(alevinArgs)) {\n",
       "            alevinArgs <- list(filterBarcodes = FALSE, tierImport = FALSE, \n",
       "                forceSlow = FALSE, dropMeanVar = FALSE)\n",
       "        }\n",
       "        stopifnot(is(alevinArgs, \"list\"))\n",
       "        stopifnot(all(sapply(alevinArgs, is.logical)))\n",
       "        alevinArgNms <- c(\"filterBarcodes\", \"tierImport\", \"forceSlow\", \n",
       "            \"dropMeanVar\")\n",
       "        stopifnot(all(names(alevinArgs) %in% alevinArgNms))\n",
       "        filterBarcodes <- if (is.null(alevinArgs$filterBarcodes)) \n",
       "            FALSE\n",
       "        else alevinArgs$filterBarcodes\n",
       "        tierImport <- if (is.null(alevinArgs$tierImport)) \n",
       "            FALSE\n",
       "        else alevinArgs$tierImport\n",
       "        forceSlow <- if (is.null(alevinArgs$forceSlow)) \n",
       "            FALSE\n",
       "        else alevinArgs$forceSlow\n",
       "        dropMeanVar <- if (is.null(alevinArgs$dropMeanVar)) \n",
       "            FALSE\n",
       "        else alevinArgs$dropMeanVar\n",
       "        if (length(files) > 1) \n",
       "            stop(\"alevin import currently only supports a single experiment\")\n",
       "        if (compareVersion(getAlevinVersion(files), \"0.14.0\") == \n",
       "            -1) {\n",
       "            stop(\"use of tximport version >= 1.18 requires alevin version >= 0.14\")\n",
       "        }\n",
       "        mat <- readAlevin(files, dropInfReps, filterBarcodes, \n",
       "            tierImport, forceSlow, dropMeanVar)\n",
       "        if (!is.list(mat)) {\n",
       "            txi <- list(abundance = NULL, counts = mat)\n",
       "        }\n",
       "        else {\n",
       "            txi <- list(abundance = NULL, counts = mat$counts, \n",
       "                mean = mat$mean, variance = mat$variance)\n",
       "            if (tierImport) {\n",
       "                txi$tier <- mat$tier\n",
       "            }\n",
       "            if (\"infReps\" %in% names(mat)) {\n",
       "                txi$infReps <- mat$infReps\n",
       "            }\n",
       "        }\n",
       "        txi$length <- NULL\n",
       "        txi$countsFromAbundance = \"no\"\n",
       "        return(txi)\n",
       "    }\n",
       "    readrStatus <- FALSE\n",
       "    if (is.null(importer) & !kallisto.h5) {\n",
       "        if (!requireNamespace(\"readr\", quietly = TRUE)) {\n",
       "            message(\"reading in files with read.delim (install 'readr' package for speed up)\")\n",
       "            importer <- read.delim\n",
       "        }\n",
       "        else {\n",
       "            timeZoneMissing <- length(suppressWarnings(OlsonNames())) == \n",
       "                0 | is.na(Sys.timezone())\n",
       "            if (timeZoneMissing) {\n",
       "                message(\"reading in files with read.delim ('readr' installed but won't work w/o timezones)\")\n",
       "                importer <- read.delim\n",
       "            }\n",
       "            else {\n",
       "                message(\"reading in files with read_tsv\")\n",
       "                readrStatus <- TRUE\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "    if (type %in% c(\"salmon\", \"sailfish\")) {\n",
       "        txIdCol <- \"Name\"\n",
       "        abundanceCol <- \"TPM\"\n",
       "        countsCol <- \"NumReads\"\n",
       "        lengthCol <- \"EffectiveLength\"\n",
       "        if (readrStatus & is.null(importer)) {\n",
       "            col.types <- readr::cols(readr::col_character(), \n",
       "                readr::col_integer(), readr::col_double(), readr::col_double(), \n",
       "                readr::col_double())\n",
       "            importer <- function(x) readr::read_tsv(x, progress = FALSE, \n",
       "                col_types = col.types)\n",
       "        }\n",
       "        infRepImporter <- if (dropInfReps) {\n",
       "            NULL\n",
       "        }\n",
       "        else {\n",
       "            function(x) readInfRepFish(x, type)\n",
       "        }\n",
       "    }\n",
       "    if (type == \"kallisto\") {\n",
       "        txIdCol <- \"target_id\"\n",
       "        abundanceCol <- \"tpm\"\n",
       "        countsCol <- \"est_counts\"\n",
       "        lengthCol <- \"eff_length\"\n",
       "        if (kallisto.h5) {\n",
       "            importer <- read_kallisto_h5\n",
       "        }\n",
       "        else if (readrStatus & is.null(importer)) {\n",
       "            col.types <- readr::cols(readr::col_character(), \n",
       "                readr::col_integer(), readr::col_double(), readr::col_double(), \n",
       "                readr::col_double())\n",
       "            importer <- function(x) readr::read_tsv(x, progress = FALSE, \n",
       "                col_types = col.types)\n",
       "        }\n",
       "        infRepImporter <- if (dropInfReps) {\n",
       "            NULL\n",
       "        }\n",
       "        else {\n",
       "            readInfRepKallisto\n",
       "        }\n",
       "    }\n",
       "    if (type == \"rsem\") {\n",
       "        if (txIn) {\n",
       "            txIdCol <- \"transcript_id\"\n",
       "            abundanceCol <- \"TPM\"\n",
       "            countsCol <- \"expected_count\"\n",
       "            lengthCol <- \"effective_length\"\n",
       "            if (readrStatus & is.null(importer)) {\n",
       "                col.types <- readr::cols(readr::col_character(), \n",
       "                  readr::col_character(), readr::col_integer(), \n",
       "                  readr::col_double(), readr::col_double(), readr::col_double(), \n",
       "                  readr::col_double(), readr::col_double())\n",
       "                importer <- function(x) readr::read_tsv(x, progress = FALSE, \n",
       "                  col_types = col.types)\n",
       "            }\n",
       "        }\n",
       "        else {\n",
       "            geneIdCol <- \"gene_id\"\n",
       "            abundanceCol <- \"TPM\"\n",
       "            countsCol <- \"expected_count\"\n",
       "            lengthCol <- \"effective_length\"\n",
       "            if (readrStatus & is.null(importer)) {\n",
       "                col.types <- readr::cols(readr::col_character(), \n",
       "                  readr::col_character(), readr::col_double(), \n",
       "                  readr::col_double(), readr::col_double(), readr::col_double(), \n",
       "                  readr::col_double())\n",
       "                importer <- function(x) readr::read_tsv(x, progress = FALSE, \n",
       "                  col_types = col.types)\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "    if (type == c(\"stringtie\")) {\n",
       "        txIdCol <- \"t_name\"\n",
       "        geneIdCol <- \"gene_name\"\n",
       "        abundanceCol <- \"FPKM\"\n",
       "        countsCol <- \"cov\"\n",
       "        lengthCol <- \"length\"\n",
       "        if (readrStatus & is.null(importer)) {\n",
       "            col.types <- readr::cols(readr::col_character(), \n",
       "                readr::col_character(), readr::col_character(), \n",
       "                readr::col_integer(), readr::col_integer(), readr::col_character(), \n",
       "                readr::col_integer(), readr::col_integer(), readr::col_character(), \n",
       "                readr::col_character(), readr::col_double(), \n",
       "                readr::col_double())\n",
       "            importer <- function(x) readr::read_tsv(x, progress = FALSE, \n",
       "                col_types = col.types)\n",
       "        }\n",
       "    }\n",
       "    infRepType <- \"none\"\n",
       "    if (type %in% c(\"salmon\", \"sailfish\", \"kallisto\") & !dropInfReps) {\n",
       "        infRepType <- if (varReduce & txOut) {\n",
       "            \"var\"\n",
       "        }\n",
       "        else {\n",
       "            \"full\"\n",
       "        }\n",
       "    }\n",
       "    if (!txIn) {\n",
       "        txi <- computeRsemGeneLevel(files, importer, geneIdCol, \n",
       "            abundanceCol, countsCol, lengthCol, countsFromAbundance)\n",
       "        message(\"\")\n",
       "        return(txi)\n",
       "    }\n",
       "    if (is.null(tx2gene) & !txOut) {\n",
       "        summarizeFail()\n",
       "    }\n",
       "    repInfo <- NULL\n",
       "    if (infRepType != \"none\") {\n",
       "        repInfo <- infRepImporter(dirname(files[1]))\n",
       "        if (is.null(repInfo)) {\n",
       "            infRepType <- \"none\"\n",
       "        }\n",
       "    }\n",
       "    if (sparse) {\n",
       "        if (!requireNamespace(\"Matrix\", quietly = TRUE)) {\n",
       "            stop(\"sparse import requires core R package `Matrix`\")\n",
       "        }\n",
       "        message(\"importing sparsely, only counts and abundances returned, support limited to\\ntxOut=TRUE, CFA either 'no' or 'scaledTPM', and no inferential replicates\")\n",
       "        stopifnot(txOut)\n",
       "        stopifnot(infRepType == \"none\")\n",
       "        stopifnot(countsFromAbundance %in% c(\"no\", \"scaledTPM\"))\n",
       "    }\n",
       "    for (i in seq_along(files)) {\n",
       "        message(i, \" \", appendLF = FALSE)\n",
       "        raw <- as.data.frame(importer(files[i]))\n",
       "        if (infRepType != \"none\") {\n",
       "            repInfo <- infRepImporter(dirname(files[i]))\n",
       "        }\n",
       "        else {\n",
       "            repInfo <- NULL\n",
       "        }\n",
       "        stopifnot(all(c(abundanceCol, countsCol, lengthCol) %in% \n",
       "            names(raw)))\n",
       "        if (i == 1) {\n",
       "            txId <- raw[[txIdCol]]\n",
       "        }\n",
       "        else {\n",
       "            stopifnot(all(txId == raw[[txIdCol]]))\n",
       "        }\n",
       "        if (!sparse) {\n",
       "            if (i == 1) {\n",
       "                mat <- matrix(nrow = nrow(raw), ncol = length(files))\n",
       "                rownames(mat) <- raw[[txIdCol]]\n",
       "                colnames(mat) <- names(files)\n",
       "                abundanceMatTx <- mat\n",
       "                countsMatTx <- mat\n",
       "                lengthMatTx <- mat\n",
       "                if (infRepType == \"var\") {\n",
       "                  varMatTx <- mat\n",
       "                }\n",
       "                else if (infRepType == \"full\") {\n",
       "                  infRepMatTx <- list()\n",
       "                }\n",
       "            }\n",
       "            abundanceMatTx[, i] <- raw[[abundanceCol]]\n",
       "            countsMatTx[, i] <- raw[[countsCol]]\n",
       "            lengthMatTx[, i] <- raw[[lengthCol]]\n",
       "            if (infRepType == \"var\") {\n",
       "                varMatTx[, i] <- repInfo$vars\n",
       "            }\n",
       "            else if (infRepType == \"full\") {\n",
       "                infRepMatTx[[i]] <- repInfo$reps\n",
       "            }\n",
       "            if (!is.null(infRepStat)) {\n",
       "                countsMatTx[, i] <- infRepStat(repInfo$reps)\n",
       "                tpm <- countsMatTx[, i]/lengthMatTx[, i]\n",
       "                abundanceMatTx[, i] <- tpm * 1e+06/sum(tpm)\n",
       "            }\n",
       "        }\n",
       "        else {\n",
       "            if (i == 1) {\n",
       "                txId <- raw[[txIdCol]]\n",
       "                countsListI <- list()\n",
       "                countsListX <- list()\n",
       "                abundanceListX <- list()\n",
       "                numNonzero <- c()\n",
       "            }\n",
       "            stopifnot(all(txId == raw[[txIdCol]]))\n",
       "            sparse.idx <- which(raw[[countsCol]] >= sparseThreshold)\n",
       "            countsListI <- c(countsListI, sparse.idx)\n",
       "            countsListX <- c(countsListX, raw[[countsCol]][sparse.idx])\n",
       "            numNonzero <- c(numNonzero, length(sparse.idx))\n",
       "            if (countsFromAbundance == \"scaledTPM\") {\n",
       "                abundanceListX <- c(abundanceListX, raw[[abundanceCol]][sparse.idx])\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "    if (sparse) {\n",
       "        countsMatTx <- Matrix::sparseMatrix(i = unlist(countsListI), \n",
       "            j = rep(seq_along(numNonzero), numNonzero), x = unlist(countsListX), \n",
       "            dims = c(length(txId), length(files)), dimnames = list(txId, \n",
       "                names(files)))\n",
       "        if (countsFromAbundance == \"scaledTPM\") {\n",
       "            abundanceMatTx <- Matrix::sparseMatrix(i = unlist(countsListI), \n",
       "                j = rep(seq_along(numNonzero), numNonzero), x = unlist(abundanceListX), \n",
       "                dims = c(length(txId), length(files)), dimnames = list(txId, \n",
       "                  names(files)))\n",
       "        }\n",
       "        else {\n",
       "            abundanceMatTx <- NULL\n",
       "        }\n",
       "        lengthMatTx <- NULL\n",
       "    }\n",
       "    if (infRepType == \"full\") {\n",
       "        if (length(infRepMatTx) != length(files)) {\n",
       "            stop(\"Note: not all samples contain inferential replicates.\\n  tximport can only import data when either all or no samples\\n  contain inferential replicates. Instead first subset to the\\n  set of samples that all contain inferential replicates.\")\n",
       "        }\n",
       "        names(infRepMatTx) <- names(files)\n",
       "    }\n",
       "    message(\"\")\n",
       "    if (infRepType == \"none\") {\n",
       "        txi <- list(abundance = abundanceMatTx, counts = countsMatTx, \n",
       "            length = lengthMatTx, countsFromAbundance = countsFromAbundance)\n",
       "    }\n",
       "    else if (infRepType == \"var\") {\n",
       "        txi <- list(abundance = abundanceMatTx, counts = countsMatTx, \n",
       "            variance = varMatTx, length = lengthMatTx, countsFromAbundance = countsFromAbundance)\n",
       "    }\n",
       "    else if (infRepType == \"full\") {\n",
       "        txi <- list(abundance = abundanceMatTx, counts = countsMatTx, \n",
       "            infReps = infRepMatTx, length = lengthMatTx, countsFromAbundance = countsFromAbundance)\n",
       "    }\n",
       "    if (type == \"stringtie\") {\n",
       "        txi$counts <- txi$counts * txi$length/readLength\n",
       "    }\n",
       "    if (type == \"rsem\") {\n",
       "        txi$length[txi$length < 1] <- 1\n",
       "    }\n",
       "    if (txOut) {\n",
       "        if (countsFromAbundance != \"no\") {\n",
       "            length4CFA <- txi$length\n",
       "            if (countsFromAbundance == \"dtuScaledTPM\") {\n",
       "                length4CFA <- medianLengthOverIsoform(length4CFA, \n",
       "                  tx2gene, ignoreTxVersion, ignoreAfterBar)\n",
       "                countsFromAbundance <- \"lengthScaledTPM\"\n",
       "            }\n",
       "            txi$counts <- makeCountsFromAbundance(countsMat = txi$counts, \n",
       "                abundanceMat = txi$abundance, lengthMat = length4CFA, \n",
       "                countsFromAbundance = countsFromAbundance)\n",
       "        }\n",
       "        if (ignoreAfterBar) {\n",
       "            for (matNm in c(\"counts\", \"abundance\", \"length\")) {\n",
       "                rowNms <- rownames(txi[[matNm]])\n",
       "                rownames(txi[[matNm]]) <- sub(\"\\\\|.*\", \"\", rowNms)\n",
       "            }\n",
       "        }\n",
       "        return(txi)\n",
       "    }\n",
       "    txi[[\"countsFromAbundance\"]] <- NULL\n",
       "    txiGene <- summarizeToGene(txi, tx2gene, varReduce, ignoreTxVersion, \n",
       "        ignoreAfterBar, countsFromAbundance)\n",
       "    return(txiGene)\n",
       "}\n",
       "```"
      ],
      "text/plain": [
       "function (files, type = c(\"none\", \"salmon\", \"sailfish\", \"alevin\", \n",
       "    \"kallisto\", \"rsem\", \"stringtie\"), txIn = TRUE, txOut = FALSE, \n",
       "    countsFromAbundance = c(\"no\", \"scaledTPM\", \"lengthScaledTPM\", \n",
       "        \"dtuScaledTPM\"), tx2gene = NULL, varReduce = FALSE, dropInfReps = FALSE, \n",
       "    infRepStat = NULL, ignoreTxVersion = FALSE, ignoreAfterBar = FALSE, \n",
       "    geneIdCol, txIdCol, abundanceCol, countsCol, lengthCol, importer = NULL, \n",
       "    existenceOptional = FALSE, sparse = FALSE, sparseThreshold = 1, \n",
       "    readLength = 75, alevinArgs = NULL) \n",
       "{\n",
       "    infRepImporter <- NULL\n",
       "    type <- match.arg(type)\n",
       "    countsFromAbundance <- match.arg(countsFromAbundance)\n",
       "    if (countsFromAbundance == \"dtuScaledTPM\") {\n",
       "        stopifnot(txOut)\n",
       "        if (is.null(tx2gene)) \n",
       "            stop(\"'dtuScaledTPM' requires 'tx2gene' input\")\n",
       "    }\n",
       "    if (!existenceOptional) \n",
       "        stopifnot(all(file.exists(files)))\n",
       "    if (!txIn & txOut) \n",
       "        stop(\"txOut only an option when transcript-level data is read in (txIn=TRUE)\")\n",
       "    stopifnot(length(files) > 0)\n",
       "    kallisto.h5 <- basename(files[1]) == \"abundance.h5\"\n",
       "    if (type == \"kallisto\" & !kallisto.h5) {\n",
       "        message(\"Note: importing `abundance.h5` is typically faster than `abundance.tsv`\")\n",
       "    }\n",
       "    if (type == \"rsem\" & txIn & grepl(\"genes\", files[1])) {\n",
       "        message(\"It looks like you are importing RSEM genes.results files, setting txIn=FALSE\")\n",
       "        txIn <- FALSE\n",
       "    }\n",
       "    if (!is.null(infRepStat)) {\n",
       "        if (dropInfReps) \n",
       "            stop(\"infRepStat requires infReps\")\n",
       "        if (type == \"alevin\") \n",
       "            stop(\"infRepStat does not currently work with alevin input\")\n",
       "        if (sparse) \n",
       "            stop(\"infRepStat does not currently work with sparse output\")\n",
       "    }\n",
       "    if (type == \"alevin\") {\n",
       "        if (is.null(alevinArgs)) {\n",
       "            alevinArgs <- list(filterBarcodes = FALSE, tierImport = FALSE, \n",
       "                forceSlow = FALSE, dropMeanVar = FALSE)\n",
       "        }\n",
       "        stopifnot(is(alevinArgs, \"list\"))\n",
       "        stopifnot(all(sapply(alevinArgs, is.logical)))\n",
       "        alevinArgNms <- c(\"filterBarcodes\", \"tierImport\", \"forceSlow\", \n",
       "            \"dropMeanVar\")\n",
       "        stopifnot(all(names(alevinArgs) %in% alevinArgNms))\n",
       "        filterBarcodes <- if (is.null(alevinArgs$filterBarcodes)) \n",
       "            FALSE\n",
       "        else alevinArgs$filterBarcodes\n",
       "        tierImport <- if (is.null(alevinArgs$tierImport)) \n",
       "            FALSE\n",
       "        else alevinArgs$tierImport\n",
       "        forceSlow <- if (is.null(alevinArgs$forceSlow)) \n",
       "            FALSE\n",
       "        else alevinArgs$forceSlow\n",
       "        dropMeanVar <- if (is.null(alevinArgs$dropMeanVar)) \n",
       "            FALSE\n",
       "        else alevinArgs$dropMeanVar\n",
       "        if (length(files) > 1) \n",
       "            stop(\"alevin import currently only supports a single experiment\")\n",
       "        if (compareVersion(getAlevinVersion(files), \"0.14.0\") == \n",
       "            -1) {\n",
       "            stop(\"use of tximport version >= 1.18 requires alevin version >= 0.14\")\n",
       "        }\n",
       "        mat <- readAlevin(files, dropInfReps, filterBarcodes, \n",
       "            tierImport, forceSlow, dropMeanVar)\n",
       "        if (!is.list(mat)) {\n",
       "            txi <- list(abundance = NULL, counts = mat)\n",
       "        }\n",
       "        else {\n",
       "            txi <- list(abundance = NULL, counts = mat$counts, \n",
       "                mean = mat$mean, variance = mat$variance)\n",
       "            if (tierImport) {\n",
       "                txi$tier <- mat$tier\n",
       "            }\n",
       "            if (\"infReps\" %in% names(mat)) {\n",
       "                txi$infReps <- mat$infReps\n",
       "            }\n",
       "        }\n",
       "        txi$length <- NULL\n",
       "        txi$countsFromAbundance = \"no\"\n",
       "        return(txi)\n",
       "    }\n",
       "    readrStatus <- FALSE\n",
       "    if (is.null(importer) & !kallisto.h5) {\n",
       "        if (!requireNamespace(\"readr\", quietly = TRUE)) {\n",
       "            message(\"reading in files with read.delim (install 'readr' package for speed up)\")\n",
       "            importer <- read.delim\n",
       "        }\n",
       "        else {\n",
       "            timeZoneMissing <- length(suppressWarnings(OlsonNames())) == \n",
       "                0 | is.na(Sys.timezone())\n",
       "            if (timeZoneMissing) {\n",
       "                message(\"reading in files with read.delim ('readr' installed but won't work w/o timezones)\")\n",
       "                importer <- read.delim\n",
       "            }\n",
       "            else {\n",
       "                message(\"reading in files with read_tsv\")\n",
       "                readrStatus <- TRUE\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "    if (type %in% c(\"salmon\", \"sailfish\")) {\n",
       "        txIdCol <- \"Name\"\n",
       "        abundanceCol <- \"TPM\"\n",
       "        countsCol <- \"NumReads\"\n",
       "        lengthCol <- \"EffectiveLength\"\n",
       "        if (readrStatus & is.null(importer)) {\n",
       "            col.types <- readr::cols(readr::col_character(), \n",
       "                readr::col_integer(), readr::col_double(), readr::col_double(), \n",
       "                readr::col_double())\n",
       "            importer <- function(x) readr::read_tsv(x, progress = FALSE, \n",
       "                col_types = col.types)\n",
       "        }\n",
       "        infRepImporter <- if (dropInfReps) {\n",
       "            NULL\n",
       "        }\n",
       "        else {\n",
       "            function(x) readInfRepFish(x, type)\n",
       "        }\n",
       "    }\n",
       "    if (type == \"kallisto\") {\n",
       "        txIdCol <- \"target_id\"\n",
       "        abundanceCol <- \"tpm\"\n",
       "        countsCol <- \"est_counts\"\n",
       "        lengthCol <- \"eff_length\"\n",
       "        if (kallisto.h5) {\n",
       "            importer <- read_kallisto_h5\n",
       "        }\n",
       "        else if (readrStatus & is.null(importer)) {\n",
       "            col.types <- readr::cols(readr::col_character(), \n",
       "                readr::col_integer(), readr::col_double(), readr::col_double(), \n",
       "                readr::col_double())\n",
       "            importer <- function(x) readr::read_tsv(x, progress = FALSE, \n",
       "                col_types = col.types)\n",
       "        }\n",
       "        infRepImporter <- if (dropInfReps) {\n",
       "            NULL\n",
       "        }\n",
       "        else {\n",
       "            readInfRepKallisto\n",
       "        }\n",
       "    }\n",
       "    if (type == \"rsem\") {\n",
       "        if (txIn) {\n",
       "            txIdCol <- \"transcript_id\"\n",
       "            abundanceCol <- \"TPM\"\n",
       "            countsCol <- \"expected_count\"\n",
       "            lengthCol <- \"effective_length\"\n",
       "            if (readrStatus & is.null(importer)) {\n",
       "                col.types <- readr::cols(readr::col_character(), \n",
       "                  readr::col_character(), readr::col_integer(), \n",
       "                  readr::col_double(), readr::col_double(), readr::col_double(), \n",
       "                  readr::col_double(), readr::col_double())\n",
       "                importer <- function(x) readr::read_tsv(x, progress = FALSE, \n",
       "                  col_types = col.types)\n",
       "            }\n",
       "        }\n",
       "        else {\n",
       "            geneIdCol <- \"gene_id\"\n",
       "            abundanceCol <- \"TPM\"\n",
       "            countsCol <- \"expected_count\"\n",
       "            lengthCol <- \"effective_length\"\n",
       "            if (readrStatus & is.null(importer)) {\n",
       "                col.types <- readr::cols(readr::col_character(), \n",
       "                  readr::col_character(), readr::col_double(), \n",
       "                  readr::col_double(), readr::col_double(), readr::col_double(), \n",
       "                  readr::col_double())\n",
       "                importer <- function(x) readr::read_tsv(x, progress = FALSE, \n",
       "                  col_types = col.types)\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "    if (type == c(\"stringtie\")) {\n",
       "        txIdCol <- \"t_name\"\n",
       "        geneIdCol <- \"gene_name\"\n",
       "        abundanceCol <- \"FPKM\"\n",
       "        countsCol <- \"cov\"\n",
       "        lengthCol <- \"length\"\n",
       "        if (readrStatus & is.null(importer)) {\n",
       "            col.types <- readr::cols(readr::col_character(), \n",
       "                readr::col_character(), readr::col_character(), \n",
       "                readr::col_integer(), readr::col_integer(), readr::col_character(), \n",
       "                readr::col_integer(), readr::col_integer(), readr::col_character(), \n",
       "                readr::col_character(), readr::col_double(), \n",
       "                readr::col_double())\n",
       "            importer <- function(x) readr::read_tsv(x, progress = FALSE, \n",
       "                col_types = col.types)\n",
       "        }\n",
       "    }\n",
       "    infRepType <- \"none\"\n",
       "    if (type %in% c(\"salmon\", \"sailfish\", \"kallisto\") & !dropInfReps) {\n",
       "        infRepType <- if (varReduce & txOut) {\n",
       "            \"var\"\n",
       "        }\n",
       "        else {\n",
       "            \"full\"\n",
       "        }\n",
       "    }\n",
       "    if (!txIn) {\n",
       "        txi <- computeRsemGeneLevel(files, importer, geneIdCol, \n",
       "            abundanceCol, countsCol, lengthCol, countsFromAbundance)\n",
       "        message(\"\")\n",
       "        return(txi)\n",
       "    }\n",
       "    if (is.null(tx2gene) & !txOut) {\n",
       "        summarizeFail()\n",
       "    }\n",
       "    repInfo <- NULL\n",
       "    if (infRepType != \"none\") {\n",
       "        repInfo <- infRepImporter(dirname(files[1]))\n",
       "        if (is.null(repInfo)) {\n",
       "            infRepType <- \"none\"\n",
       "        }\n",
       "    }\n",
       "    if (sparse) {\n",
       "        if (!requireNamespace(\"Matrix\", quietly = TRUE)) {\n",
       "            stop(\"sparse import requires core R package `Matrix`\")\n",
       "        }\n",
       "        message(\"importing sparsely, only counts and abundances returned, support limited to\\ntxOut=TRUE, CFA either 'no' or 'scaledTPM', and no inferential replicates\")\n",
       "        stopifnot(txOut)\n",
       "        stopifnot(infRepType == \"none\")\n",
       "        stopifnot(countsFromAbundance %in% c(\"no\", \"scaledTPM\"))\n",
       "    }\n",
       "    for (i in seq_along(files)) {\n",
       "        message(i, \" \", appendLF = FALSE)\n",
       "        raw <- as.data.frame(importer(files[i]))\n",
       "        if (infRepType != \"none\") {\n",
       "            repInfo <- infRepImporter(dirname(files[i]))\n",
       "        }\n",
       "        else {\n",
       "            repInfo <- NULL\n",
       "        }\n",
       "        stopifnot(all(c(abundanceCol, countsCol, lengthCol) %in% \n",
       "            names(raw)))\n",
       "        if (i == 1) {\n",
       "            txId <- raw[[txIdCol]]\n",
       "        }\n",
       "        else {\n",
       "            stopifnot(all(txId == raw[[txIdCol]]))\n",
       "        }\n",
       "        if (!sparse) {\n",
       "            if (i == 1) {\n",
       "                mat <- matrix(nrow = nrow(raw), ncol = length(files))\n",
       "                rownames(mat) <- raw[[txIdCol]]\n",
       "                colnames(mat) <- names(files)\n",
       "                abundanceMatTx <- mat\n",
       "                countsMatTx <- mat\n",
       "                lengthMatTx <- mat\n",
       "                if (infRepType == \"var\") {\n",
       "                  varMatTx <- mat\n",
       "                }\n",
       "                else if (infRepType == \"full\") {\n",
       "                  infRepMatTx <- list()\n",
       "                }\n",
       "            }\n",
       "            abundanceMatTx[, i] <- raw[[abundanceCol]]\n",
       "            countsMatTx[, i] <- raw[[countsCol]]\n",
       "            lengthMatTx[, i] <- raw[[lengthCol]]\n",
       "            if (infRepType == \"var\") {\n",
       "                varMatTx[, i] <- repInfo$vars\n",
       "            }\n",
       "            else if (infRepType == \"full\") {\n",
       "                infRepMatTx[[i]] <- repInfo$reps\n",
       "            }\n",
       "            if (!is.null(infRepStat)) {\n",
       "                countsMatTx[, i] <- infRepStat(repInfo$reps)\n",
       "                tpm <- countsMatTx[, i]/lengthMatTx[, i]\n",
       "                abundanceMatTx[, i] <- tpm * 1e+06/sum(tpm)\n",
       "            }\n",
       "        }\n",
       "        else {\n",
       "            if (i == 1) {\n",
       "                txId <- raw[[txIdCol]]\n",
       "                countsListI <- list()\n",
       "                countsListX <- list()\n",
       "                abundanceListX <- list()\n",
       "                numNonzero <- c()\n",
       "            }\n",
       "            stopifnot(all(txId == raw[[txIdCol]]))\n",
       "            sparse.idx <- which(raw[[countsCol]] >= sparseThreshold)\n",
       "            countsListI <- c(countsListI, sparse.idx)\n",
       "            countsListX <- c(countsListX, raw[[countsCol]][sparse.idx])\n",
       "            numNonzero <- c(numNonzero, length(sparse.idx))\n",
       "            if (countsFromAbundance == \"scaledTPM\") {\n",
       "                abundanceListX <- c(abundanceListX, raw[[abundanceCol]][sparse.idx])\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "    if (sparse) {\n",
       "        countsMatTx <- Matrix::sparseMatrix(i = unlist(countsListI), \n",
       "            j = rep(seq_along(numNonzero), numNonzero), x = unlist(countsListX), \n",
       "            dims = c(length(txId), length(files)), dimnames = list(txId, \n",
       "                names(files)))\n",
       "        if (countsFromAbundance == \"scaledTPM\") {\n",
       "            abundanceMatTx <- Matrix::sparseMatrix(i = unlist(countsListI), \n",
       "                j = rep(seq_along(numNonzero), numNonzero), x = unlist(abundanceListX), \n",
       "                dims = c(length(txId), length(files)), dimnames = list(txId, \n",
       "                  names(files)))\n",
       "        }\n",
       "        else {\n",
       "            abundanceMatTx <- NULL\n",
       "        }\n",
       "        lengthMatTx <- NULL\n",
       "    }\n",
       "    if (infRepType == \"full\") {\n",
       "        if (length(infRepMatTx) != length(files)) {\n",
       "            stop(\"Note: not all samples contain inferential replicates.\\n  tximport can only import data when either all or no samples\\n  contain inferential replicates. Instead first subset to the\\n  set of samples that all contain inferential replicates.\")\n",
       "        }\n",
       "        names(infRepMatTx) <- names(files)\n",
       "    }\n",
       "    message(\"\")\n",
       "    if (infRepType == \"none\") {\n",
       "        txi <- list(abundance = abundanceMatTx, counts = countsMatTx, \n",
       "            length = lengthMatTx, countsFromAbundance = countsFromAbundance)\n",
       "    }\n",
       "    else if (infRepType == \"var\") {\n",
       "        txi <- list(abundance = abundanceMatTx, counts = countsMatTx, \n",
       "            variance = varMatTx, length = lengthMatTx, countsFromAbundance = countsFromAbundance)\n",
       "    }\n",
       "    else if (infRepType == \"full\") {\n",
       "        txi <- list(abundance = abundanceMatTx, counts = countsMatTx, \n",
       "            infReps = infRepMatTx, length = lengthMatTx, countsFromAbundance = countsFromAbundance)\n",
       "    }\n",
       "    if (type == \"stringtie\") {\n",
       "        txi$counts <- txi$counts * txi$length/readLength\n",
       "    }\n",
       "    if (type == \"rsem\") {\n",
       "        txi$length[txi$length < 1] <- 1\n",
       "    }\n",
       "    if (txOut) {\n",
       "        if (countsFromAbundance != \"no\") {\n",
       "            length4CFA <- txi$length\n",
       "            if (countsFromAbundance == \"dtuScaledTPM\") {\n",
       "                length4CFA <- medianLengthOverIsoform(length4CFA, \n",
       "                  tx2gene, ignoreTxVersion, ignoreAfterBar)\n",
       "                countsFromAbundance <- \"lengthScaledTPM\"\n",
       "            }\n",
       "            txi$counts <- makeCountsFromAbundance(countsMat = txi$counts, \n",
       "                abundanceMat = txi$abundance, lengthMat = length4CFA, \n",
       "                countsFromAbundance = countsFromAbundance)\n",
       "        }\n",
       "        if (ignoreAfterBar) {\n",
       "            for (matNm in c(\"counts\", \"abundance\", \"length\")) {\n",
       "                rowNms <- rownames(txi[[matNm]])\n",
       "                rownames(txi[[matNm]]) <- sub(\"\\\\|.*\", \"\", rowNms)\n",
       "            }\n",
       "        }\n",
       "        return(txi)\n",
       "    }\n",
       "    txi[[\"countsFromAbundance\"]] <- NULL\n",
       "    txiGene <- summarizeToGene(txi, tx2gene, varReduce, ignoreTxVersion, \n",
       "        ignoreAfterBar, countsFromAbundance)\n",
       "    return(txiGene)\n",
       "}\n",
       "<bytecode: 0x555599376dd0>\n",
       "<environment: namespace:tximport>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tximport"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d119c278-75b7-45e3-abb5-67a053799801",
   "metadata": {},
   "outputs": [],
   "source": [
    "install.packages(\"cowplot\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dd174a9-c0bd-4f55-afcf-ee4a03a8434c",
   "metadata": {},
   "outputs": [],
   "source": [
    "BiocManager::install(\"TxDb.Hsapiens.UCSC.hg38.knownGene\")\n",
    "library (TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene\n",
    "k <- keys(txdb, keytype = \"TXNAME\")\n",
    "tx2gene <- select(txdb, k, \"GENEID\", \"TXNAME\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b8a8225-747e-491f-accd-d23f035028e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"cowplot\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f11fb1c-4377-42a2-84ec-48029872462a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if (!require(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "BiocManager::install(version = \"3.14\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e3340e5-8b57-4eab-8191-d12240296a3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "BiocManager::install(\"biomaRt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae91c4a7-6dc4-4b7b-a0f3-445281753dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "BiocManager::install(\"DESeq2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48a60bc9-e583-4d3d-ad1c-0706740fe6ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DESeq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5d10168-4933-411c-98f7-289eb247447a",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"devtools\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05532e3f-6097-4f15-816b-6571a5993a91",
   "metadata": {},
   "source": [
    "#### First time through - Received a Warning that rhdf5 not available for this version of R\n",
    "\n",
    "Looked up our version and google searched R 4.1.1 rhdf5\n",
    "    \n",
    "Can install using `BiocManager::install(\"rhdf5\")`\n",
    "\n",
    "Pactherlab says to install `*rhdf5*` first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a43a9e1-7ce6-4b68-860b-5e3e1876ab3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "BiocManager::install(\"rhdf5\", force=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7b03ce8-e5d8-4be8-be2b-e40c428d4541",
   "metadata": {},
   "outputs": [],
   "source": [
    "library (rhdf5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09f66a8d-241c-4f86-8b23-3fcf14859d4f",
   "metadata": {},
   "source": [
    "#### Issues\n",
    "\n",
    "Noted in issues https://github.com/pachterlab/sleuth/issues/259 -- follow the instructions from [Paast](https://github.com/pachterlab/sleuth/issues/259#issuecomment-966270599)\n",
    "\n",
    "Install rhdf5 as noted above.\n",
    "\n",
    "Load the library\n",
    "\n",
    "##### Clone sleuth and install after editing the file\n",
    "\n",
    "Change directory to the top working directory in this etheral machine.\n",
    "\n",
    "```bash\n",
    "cd /sbgenomics/workspace\n",
    "```\n",
    "\n",
    "now clone the library \n",
    "\n",
    "```bash\n",
    "git clone https://github.com/pachterlab/sleuth.git\n",
    "```\n",
    "\n",
    "edit NAMESPACE as the instructions note - to remove the dependency remove the last line to remove the reference to **rhdf5**\n",
    "\n",
    "And then run the install.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cace19ad-5aea-475b-8fe2-d6cc962eeba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "devtools::install('../../sleuth/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a69aba07-0138-427f-a8e8-a28bd30b8456",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(sleuth)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5947268-42ff-49ef-903b-980d1427b68a",
   "metadata": {},
   "source": [
    "We have successfully run Kallisto with Kallisto Quantitation.\n",
    "\n",
    "Results may be found after running an application on Cavatica here:\n",
    "\n",
    "```bash\n",
    "/sbgenomics/project-files/\n",
    "```\n",
    "\n",
    "For this analysis we will use the results from the run using `metadata_ten_samples_only_txt`\n",
    "\n",
    "Results are in:\n",
    "\n",
    "```bash\n",
    "/sbgenomics/project-files/ten_samples_expression_matrix.tpm.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea7bc5b5-9e72-4bb8-92bd-d24143bb2479",
   "metadata": {},
   "source": [
    "### Parsing metadata\n",
    "\n",
    "A sleuth analysis is dependent on a metadata file, which describes the experimental design, the sample names, conditions and covariates. The metadata file is external to sleuth, and must be prepared prior to analysis. A metadata file should have been downloaded along with the kallisto quantifications. The first step in a sleuth analysis is loading of the metadata file. You might need the path in read_table below to where you have downloaded the kallisto dataset, so that the path directs to the sample_table.txt. We then select the relevant columns of the metadata.\n",
    "\n",
    "In our case, I used:\n",
    "\n",
    "```bash\n",
    "/sbgenomics/project-files/metadata_ten_samples_only.csv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a430a324-b6b5-4255-81eb-b15734e115bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- read.table('/sbgenomics/project-files/metadata_ten_samples_only.csv', sep=\",\", header=TRUE, stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "809234a4-57cf-441d-a2c4-e1fd97127e90",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(metadata, n=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83d55633-bf9e-4b00-90d0-ef75788cef3a",
   "metadata": {},
   "source": [
    "There is an error in the last sample detail - where the paired should read `2` and not NA.  So I copied the file to a local directory and corrected it -- it is corrected permanently now - but for this run through you can see:\n",
    "```bash\n",
    "cp /sbgenomics/project-files/metadata_ten_samples_only.csv /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data\n",
    "```\n",
    "\n",
    "where I edited the file and now will read this one in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d94ba4a-a191-494e-b2ed-d589c47843c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- read.table('/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/metadata_ten_samples_only.csv', sep=\",\", header=TRUE, stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf79d1bc-1a14-48c8-a57a-83f8116bd9ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9209ebbd-dcfa-49b7-bb8e-413004406e8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dim(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f907e43b-7a13-47bf-8aad-dc88eb09f022",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- dplyr::select(metadata, c('Case.ID', 'Sample.ID', 'Gender', 'Disease.type', 'Abundance'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fea7756a-b5e7-4707-909c-10de1fefc960",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9be16242-cd08-413b-a17e-8958d431c4f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- dplyr::distinct(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b4a9c46-262d-4daa-9c61-7c43ecde421d",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "679e4004-878f-4937-b888-ef9fd43f01fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "dim(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0656197e-93f0-4008-b3bd-35c14adf1522",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- dplyr::rename(metadata, sample = Sample.ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0617541a-0664-46bf-896e-4838af463e9a",
   "metadata": {},
   "source": [
    "Need to rename a colump as well to `path` where we have `Abundance`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11e98d5a-534a-4021-bfed-ffa7768cd8a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- dplyr::rename(metadata, path = Abundance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edc62a33-02e1-4fbb-9b80-6a53a66a41e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f8ecb14-d688-47ec-bbee-c84bdc5bdb4d",
   "metadata": {},
   "source": [
    "#### biomaRt - how to use\n",
    "\n",
    "Following instructions from the [ensembl site](https://grch37.ensembl.org/info/data/biomart/biomart_r_package.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b315a983-b7eb-47c5-a618-3558f1988f21",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(biomaRt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf6b61ce-eaf9-4ede-b84f-a498ab0e16eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "mart <- biomaRt::useMart(biomart=\"ensembl\", \n",
    "                     dataset = \"hsapiens_gene_ensembl\",\n",
    "                        host = \"https://useast.ensembl.org\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09a5d947-397f-46de-a8e6-88658fb08d32",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttg <- biomaRt::getBM(\n",
    "  attributes = c(\"ensembl_transcript_id\", \"transcript_version\",\n",
    "  \"ensembl_gene_id\", \"external_gene_name\", \"description\",\n",
    "  \"transcript_biotype\"),\n",
    "  mart = mart)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7386de62-67da-4ca9-a0d4-d5d7ad831a05",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttg <- dplyr::rename(ttg, target_id = ensembl_transcript_id,\n",
    "  ens_gene = ensembl_gene_id, ext_gene = external_gene_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0521385a-1bd1-402f-a9b0-53b52a1a4aae",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttg <- dplyr::select(ttg, c('target_id', 'ens_gene', 'ext_gene'))\n",
    "head(ttg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d913a41-9424-440e-8f8d-218c8001fd8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttg <- read.table('/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/ttg.csv', sep=\",\", header=TRUE, stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b09c920-8118-4f34-a303-68e07003a868",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(ttg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ad386e6-6c48-4e78-94f2-dc2b76581643",
   "metadata": {},
   "source": [
    "The resulting table contains Ensembl gene names (‘ens_gene’) and the associated transcripts (‘target_id’). Note that the gene-transcript mapping must be compatible with the transcriptome used with kallisto. In other words, to use Ensembl transcript-gene associations kallisto was run using the Ensembl transcriptome.\n",
    "\n",
    "#### Preparing the analysis\n",
    "\n",
    "The next step is to build a sleuth object. The sleuth object contains specification of the experimental design, a map describing grouping of transcripts into genes (or other groups), and a number of user specific parameters. In the example that follows, metadata is the experimental design and target_mapping describes the transcript groupings into genes previously constructed. Furthermore, we provide an aggregation_column, the column name of in ‘target_mapping’ table that is used to aggregate the transcripts. When both ‘target_mapping’ and ‘aggregation_column’ are provided, sleuth will automatically run in gene mode, returning gene differential expression results that came from the aggregation of transcript p-values.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d32d1dce-7746-4fee-a4ab-7af9d5a84aca",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttg      <- data.frame(ttg)\n",
    "metadata <- data.frame(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92622045-2bc5-4756-bf26-218a3244c69b",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(ttg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "127a5e33-2837-4687-ab45-8786e87e4bf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11c20432-066b-4499-add6-71ea21f8fcf7",
   "metadata": {},
   "source": [
    "#### Model (Design) Matrix Required\n",
    "We need to supply a model matrix -- and Sleuth implicitly uses DESeq2\n",
    "\n",
    "[How to use DESeq2](https://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)\n",
    "\n",
    "We have the following to compare condition effects of gender (Male, Female) with Disease type(TAM, DS-AML) in our cases.\n",
    "\n",
    "We have two groups (Male, Female) and two conditions (TAM, DS-AML)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d5e229a-213b-4b33-93f7-f6339b5fbafb",
   "metadata": {},
   "outputs": [],
   "source": [
    "group <- factor(metadata$Gender)\n",
    "group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31d9e6f1-af6d-47be-8b30-16a56c3cb5f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "condition <- factor(metadata$Disease.type)\n",
    "condition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31e11dd7-998d-46b0-aca7-16011214d33b",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_model <- model.matrix(~group + condition + group:condition)\n",
    "full_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "839a88a3-6c2f-457f-8b46-8a577f42de61",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_to_covariates = metadata\n",
    "target_mapping = ttg\n",
    "aggregation_column = \"ens_gene\"\n",
    "gene_mode = TRUE\n",
    "extra_bootstrap_summary = TRUE\n",
    "read_bootstrap_tpm = TRUE\n",
    "full_model = full_model\n",
    "normalize = TRUE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5946281-e3e5-4868-ba3f-7a06e83bd4f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "extra_opts <- list(gene_mode, extra_bootstrap_summary, read_bootstrap_tpm, full_model, normalize)\n",
    "names(extra_opts) <- c(\"gene_mode\",\n",
    "                       \"extra_bootstrap_summary\", \n",
    "                       \"read_bootstrap_tpm\", \n",
    "                       \"full_model\",\n",
    "                       \"normalize\")\n",
    "  if (\"extra_bootstrap_summary\" %in% names(extra_opts)) {\n",
    "    extra_bootstrap_summary <- extra_opts$extra_bootstrap_summary\n",
    "  } else {\n",
    "    extra_bootstrap_summary <- FALSE\n",
    "  }\n",
    "  if (\"read_bootstrap_tpm\" %in% names(extra_opts)) {\n",
    "    read_bootstrap_tpm <- extra_opts$read_bootstrap_tpm\n",
    "  } else {\n",
    "    read_bootstrap_tpm <- FALSE\n",
    "  }\n",
    "  if (\"max_bootstrap\" %in% names(extra_opts)) {\n",
    "    max_bootstrap <- extra_opts$max_bootstrap\n",
    "  } else {\n",
    "    max_bootstrap <- NULL\n",
    "  }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63eaaa55-a668-4ef6-911b-788472fd7e93",
   "metadata": {},
   "outputs": [],
   "source": [
    "extra_bootstrap_summary\n",
    "read_bootstrap_tpm\n",
    "max_bootstrap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae46d60c-45f5-4cd2-8e48-c561bcab177e",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(extra_opts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f090924-7335-4626-9bd0-a9e3d46db878",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- sleuth_prep(sample_to_covariates    = metadata, \n",
    "                  target_mapping          = ttg, \n",
    "                  aggregation_column      = 'ens_gene',\n",
    "                  gene_mode               = TRUE,\n",
    "                  extra_bootstrap_summary = TRUE,\n",
    "                  read_bootstrap_tpm      = TRUE,\n",
    "                  full_model              = full_model,\n",
    "                  normalize               = TRUE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "936b2528-36a5-47da-a298-efab33f94241",
   "metadata": {},
   "source": [
    " \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4442bf52-6431-47e9-877f-7987ad4055b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttg.df <- data.frame (ttg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfc64998-e3ff-4295-9cd7-21121236ef12",
   "metadata": {},
   "outputs": [],
   "source": [
    "  sample_to_covariates <- as.data.frame(sample_to_covariates)\n",
    "  sample_to_covariates$sample <- as.character(sample_to_covariates$sample)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0bb6456-3a3c-4f4e-a29a-e9daddacf867",
   "metadata": {},
   "outputs": [],
   "source": [
    "nrow(sample_to_covariates)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
