{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "97f9ac63-1c15-481c-af9a-ba032429149d",
   "metadata": {},
   "source": [
    "# Pediatric DS AML vs TAM\n",
    "\n",
    "Sleuth Analysis with Kallisto Quantitation Input\n",
    "\n",
    "## Following the Pachterlab walkthrough\n",
    "[Using p-value aggregation to obtain gene differential expression in datasets with multiple experimental conditions](\n",
    "https://pachterlab.github.io/sleuth_walkthroughs/pval_agg/analysis.html)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ac92837-8970-4820-8557-342a4993df84",
   "metadata": {},
   "source": [
    "### Getting started with Sleuth\n",
    "\n",
    "**sleuth** is a tool for the analysis and comparison of multiple related RNA-Seq experiments. Key features include:\n",
    "\n",
    "* The ability to perform both transcript-level and gene-level analysis.\n",
    "* Compatibility with kallisto enabling a fast and accurate workflow from reads to results.\n",
    "* The use of boostraps to ascertain and correct for technical variation in experiments.\n",
    "* An interactive app for exploratory data analysis.\n",
    "\n",
    "To use sleuth, RNA-Seq data must first be quantified with kallisto ( which we did so with the CAVATICA workflow), which is a program for very fast RNA-Seq quantification based on pseudo-alignment. An important feature of kallisto is that it outputs bootstraps along with the estimates of transcript abundances. These can serve as proxies for technical replicates, allowing for an ascertainment of the variability in estimates due to the random processes underlying RNA-Seq as well as the statistical procedure of read assignment. kallisto can quantify 30 million human reads in less than 3 minutes on a Mac desktop computer using only the read sequences and a transcriptome index that itself takes less than 10 minutes to build. sleuth has been designed to work seamlessly and efficiently with kallisto, and therefore RNA-Seq analysis with kallisto and sleuth is tractable on a laptop computer in a matter of minutes. More details about kallisto and sleuth are provided the papers describing the methods:\n",
    "\n",
    "#### Citations\n",
    "\n",
    "* Nicolas L Bray, Harold Pimentel, Páll Melsted and Lior Pachter, Near-optimal probabilistic RNA-seq quantification, Nature Biotechnology 34, 525–527 (2016), doi:10.1038/nbt.3519\n",
    "\n",
    "* Harold Pimentel, Nicolas L Bray, Suzette Puente, Páll Melsted and Lior Pachter, Differential analysis of RNA-seq incorporating quantification uncertainty, in press.\n",
    "\n",
    "sleuth has been designed to facilitate the exploration of RNA-Seq data by utilizing the Shiny web application framework by RStudio. The worked example below illustrates how to load data into sleuth and how to open Shiny plots for exploratory data analysis. The code underlying all plots is available via the Shiny interface so that analyses can be fully “open source”.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42a7cbed-f424-4492-b290-da466df30841",
   "metadata": {},
   "source": [
    "### Introduction\n",
    "\n",
    "Applying the same techniques used in the walk through - apply it to our own AML vs TAM\n",
    "\n",
    "We need to set up our working directory.\n",
    "And we need \n",
    "\n",
    "* `cowplot` - for making prettier plots and plots with grids.  and \n",
    "* `biomaRt` - for extracting the Ensembl transcript to gene mapping\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d119c278-75b7-45e3-abb5-67a053799801",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"cowplot\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9b8a8225-747e-491f-accd-d23f035028e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"cowplot\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3f11fb1c-4377-42a2-84ec-48029872462a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.18), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Installing package(s) 'BiocVersion'\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'backports', 'BiocGenerics', 'blob', 'brew', 'brio', 'broom',\n",
      "  'bslib', 'callr', 'caret', 'class', 'cli', 'clipr', 'colorspace',\n",
      "  'commonmark', 'conflicted', 'covr', 'cpp11', 'crayon', 'credentials',\n",
      "  'crosstalk', 'curl', 'data.table', 'DBI', 'dbplyr', 'desc', 'devtools',\n",
      "  'dials', 'diffobj', 'digest', 'dplyr', 'DT', 'dtplyr', 'e1071', 'evaluate',\n",
      "  'fansi', 'farver', 'forcats', 'foreach', 'forecast', 'fs', 'furrr', 'future',\n",
      "  'future.apply', 'gargle', 'generics', 'gert', 'ggplot2', 'gh', 'git2r',\n",
      "  'gitcreds', 'globals', 'glue', 'googlesheets4', 'gower', 'gtable', 'hardhat',\n",
      "  'haven', 'hms', 'htmltools', 'httpuv', 'httr', 'infer', 'ipred', 'IRdisplay',\n",
      "  'IRkernel', 'isoband', 'iterators', 'jsonlite', 'knitr', 'later', 'lattice',\n",
      "  'lhs', 'lifecycle', 'lmtest', 'lubridate', 'magrittr', 'MASS', 'Matrix',\n",
      "  'memoise', 'mgcv', 'mime', 'modeldata', 'modelr', 'nlme', 'nnet',\n",
      "  'objectProperties', 'objectSignals', 'openssl', 'parallelly', 'parsnip',\n",
      "  'patchwork', 'pbdZMQ', 'pillar', 'pkgbuild', 'pkgload', 'plyr', 'processx',\n",
      "  'progressr', 'proxy', 'ps', 'purrr', 'quantmod', 'randomForest', 'rcmdcheck',\n",
      "  'RColorBrewer', 'Rcpp', 'RcppArmadillo', 'RCurl', 'readr', 'readxl',\n",
      "  'recipes', 'remotes', 'repr', 'reprex', 'rex', 'rlang', 'rmarkdown', 'RODBC',\n",
      "  'roxygen2', 'rpart', 'rprojroot', 'rsample', 'RSQLite', 'rstudioapi',\n",
      "  'rversions', 'rvest', 'S4Vectors', 'sass', 'scales', 'sessioninfo',\n",
      "  'sevenbridges', 'shiny', 'stringi', 'stringr', 'survival', 'testthat',\n",
      "  'tibble', 'tidymodels', 'tidyr', 'tidyselect', 'tidyverse', 'timeDate',\n",
      "  'tinytex', 'tseries', 'TTR', 'tune', 'tzdb', 'urca', 'usethis', 'uuid',\n",
      "  'vctrs', 'viridisLite', 'vroom', 'waldo', 'withr', 'workflows',\n",
      "  'workflowsets', 'xfun', 'xml2', 'xts', 'yaml', 'yardstick', 'zip', 'zoo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!require(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "BiocManager::install(version = \"3.14\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4e3340e5-8b57-4eab-8191-d12240296a3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.18), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Installing package(s) 'biomaRt'\n",
      "\n",
      "also installing the dependencies ‘zlibbioc’, ‘GenomeInfoDbData’, ‘XVector’, ‘GenomeInfoDb’, ‘BiocGenerics’, ‘png’, ‘Biostrings’, ‘Biobase’, ‘IRanges’, ‘KEGGREST’, ‘filelock’, ‘XML’, ‘AnnotationDbi’, ‘BiocFileCache’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'backports', 'blob', 'brew', 'brio', 'broom', 'bslib', 'callr',\n",
      "  'caret', 'class', 'cli', 'clipr', 'colorspace', 'commonmark', 'conflicted',\n",
      "  'covr', 'cpp11', 'crayon', 'credentials', 'crosstalk', 'curl', 'data.table',\n",
      "  'DBI', 'dbplyr', 'desc', 'devtools', 'dials', 'diffobj', 'digest', 'dplyr',\n",
      "  'DT', 'dtplyr', 'e1071', 'evaluate', 'fansi', 'farver', 'forcats', 'foreach',\n",
      "  'forecast', 'fs', 'furrr', 'future', 'future.apply', 'gargle', 'generics',\n",
      "  'gert', 'ggplot2', 'gh', 'git2r', 'gitcreds', 'globals', 'glue',\n",
      "  'googlesheets4', 'gower', 'gtable', 'hardhat', 'haven', 'hms', 'htmltools',\n",
      "  'httpuv', 'httr', 'infer', 'ipred', 'IRdisplay', 'IRkernel', 'isoband',\n",
      "  'iterators', 'jsonlite', 'knitr', 'later', 'lattice', 'lhs', 'lifecycle',\n",
      "  'lmtest', 'lubridate', 'magrittr', 'MASS', 'Matrix', 'memoise', 'mgcv',\n",
      "  'mime', 'modeldata', 'modelr', 'nlme', 'nnet', 'objectProperties',\n",
      "  'objectSignals', 'openssl', 'parallelly', 'parsnip', 'patchwork', 'pbdZMQ',\n",
      "  'pillar', 'pkgbuild', 'pkgload', 'plyr', 'processx', 'progressr', 'proxy',\n",
      "  'ps', 'purrr', 'quantmod', 'randomForest', 'rcmdcheck', 'RColorBrewer',\n",
      "  'Rcpp', 'RcppArmadillo', 'RCurl', 'readr', 'readxl', 'recipes', 'remotes',\n",
      "  'repr', 'reprex', 'rex', 'rlang', 'rmarkdown', 'RODBC', 'roxygen2', 'rpart',\n",
      "  'rprojroot', 'rsample', 'RSQLite', 'rstudioapi', 'rversions', 'rvest',\n",
      "  'S4Vectors', 'sass', 'scales', 'sessioninfo', 'sevenbridges', 'shiny',\n",
      "  'stringi', 'stringr', 'survival', 'testthat', 'tibble', 'tidymodels',\n",
      "  'tidyr', 'tidyselect', 'tidyverse', 'timeDate', 'tinytex', 'tseries', 'TTR',\n",
      "  'tune', 'tzdb', 'urca', 'usethis', 'uuid', 'vctrs', 'viridisLite', 'vroom',\n",
      "  'waldo', 'withr', 'workflows', 'workflowsets', 'xfun', 'xml2', 'xts', 'yaml',\n",
      "  'yardstick', 'zip', 'zoo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(\"biomaRt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e5d10168-4933-411c-98f7-289eb247447a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: usethis\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"devtools\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05532e3f-6097-4f15-816b-6571a5993a91",
   "metadata": {},
   "source": [
    "#### First time through - Received a Warning that rhdf5 not available for this version of R\n",
    "\n",
    "Looked up our version and google searched R 4.1.1 rhdf5\n",
    "    \n",
    "Can install using `BiocManager::install(\"rhdf5\")`\n",
    "\n",
    "Pactherlab says to install `*rhdf5*` first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0a43a9e1-7ce6-4b68-860b-5e3e1876ab3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.18), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Installing package(s) 'rhdf5'\n",
      "\n",
      "also installing the dependencies ‘Rhdf5lib’, ‘rhdf5filters’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'backports', 'blob', 'brew', 'brio', 'broom', 'bslib', 'callr',\n",
      "  'caret', 'class', 'cli', 'clipr', 'colorspace', 'commonmark', 'conflicted',\n",
      "  'covr', 'cpp11', 'crayon', 'credentials', 'crosstalk', 'curl', 'data.table',\n",
      "  'DBI', 'dbplyr', 'desc', 'devtools', 'dials', 'diffobj', 'digest', 'dplyr',\n",
      "  'DT', 'dtplyr', 'e1071', 'evaluate', 'fansi', 'farver', 'forcats', 'foreach',\n",
      "  'forecast', 'fs', 'furrr', 'future', 'future.apply', 'gargle', 'generics',\n",
      "  'gert', 'ggplot2', 'gh', 'git2r', 'gitcreds', 'globals', 'glue',\n",
      "  'googlesheets4', 'gower', 'gtable', 'hardhat', 'haven', 'hms', 'htmltools',\n",
      "  'httpuv', 'httr', 'infer', 'ipred', 'IRdisplay', 'IRkernel', 'isoband',\n",
      "  'iterators', 'jsonlite', 'knitr', 'later', 'lattice', 'lhs', 'lifecycle',\n",
      "  'lmtest', 'lubridate', 'magrittr', 'MASS', 'Matrix', 'memoise', 'mgcv',\n",
      "  'mime', 'modeldata', 'modelr', 'nlme', 'nnet', 'objectProperties',\n",
      "  'objectSignals', 'openssl', 'parallelly', 'parsnip', 'patchwork', 'pbdZMQ',\n",
      "  'pillar', 'pkgbuild', 'pkgload', 'plyr', 'processx', 'progressr', 'proxy',\n",
      "  'ps', 'purrr', 'quantmod', 'randomForest', 'rcmdcheck', 'RColorBrewer',\n",
      "  'Rcpp', 'RcppArmadillo', 'RCurl', 'readr', 'readxl', 'recipes', 'remotes',\n",
      "  'repr', 'reprex', 'rex', 'rlang', 'rmarkdown', 'RODBC', 'roxygen2', 'rpart',\n",
      "  'rprojroot', 'rsample', 'RSQLite', 'rstudioapi', 'rversions', 'rvest',\n",
      "  'S4Vectors', 'sass', 'scales', 'sessioninfo', 'sevenbridges', 'shiny',\n",
      "  'stringi', 'stringr', 'survival', 'testthat', 'tibble', 'tidymodels',\n",
      "  'tidyr', 'tidyselect', 'tidyverse', 'timeDate', 'tinytex', 'tseries', 'TTR',\n",
      "  'tune', 'tzdb', 'urca', 'usethis', 'uuid', 'vctrs', 'viridisLite', 'vroom',\n",
      "  'waldo', 'withr', 'workflows', 'workflowsets', 'xfun', 'xml2', 'xts', 'yaml',\n",
      "  'yardstick', 'zip', 'zoo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(\"rhdf5\", force=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e7b03ce8-e5d8-4be8-be2b-e40c428d4541",
   "metadata": {},
   "outputs": [],
   "source": [
    "library (rhdf5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09f66a8d-241c-4f86-8b23-3fcf14859d4f",
   "metadata": {},
   "source": [
    "#### Issues\n",
    "\n",
    "Noted in issues https://github.com/pachterlab/sleuth/issues/259 -- follow the instructions from [Paast](https://github.com/pachterlab/sleuth/issues/259#issuecomment-966270599)\n",
    "\n",
    "Install rhdf5 as noted above.\n",
    "\n",
    "Load the library\n",
    "\n",
    "##### Clone sleuth and install after editing the file\n",
    "\n",
    "Change directory to the top working directory in this etheral machine.\n",
    "\n",
    "```bash\n",
    "cd /sbgenomics/workspace\n",
    "```\n",
    "\n",
    "now clone the library \n",
    "\n",
    "```bash\n",
    "git clone https://github.com/pachterlab/sleuth.git\n",
    "```\n",
    "\n",
    "edit NAMESPACE as the instructions note - to remove the dependency remove the last line to remove the reference to **rhdf5**\n",
    "\n",
    "And then run the install.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cace19ad-5aea-475b-8fe2-d6cc962eeba3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rlang        (0.4.11 -> 1.0.6 ) [CRAN]\n",
      "fs           (1.5.0  -> 1.5.2 ) [CRAN]\n",
      "digest       (0.6.27 -> 0.6.29) [CRAN]\n",
      "memoise      (2.0.0  -> 2.0.1 ) [CRAN]\n",
      "sass         (0.4.0  -> 0.4.2 ) [CRAN]\n",
      "jsonlite     (1.7.2  -> 1.8.2 ) [CRAN]\n",
      "htmltools    (0.5.2  -> 0.5.3 ) [CRAN]\n",
      "magrittr     (2.0.1  -> 2.0.3 ) [CRAN]\n",
      "later        (1.2.0  -> 1.3.0 ) [CRAN]\n",
      "Rcpp         (1.0.7  -> 1.0.9 ) [CRAN]\n",
      "stringi      (1.7.4  -> 1.7.8 ) [CRAN]\n",
      "glue         (1.4.2  -> 1.6.2 ) [CRAN]\n",
      "withr        (2.4.2  -> 2.5.0 ) [CRAN]\n",
      "fansi        (0.5.0  -> 1.0.3 ) [CRAN]\n",
      "cli          (3.0.1  -> 3.4.1 ) [CRAN]\n",
      "pillar       (1.6.2  -> 1.8.1 ) [CRAN]\n",
      "vctrs        (0.3.8  -> 0.4.2 ) [CRAN]\n",
      "tidyselect   (1.1.1  -> 1.2.0 ) [CRAN]\n",
      "tibble       (3.1.4  -> 3.1.8 ) [CRAN]\n",
      "lifecycle    (1.0.0  -> 1.0.3 ) [CRAN]\n",
      "generics     (0.1.0  -> 0.1.3 ) [CRAN]\n",
      "colorspace   (2.0-2  -> 2.0-3 ) [CRAN]\n",
      "viridisLite  (0.4.0  -> 0.4.1 ) [CRAN]\n",
      "RColorBrewer (1.1-2  -> 1.1-3 ) [CRAN]\n",
      "farver       (2.1.0  -> 2.1.1 ) [CRAN]\n",
      "bslib        (0.3.0  -> 0.4.0 ) [CRAN]\n",
      "commonmark   (1.7    -> 1.8.1 ) [CRAN]\n",
      "crayon       (1.4.1  -> 1.5.2 ) [CRAN]\n",
      "fontawesome  (NA     -> 0.3.0 ) [CRAN]\n",
      "mime         (0.11   -> 0.12  ) [CRAN]\n",
      "httpuv       (1.6.3  -> 1.6.6 ) [CRAN]\n",
      "gtable       (0.3.0  -> 0.3.1 ) [CRAN]\n",
      "scales       (1.1.1  -> 1.2.1 ) [CRAN]\n",
      "stringr      (1.4.0  -> 1.4.1 ) [CRAN]\n",
      "plyr         (1.8.6  -> 1.8.7 ) [CRAN]\n",
      "cpp11        (0.3.1  -> 0.4.3 ) [CRAN]\n",
      "purrr        (0.3.4  -> 0.3.5 ) [CRAN]\n",
      "dplyr        (1.0.7  -> 1.0.10) [CRAN]\n",
      "isoband      (0.2.5  -> 0.2.6 ) [CRAN]\n",
      "aggregation  (NA     -> 1.0.1 ) [CRAN]\n",
      "shiny        (1.6.0  -> 1.7.2 ) [CRAN]\n",
      "pheatmap     (NA     -> 1.0.12) [CRAN]\n",
      "matrixStats  (NA     -> 0.62.0) [CRAN]\n",
      "tidyr        (1.1.3  -> 1.2.1 ) [CRAN]\n",
      "data.table   (1.14.0 -> 1.14.4) [CRAN]\n",
      "ggplot2      (3.3.5  -> 3.3.6 ) [CRAN]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Skipping 1 packages not available: rhdf5\n",
      "\n",
      "Installing 46 packages: rlang, fs, digest, memoise, sass, jsonlite, htmltools, magrittr, later, Rcpp, stringi, glue, withr, fansi, cli, pillar, vctrs, tidyselect, tibble, lifecycle, generics, colorspace, viridisLite, RColorBrewer, farver, bslib, commonmark, crayon, fontawesome, mime, httpuv, gtable, scales, stringr, plyr, cpp11, purrr, dplyr, isoband, aggregation, shiny, pheatmap, matrixStats, tidyr, data.table, ggplot2\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m✔\u001b[39m  \u001b[90mchecking for file ‘/sbgenomics/workspace/sleuth/DESCRIPTION’\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mpreparing ‘sleuth’:\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[32m✔\u001b[39m  \u001b[90mchecking DESCRIPTION meta-information\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for LF line-endings in source and make files and shell scripts\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for empty or unneeded directories\u001b[39m\u001b[36m\u001b[39m\n",
      "   Omitted ‘LazyData’ from DESCRIPTION\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mbuilding ‘sleuth_0.30.0.tar.gz’\u001b[39m\u001b[36m\u001b[39m\n",
      "   \n",
      "Running /opt/conda/lib/R/bin/R CMD INSTALL \\\n",
      "  /tmp/RtmpHvladK/sleuth_0.30.0.tar.gz --install-tests \n",
      "* installing to library ‘/opt/conda/lib/R/library’\n",
      "* installing *source* package ‘sleuth’ ...\n",
      "** using staged installation\n",
      "** R\n",
      "** inst\n",
      "** tests\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (sleuth)\n"
     ]
    }
   ],
   "source": [
    "devtools::install('../../sleuth/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a69aba07-0138-427f-a8e8-a28bd30b8456",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(sleuth)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5947268-42ff-49ef-903b-980d1427b68a",
   "metadata": {},
   "source": [
    "We have successfully run Kallisto with Kallisto Quantitation.\n",
    "\n",
    "Results may be found after running an application on Cavatica here:\n",
    "\n",
    "```bash\n",
    "/sbgenomics/project-files/\n",
    "```\n",
    "\n",
    "For this analysis we will use the results from the run using `metadata_ten_samples_only_txt`\n",
    "\n",
    "Results are in:\n",
    "\n",
    "```bash\n",
    "/sbgenomics/project-files/ten_samples_expression_matrix.tpm.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea7bc5b5-9e72-4bb8-92bd-d24143bb2479",
   "metadata": {},
   "source": [
    "### Parsing metadata\n",
    "\n",
    "A sleuth analysis is dependent on a metadata file, which describes the experimental design, the sample names, conditions and covariates. The metadata file is external to sleuth, and must be prepared prior to analysis. A metadata file should have been downloaded along with the kallisto quantifications. The first step in a sleuth analysis is loading of the metadata file. You might need the path in read_table below to where you have downloaded the kallisto dataset, so that the path directs to the sample_table.txt. We then select the relevant columns of the metadata.\n",
    "\n",
    "In our case, I used:\n",
    "\n",
    "```bash\n",
    "/sbgenomics/project-files/metadata_ten_samples_only.csv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a430a324-b6b5-4255-81eb-b15734e115bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- read.table('/sbgenomics/project-files/metadata_ten_samples_only.csv', sep=\",\", header=TRUE, stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "809234a4-57cf-441d-a2c4-e1fd97127e90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 20 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>File.name</th><th scope=col>Case.ID</th><th scope=col>subject</th><th scope=col>Sample.ID</th><th scope=col>sample</th><th scope=col>Gender</th><th scope=col>Disease.type</th><th scope=col>Paired.end</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PAXSBH</td><td>PAXSBH</td><td>PAXSBH-03A-01R</td><td>PAXSBH-03A-01R</td><td>Female</td><td>TAM   </td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAXSBH</td><td>PAXSBH</td><td>PAXSBH-03A-01R</td><td>PAXSBH-03A-01R</td><td>Female</td><td>TAM   </td><td> 2</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PAXWGW</td><td>PAXWGW</td><td>PAXWGW-03A-01R</td><td>PAXWGW-03A-01R</td><td>Female</td><td>TAM   </td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAXWGW</td><td>PAXWGW</td><td>PAXWGW-03A-01R</td><td>PAXWGW-03A-01R</td><td>Female</td><td>TAM   </td><td> 2</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PASNSP</td><td>PASNSP</td><td>PASNSP-03A-01R</td><td>PASNSP-03A-01R</td><td>Male  </td><td>TAM   </td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PASNSP</td><td>PASNSP</td><td>PASNSP-03A-01R</td><td>PASNSP-03A-01R</td><td>Male  </td><td>TAM   </td><td> 2</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>PASWXF-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PASWXF</td><td>PASWXF</td><td>PASWXF-03A-01R</td><td>PASWXF-03A-01R</td><td>Male  </td><td>TAM   </td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>PASWXF-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PASWXF</td><td>PASWXF</td><td>PASWXF-03A-01R</td><td>PASWXF-03A-01R</td><td>Male  </td><td>TAM   </td><td> 2</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>PASXCL-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PASXCL</td><td>PASXCL</td><td>PASXCL-03A-01R</td><td>PASXCL-03A-01R</td><td>Male  </td><td>TAM   </td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>PASXCL-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PASXCL</td><td>PASXCL</td><td>PASXCL-03A-01R</td><td>PASXCL-03A-01R</td><td>Male  </td><td>TAM   </td><td> 2</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>PAVZTK-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAVZTK</td><td>PAVZTK</td><td>PAVZTK-09A-01R</td><td>PAVZTK-09A-01R</td><td>Female</td><td>DS-AML</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>PAVZTK-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PAVZTK</td><td>PAVZTK</td><td>PAVZTK-09A-01R</td><td>PAVZTK-09A-01R</td><td>Female</td><td>DS-AML</td><td> 2</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>PAWIYJ-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAWIYJ</td><td>PAWIYJ</td><td>PAWIYJ-09A-01R</td><td>PAWIYJ-09A-01R</td><td>Male  </td><td>DS-AML</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>PAWIYJ-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PAWIYJ</td><td>PAWIYJ</td><td>PAWIYJ-09A-01R</td><td>PAWIYJ-09A-01R</td><td>Male  </td><td>DS-AML</td><td> 2</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>PAXCGV-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAXCGV</td><td>PAXCGV</td><td>PAXCGV-09A-01R</td><td>PAXCGV-09A-01R</td><td>Female</td><td>DS-AML</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>PAXCGV-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PAXCGV</td><td>PAXCGV</td><td>PAXCGV-09A-01R</td><td>PAXCGV-09A-01R</td><td>Female</td><td>DS-AML</td><td> 2</td></tr>\n",
       "\t<tr><th scope=row>17</th><td>PAXLGI-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAXLGI</td><td>PAXLGI</td><td>PAXLGI-09A-01R</td><td>PAXLGI-09A-01R</td><td>Female</td><td>DS-AML</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>18</th><td>PAXLGI-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PAXLGI</td><td>PAXLGI</td><td>PAXLGI-09A-01R</td><td>PAXLGI-09A-01R</td><td>Female</td><td>DS-AML</td><td> 2</td></tr>\n",
       "\t<tr><th scope=row>19</th><td>PAXPEY-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAXPEY</td><td>PAXPEY</td><td>PAXPEY-09A-01R</td><td>PAXPEY-09A-01R</td><td>Male  </td><td>DS-AML</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>20</th><td>PAXPEY-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PAXPEY</td><td>PAXPEY</td><td>PAXPEY-09A-01R</td><td>PAXPEY-09A-01R</td><td>Male  </td><td>DS-AML</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 20 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & File.name & Case.ID & subject & Sample.ID & sample & Gender & Disease.type & Paired.end\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & PAXSBH-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PAXSBH & PAXSBH & PAXSBH-03A-01R & PAXSBH-03A-01R & Female & TAM    &  1\\\\\n",
       "\t2 & PAXSBH-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAXSBH & PAXSBH & PAXSBH-03A-01R & PAXSBH-03A-01R & Female & TAM    &  2\\\\\n",
       "\t3 & PAXWGW-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PAXWGW & PAXWGW & PAXWGW-03A-01R & PAXWGW-03A-01R & Female & TAM    &  1\\\\\n",
       "\t4 & PAXWGW-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAXWGW & PAXWGW & PAXWGW-03A-01R & PAXWGW-03A-01R & Female & TAM    &  2\\\\\n",
       "\t5 & PASNSP-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PASNSP & PASNSP & PASNSP-03A-01R & PASNSP-03A-01R & Male   & TAM    &  1\\\\\n",
       "\t6 & PASNSP-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PASNSP & PASNSP & PASNSP-03A-01R & PASNSP-03A-01R & Male   & TAM    &  2\\\\\n",
       "\t7 & PASWXF-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PASWXF & PASWXF & PASWXF-03A-01R & PASWXF-03A-01R & Male   & TAM    &  1\\\\\n",
       "\t8 & PASWXF-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PASWXF & PASWXF & PASWXF-03A-01R & PASWXF-03A-01R & Male   & TAM    &  2\\\\\n",
       "\t9 & PASXCL-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PASXCL & PASXCL & PASXCL-03A-01R & PASXCL-03A-01R & Male   & TAM    &  1\\\\\n",
       "\t10 & PASXCL-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PASXCL & PASXCL & PASXCL-03A-01R & PASXCL-03A-01R & Male   & TAM    &  2\\\\\n",
       "\t11 & PAVZTK-09A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAVZTK & PAVZTK & PAVZTK-09A-01R & PAVZTK-09A-01R & Female & DS-AML &  1\\\\\n",
       "\t12 & PAVZTK-09A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PAVZTK & PAVZTK & PAVZTK-09A-01R & PAVZTK-09A-01R & Female & DS-AML &  2\\\\\n",
       "\t13 & PAWIYJ-09A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAWIYJ & PAWIYJ & PAWIYJ-09A-01R & PAWIYJ-09A-01R & Male   & DS-AML &  1\\\\\n",
       "\t14 & PAWIYJ-09A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PAWIYJ & PAWIYJ & PAWIYJ-09A-01R & PAWIYJ-09A-01R & Male   & DS-AML &  2\\\\\n",
       "\t15 & PAXCGV-09A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAXCGV & PAXCGV & PAXCGV-09A-01R & PAXCGV-09A-01R & Female & DS-AML &  1\\\\\n",
       "\t16 & PAXCGV-09A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PAXCGV & PAXCGV & PAXCGV-09A-01R & PAXCGV-09A-01R & Female & DS-AML &  2\\\\\n",
       "\t17 & PAXLGI-09A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAXLGI & PAXLGI & PAXLGI-09A-01R & PAXLGI-09A-01R & Female & DS-AML &  1\\\\\n",
       "\t18 & PAXLGI-09A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PAXLGI & PAXLGI & PAXLGI-09A-01R & PAXLGI-09A-01R & Female & DS-AML &  2\\\\\n",
       "\t19 & PAXPEY-09A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAXPEY & PAXPEY & PAXPEY-09A-01R & PAXPEY-09A-01R & Male   & DS-AML &  1\\\\\n",
       "\t20 & PAXPEY-09A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PAXPEY & PAXPEY & PAXPEY-09A-01R & PAXPEY-09A-01R & Male   & DS-AML & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 20 × 8\n",
       "\n",
       "| <!--/--> | File.name &lt;chr&gt; | Case.ID &lt;chr&gt; | subject &lt;chr&gt; | Sample.ID &lt;chr&gt; | sample &lt;chr&gt; | Gender &lt;chr&gt; | Disease.type &lt;chr&gt; | Paired.end &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PAXSBH | PAXSBH | PAXSBH-03A-01R | PAXSBH-03A-01R | Female | TAM    |  1 |\n",
       "| 2 | PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAXSBH | PAXSBH | PAXSBH-03A-01R | PAXSBH-03A-01R | Female | TAM    |  2 |\n",
       "| 3 | PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PAXWGW | PAXWGW | PAXWGW-03A-01R | PAXWGW-03A-01R | Female | TAM    |  1 |\n",
       "| 4 | PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAXWGW | PAXWGW | PAXWGW-03A-01R | PAXWGW-03A-01R | Female | TAM    |  2 |\n",
       "| 5 | PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PASNSP | PASNSP | PASNSP-03A-01R | PASNSP-03A-01R | Male   | TAM    |  1 |\n",
       "| 6 | PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PASNSP | PASNSP | PASNSP-03A-01R | PASNSP-03A-01R | Male   | TAM    |  2 |\n",
       "| 7 | PASWXF-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PASWXF | PASWXF | PASWXF-03A-01R | PASWXF-03A-01R | Male   | TAM    |  1 |\n",
       "| 8 | PASWXF-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PASWXF | PASWXF | PASWXF-03A-01R | PASWXF-03A-01R | Male   | TAM    |  2 |\n",
       "| 9 | PASXCL-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PASXCL | PASXCL | PASXCL-03A-01R | PASXCL-03A-01R | Male   | TAM    |  1 |\n",
       "| 10 | PASXCL-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PASXCL | PASXCL | PASXCL-03A-01R | PASXCL-03A-01R | Male   | TAM    |  2 |\n",
       "| 11 | PAVZTK-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAVZTK | PAVZTK | PAVZTK-09A-01R | PAVZTK-09A-01R | Female | DS-AML |  1 |\n",
       "| 12 | PAVZTK-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PAVZTK | PAVZTK | PAVZTK-09A-01R | PAVZTK-09A-01R | Female | DS-AML |  2 |\n",
       "| 13 | PAWIYJ-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAWIYJ | PAWIYJ | PAWIYJ-09A-01R | PAWIYJ-09A-01R | Male   | DS-AML |  1 |\n",
       "| 14 | PAWIYJ-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PAWIYJ | PAWIYJ | PAWIYJ-09A-01R | PAWIYJ-09A-01R | Male   | DS-AML |  2 |\n",
       "| 15 | PAXCGV-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAXCGV | PAXCGV | PAXCGV-09A-01R | PAXCGV-09A-01R | Female | DS-AML |  1 |\n",
       "| 16 | PAXCGV-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PAXCGV | PAXCGV | PAXCGV-09A-01R | PAXCGV-09A-01R | Female | DS-AML |  2 |\n",
       "| 17 | PAXLGI-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAXLGI | PAXLGI | PAXLGI-09A-01R | PAXLGI-09A-01R | Female | DS-AML |  1 |\n",
       "| 18 | PAXLGI-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PAXLGI | PAXLGI | PAXLGI-09A-01R | PAXLGI-09A-01R | Female | DS-AML |  2 |\n",
       "| 19 | PAXPEY-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAXPEY | PAXPEY | PAXPEY-09A-01R | PAXPEY-09A-01R | Male   | DS-AML |  1 |\n",
       "| 20 | PAXPEY-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PAXPEY | PAXPEY | PAXPEY-09A-01R | PAXPEY-09A-01R | Male   | DS-AML | NA |\n",
       "\n"
      ],
      "text/plain": [
       "   File.name                                                     Case.ID\n",
       "1  PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PAXSBH \n",
       "2  PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAXSBH \n",
       "3  PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PAXWGW \n",
       "4  PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAXWGW \n",
       "5  PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PASNSP \n",
       "6  PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PASNSP \n",
       "7  PASWXF-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PASWXF \n",
       "8  PASWXF-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PASWXF \n",
       "9  PASXCL-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PASXCL \n",
       "10 PASXCL-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PASXCL \n",
       "11 PAVZTK-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAVZTK \n",
       "12 PAVZTK-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PAVZTK \n",
       "13 PAWIYJ-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAWIYJ \n",
       "14 PAWIYJ-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PAWIYJ \n",
       "15 PAXCGV-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAXCGV \n",
       "16 PAXCGV-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PAXCGV \n",
       "17 PAXLGI-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAXLGI \n",
       "18 PAXLGI-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PAXLGI \n",
       "19 PAXPEY-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAXPEY \n",
       "20 PAXPEY-09A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PAXPEY \n",
       "   subject Sample.ID      sample         Gender Disease.type Paired.end\n",
       "1  PAXSBH  PAXSBH-03A-01R PAXSBH-03A-01R Female TAM           1        \n",
       "2  PAXSBH  PAXSBH-03A-01R PAXSBH-03A-01R Female TAM           2        \n",
       "3  PAXWGW  PAXWGW-03A-01R PAXWGW-03A-01R Female TAM           1        \n",
       "4  PAXWGW  PAXWGW-03A-01R PAXWGW-03A-01R Female TAM           2        \n",
       "5  PASNSP  PASNSP-03A-01R PASNSP-03A-01R Male   TAM           1        \n",
       "6  PASNSP  PASNSP-03A-01R PASNSP-03A-01R Male   TAM           2        \n",
       "7  PASWXF  PASWXF-03A-01R PASWXF-03A-01R Male   TAM           1        \n",
       "8  PASWXF  PASWXF-03A-01R PASWXF-03A-01R Male   TAM           2        \n",
       "9  PASXCL  PASXCL-03A-01R PASXCL-03A-01R Male   TAM           1        \n",
       "10 PASXCL  PASXCL-03A-01R PASXCL-03A-01R Male   TAM           2        \n",
       "11 PAVZTK  PAVZTK-09A-01R PAVZTK-09A-01R Female DS-AML        1        \n",
       "12 PAVZTK  PAVZTK-09A-01R PAVZTK-09A-01R Female DS-AML        2        \n",
       "13 PAWIYJ  PAWIYJ-09A-01R PAWIYJ-09A-01R Male   DS-AML        1        \n",
       "14 PAWIYJ  PAWIYJ-09A-01R PAWIYJ-09A-01R Male   DS-AML        2        \n",
       "15 PAXCGV  PAXCGV-09A-01R PAXCGV-09A-01R Female DS-AML        1        \n",
       "16 PAXCGV  PAXCGV-09A-01R PAXCGV-09A-01R Female DS-AML        2        \n",
       "17 PAXLGI  PAXLGI-09A-01R PAXLGI-09A-01R Female DS-AML        1        \n",
       "18 PAXLGI  PAXLGI-09A-01R PAXLGI-09A-01R Female DS-AML        2        \n",
       "19 PAXPEY  PAXPEY-09A-01R PAXPEY-09A-01R Male   DS-AML        1        \n",
       "20 PAXPEY  PAXPEY-09A-01R PAXPEY-09A-01R Male   DS-AML       NA        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(metadata, n=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83d55633-bf9e-4b00-90d0-ef75788cef3a",
   "metadata": {},
   "source": [
    "There is an error in the last sample detail - where the paired should read `2` and not NA.  So I copied the file to a local directory and corrected it -- it is corrected permanently now - but for this run through you can see:\n",
    "```bash\n",
    "cp /sbgenomics/project-files/metadata_ten_samples_only.csv /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data\n",
    "```\n",
    "\n",
    "where I edited the file and now will read this one in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6d94ba4a-a191-494e-b2ed-d589c47843c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- read.table('/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/metadata_ten_samples_only.csv', sep=\",\", header=TRUE, stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "cf79d1bc-1a14-48c8-a57a-83f8116bd9ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>oFile.name</th><th scope=col>Case.ID</th><th scope=col>subject</th><th scope=col>Sample.ID</th><th scope=col>sample</th><th scope=col>Gender</th><th scope=col>Disease.type</th><th scope=col>Paired.end</th><th scope=col>Abundance</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAXSBH</td><td>PAXSBH</td><td>PAXSBH-03A-01R</td><td>PAXSBH-03A-01R</td><td>Female</td><td>TAM</td><td>1</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAXSBH</td><td>PAXSBH</td><td>PAXSBH-03A-01R</td><td>PAXSBH-03A-01R</td><td>Female</td><td>TAM</td><td>2</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PAXWGW</td><td>PAXWGW</td><td>PAXWGW-03A-01R</td><td>PAXWGW-03A-01R</td><td>Female</td><td>TAM</td><td>1</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PAXWGW</td><td>PAXWGW</td><td>PAXWGW-03A-01R</td><td>PAXWGW-03A-01R</td><td>Female</td><td>TAM</td><td>2</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz</td><td>PASNSP</td><td>PASNSP</td><td>PASNSP-03A-01R</td><td>PASNSP-03A-01R</td><td>Male  </td><td>TAM</td><td>1</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz</td><td>PASNSP</td><td>PASNSP</td><td>PASNSP-03A-01R</td><td>PASNSP-03A-01R</td><td>Male  </td><td>TAM</td><td>2</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & oFile.name & Case.ID & subject & Sample.ID & sample & Gender & Disease.type & Paired.end & Abundance\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & PAXSBH-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAXSBH & PAXSBH & PAXSBH-03A-01R & PAXSBH-03A-01R & Female & TAM & 1 & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t2 & PAXSBH-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAXSBH & PAXSBH & PAXSBH-03A-01R & PAXSBH-03A-01R & Female & TAM & 2 & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t3 & PAXWGW-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PAXWGW & PAXWGW & PAXWGW-03A-01R & PAXWGW-03A-01R & Female & TAM & 1 & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t4 & PAXWGW-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PAXWGW & PAXWGW & PAXWGW-03A-01R & PAXWGW-03A-01R & Female & TAM & 2 & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t5 & PASNSP-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r1.fq.gz & PASNSP & PASNSP & PASNSP-03A-01R & PASNSP-03A-01R & Male   & TAM & 1 & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t6 & PASNSP-03A-01R\\_RBS\\_withJunctionsOnGenome\\_dupsFlagged\\_r2.fq.gz & PASNSP & PASNSP & PASNSP-03A-01R & PASNSP-03A-01R & Male   & TAM & 2 & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | oFile.name &lt;chr&gt; | Case.ID &lt;chr&gt; | subject &lt;chr&gt; | Sample.ID &lt;chr&gt; | sample &lt;chr&gt; | Gender &lt;chr&gt; | Disease.type &lt;chr&gt; | Paired.end &lt;int&gt; | Abundance &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAXSBH | PAXSBH | PAXSBH-03A-01R | PAXSBH-03A-01R | Female | TAM | 1 | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 2 | PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAXSBH | PAXSBH | PAXSBH-03A-01R | PAXSBH-03A-01R | Female | TAM | 2 | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 3 | PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PAXWGW | PAXWGW | PAXWGW-03A-01R | PAXWGW-03A-01R | Female | TAM | 1 | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 4 | PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PAXWGW | PAXWGW | PAXWGW-03A-01R | PAXWGW-03A-01R | Female | TAM | 2 | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 5 | PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz | PASNSP | PASNSP | PASNSP-03A-01R | PASNSP-03A-01R | Male   | TAM | 1 | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 6 | PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz | PASNSP | PASNSP | PASNSP-03A-01R | PASNSP-03A-01R | Male   | TAM | 2 | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv |\n",
       "\n"
      ],
      "text/plain": [
       "  oFile.name                                                    Case.ID subject\n",
       "1 PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAXSBH  PAXSBH \n",
       "2 PAXSBH-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAXSBH  PAXSBH \n",
       "3 PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PAXWGW  PAXWGW \n",
       "4 PAXWGW-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PAXWGW  PAXWGW \n",
       "5 PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r1.fq.gz PASNSP  PASNSP \n",
       "6 PASNSP-03A-01R_RBS_withJunctionsOnGenome_dupsFlagged_r2.fq.gz PASNSP  PASNSP \n",
       "  Sample.ID      sample         Gender Disease.type Paired.end\n",
       "1 PAXSBH-03A-01R PAXSBH-03A-01R Female TAM          1         \n",
       "2 PAXSBH-03A-01R PAXSBH-03A-01R Female TAM          2         \n",
       "3 PAXWGW-03A-01R PAXWGW-03A-01R Female TAM          1         \n",
       "4 PAXWGW-03A-01R PAXWGW-03A-01R Female TAM          2         \n",
       "5 PASNSP-03A-01R PASNSP-03A-01R Male   TAM          1         \n",
       "6 PASNSP-03A-01R PASNSP-03A-01R Male   TAM          2         \n",
       "  Abundance                                                                                           \n",
       "1 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv\n",
       "2 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv\n",
       "3 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv\n",
       "4 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv\n",
       "5 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv\n",
       "6 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "9209ebbd-dcfa-49b7-bb8e-413004406e8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>20</li><li>9</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 20\n",
       "\\item 9\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 20\n",
       "2. 9\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 20  9"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "f907e43b-7a13-47bf-8aad-dc88eb09f022",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- dplyr::select(metadata, c('Case.ID', 'Sample.ID', 'Gender', 'Disease.type', 'Abundance'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "fea7756a-b5e7-4707-909c-10de1fefc960",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Case.ID</th><th scope=col>Sample.ID</th><th scope=col>Gender</th><th scope=col>Disease.type</th><th scope=col>Abundance</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>PAXSBH</td><td>PAXSBH-03A-01R</td><td>Female</td><td>TAM</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>PAXSBH</td><td>PAXSBH-03A-01R</td><td>Female</td><td>TAM</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>PAXWGW</td><td>PAXWGW-03A-01R</td><td>Female</td><td>TAM</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>PAXWGW</td><td>PAXWGW-03A-01R</td><td>Female</td><td>TAM</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>PASNSP</td><td>PASNSP-03A-01R</td><td>Male  </td><td>TAM</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>PASNSP</td><td>PASNSP-03A-01R</td><td>Male  </td><td>TAM</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & Case.ID & Sample.ID & Gender & Disease.type & Abundance\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & PAXSBH & PAXSBH-03A-01R & Female & TAM & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t2 & PAXSBH & PAXSBH-03A-01R & Female & TAM & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t3 & PAXWGW & PAXWGW-03A-01R & Female & TAM & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t4 & PAXWGW & PAXWGW-03A-01R & Female & TAM & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t5 & PASNSP & PASNSP-03A-01R & Male   & TAM & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t6 & PASNSP & PASNSP-03A-01R & Male   & TAM & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | Case.ID &lt;chr&gt; | Sample.ID &lt;chr&gt; | Gender &lt;chr&gt; | Disease.type &lt;chr&gt; | Abundance &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | PAXSBH | PAXSBH-03A-01R | Female | TAM | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 2 | PAXSBH | PAXSBH-03A-01R | Female | TAM | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 3 | PAXWGW | PAXWGW-03A-01R | Female | TAM | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 4 | PAXWGW | PAXWGW-03A-01R | Female | TAM | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 5 | PASNSP | PASNSP-03A-01R | Male   | TAM | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 6 | PASNSP | PASNSP-03A-01R | Male   | TAM | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv |\n",
       "\n"
      ],
      "text/plain": [
       "  Case.ID Sample.ID      Gender Disease.type\n",
       "1 PAXSBH  PAXSBH-03A-01R Female TAM         \n",
       "2 PAXSBH  PAXSBH-03A-01R Female TAM         \n",
       "3 PAXWGW  PAXWGW-03A-01R Female TAM         \n",
       "4 PAXWGW  PAXWGW-03A-01R Female TAM         \n",
       "5 PASNSP  PASNSP-03A-01R Male   TAM         \n",
       "6 PASNSP  PASNSP-03A-01R Male   TAM         \n",
       "  Abundance                                                                                           \n",
       "1 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv\n",
       "2 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv\n",
       "3 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv\n",
       "4 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv\n",
       "5 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv\n",
       "6 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "9be16242-cd08-413b-a17e-8958d431c4f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- dplyr::distinct(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "3b4a9c46-262d-4daa-9c61-7c43ecde421d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Case.ID</th><th scope=col>Sample.ID</th><th scope=col>Gender</th><th scope=col>Disease.type</th><th scope=col>Abundance</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>PAXSBH</td><td>PAXSBH-03A-01R</td><td>Female</td><td>TAM   </td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>PAXWGW</td><td>PAXWGW-03A-01R</td><td>Female</td><td>TAM   </td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>PASNSP</td><td>PASNSP-03A-01R</td><td>Male  </td><td>TAM   </td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>PASWXF</td><td>PASWXF-03A-01R</td><td>Male  </td><td>TAM   </td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASWXF-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>PASXCL</td><td>PASXCL-03A-01R</td><td>Male  </td><td>TAM   </td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASXCL-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>PAVZTK</td><td>PAVZTK-09A-01R</td><td>Female</td><td>DS-AML</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAVZTK-09A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & Case.ID & Sample.ID & Gender & Disease.type & Abundance\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & PAXSBH & PAXSBH-03A-01R & Female & TAM    & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t2 & PAXWGW & PAXWGW-03A-01R & Female & TAM    & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t3 & PASNSP & PASNSP-03A-01R & Male   & TAM    & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t4 & PASWXF & PASWXF-03A-01R & Male   & TAM    & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASWXF-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t5 & PASXCL & PASXCL-03A-01R & Male   & TAM    & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASXCL-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t6 & PAVZTK & PAVZTK-09A-01R & Female & DS-AML & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAVZTK-09A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | Case.ID &lt;chr&gt; | Sample.ID &lt;chr&gt; | Gender &lt;chr&gt; | Disease.type &lt;chr&gt; | Abundance &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | PAXSBH | PAXSBH-03A-01R | Female | TAM    | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 2 | PAXWGW | PAXWGW-03A-01R | Female | TAM    | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 3 | PASNSP | PASNSP-03A-01R | Male   | TAM    | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 4 | PASWXF | PASWXF-03A-01R | Male   | TAM    | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASWXF-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 5 | PASXCL | PASXCL-03A-01R | Male   | TAM    | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASXCL-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 6 | PAVZTK | PAVZTK-09A-01R | Female | DS-AML | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAVZTK-09A-01R.kallisto_quant.abundance.tsv |\n",
       "\n"
      ],
      "text/plain": [
       "  Case.ID Sample.ID      Gender Disease.type\n",
       "1 PAXSBH  PAXSBH-03A-01R Female TAM         \n",
       "2 PAXWGW  PAXWGW-03A-01R Female TAM         \n",
       "3 PASNSP  PASNSP-03A-01R Male   TAM         \n",
       "4 PASWXF  PASWXF-03A-01R Male   TAM         \n",
       "5 PASXCL  PASXCL-03A-01R Male   TAM         \n",
       "6 PAVZTK  PAVZTK-09A-01R Female DS-AML      \n",
       "  Abundance                                                                                           \n",
       "1 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv\n",
       "2 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv\n",
       "3 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv\n",
       "4 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASWXF-03A-01R.kallisto_quant.abundance.tsv\n",
       "5 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASXCL-03A-01R.kallisto_quant.abundance.tsv\n",
       "6 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAVZTK-09A-01R.kallisto_quant.abundance.tsv"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "679e4004-878f-4937-b888-ef9fd43f01fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>10</li><li>5</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 10\n",
       "\\item 5\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 10\n",
       "2. 5\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 10  5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "0656197e-93f0-4008-b3bd-35c14adf1522",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- dplyr::rename(metadata, sample = Sample.ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0617541a-0664-46bf-896e-4838af463e9a",
   "metadata": {},
   "source": [
    "Need to rename a colump as well to `path` where we have `Abundance`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "11e98d5a-534a-4021-bfed-ffa7768cd8a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- dplyr::rename(metadata, path = Abundance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "edc62a33-02e1-4fbb-9b80-6a53a66a41e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Case.ID</th><th scope=col>sample</th><th scope=col>Gender</th><th scope=col>Disease.type</th><th scope=col>path</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>PAXSBH</td><td>PAXSBH-03A-01R</td><td>Female</td><td>TAM   </td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>PAXWGW</td><td>PAXWGW-03A-01R</td><td>Female</td><td>TAM   </td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>PASNSP</td><td>PASNSP-03A-01R</td><td>Male  </td><td>TAM   </td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>PASWXF</td><td>PASWXF-03A-01R</td><td>Male  </td><td>TAM   </td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASWXF-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>PASXCL</td><td>PASXCL-03A-01R</td><td>Male  </td><td>TAM   </td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASXCL-03A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>PAVZTK</td><td>PAVZTK-09A-01R</td><td>Female</td><td>DS-AML</td><td>/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAVZTK-09A-01R.kallisto_quant.abundance.tsv</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & Case.ID & sample & Gender & Disease.type & path\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & PAXSBH & PAXSBH-03A-01R & Female & TAM    & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t2 & PAXWGW & PAXWGW-03A-01R & Female & TAM    & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t3 & PASNSP & PASNSP-03A-01R & Male   & TAM    & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t4 & PASWXF & PASWXF-03A-01R & Male   & TAM    & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASWXF-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t5 & PASXCL & PASXCL-03A-01R & Male   & TAM    & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASXCL-03A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\t6 & PAVZTK & PAVZTK-09A-01R & Female & DS-AML & /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAVZTK-09A-01R.kallisto\\_quant.abundance.tsv\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | Case.ID &lt;chr&gt; | sample &lt;chr&gt; | Gender &lt;chr&gt; | Disease.type &lt;chr&gt; | path &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | PAXSBH | PAXSBH-03A-01R | Female | TAM    | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 2 | PAXWGW | PAXWGW-03A-01R | Female | TAM    | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 3 | PASNSP | PASNSP-03A-01R | Male   | TAM    | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 4 | PASWXF | PASWXF-03A-01R | Male   | TAM    | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASWXF-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 5 | PASXCL | PASXCL-03A-01R | Male   | TAM    | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASXCL-03A-01R.kallisto_quant.abundance.tsv |\n",
       "| 6 | PAVZTK | PAVZTK-09A-01R | Female | DS-AML | /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAVZTK-09A-01R.kallisto_quant.abundance.tsv |\n",
       "\n"
      ],
      "text/plain": [
       "  Case.ID sample         Gender Disease.type\n",
       "1 PAXSBH  PAXSBH-03A-01R Female TAM         \n",
       "2 PAXWGW  PAXWGW-03A-01R Female TAM         \n",
       "3 PASNSP  PASNSP-03A-01R Male   TAM         \n",
       "4 PASWXF  PASWXF-03A-01R Male   TAM         \n",
       "5 PASXCL  PASXCL-03A-01R Male   TAM         \n",
       "6 PAVZTK  PAVZTK-09A-01R Female DS-AML      \n",
       "  path                                                                                                \n",
       "1 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXSBH-03A-01R.kallisto_quant.abundance.tsv\n",
       "2 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAXWGW-03A-01R.kallisto_quant.abundance.tsv\n",
       "3 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASNSP-03A-01R.kallisto_quant.abundance.tsv\n",
       "4 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASWXF-03A-01R.kallisto_quant.abundance.tsv\n",
       "5 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PASXCL-03A-01R.kallisto_quant.abundance.tsv\n",
       "6 /sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/PAVZTK-09A-01R.kallisto_quant.abundance.tsv"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f8ecb14-d688-47ec-bbee-c84bdc5bdb4d",
   "metadata": {},
   "source": [
    "#### biomaRt - how to use\n",
    "\n",
    "Following instructions from the [ensembl site](https://grch37.ensembl.org/info/data/biomart/biomart_r_package.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b315a983-b7eb-47c5-a618-3558f1988f21",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(biomaRt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9a527c62-5125-4471-be1c-eaf43f2a0d29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 4 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>biomart</th><th scope=col>version</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>genes        </td><td>Ensembl Genes 107     </td></tr>\n",
       "\t<tr><td>mouse_strains</td><td>Mouse strains 107     </td></tr>\n",
       "\t<tr><td>snps         </td><td>Ensembl Variation 107 </td></tr>\n",
       "\t<tr><td>regulation   </td><td>Ensembl Regulation 107</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 2\n",
       "\\begin{tabular}{ll}\n",
       " biomart & version\\\\\n",
       " <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t genes         & Ensembl Genes 107     \\\\\n",
       "\t mouse\\_strains & Mouse strains 107     \\\\\n",
       "\t snps          & Ensembl Variation 107 \\\\\n",
       "\t regulation    & Ensembl Regulation 107\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 2\n",
       "\n",
       "| biomart &lt;chr&gt; | version &lt;chr&gt; |\n",
       "|---|---|\n",
       "| genes         | Ensembl Genes 107      |\n",
       "| mouse_strains | Mouse strains 107      |\n",
       "| snps          | Ensembl Variation 107  |\n",
       "| regulation    | Ensembl Regulation 107 |\n",
       "\n"
      ],
      "text/plain": [
       "  biomart       version               \n",
       "1 genes         Ensembl Genes 107     \n",
       "2 mouse_strains Mouse strains 107     \n",
       "3 snps          Ensembl Variation 107 \n",
       "4 regulation    Ensembl Regulation 107"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listEnsembl()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "cf6b61ce-eaf9-4ede-b84f-a498ab0e16eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "mart <- biomaRt::useMart(biomart=\"ensembl\", \n",
    "                     dataset = \"hsapiens_gene_ensembl\",\n",
    "                        host = \"https://useast.ensembl.org\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bf534bf-e186-4d0c-b6b4-ac930e640051",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "09a5d947-397f-46de-a8e6-88658fb08d32",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttg <- biomaRt::getBM(\n",
    "  attributes = c(\"ensembl_transcript_id\", \"transcript_version\",\n",
    "  \"ensembl_gene_id\", \"external_gene_name\", \"description\",\n",
    "  \"transcript_biotype\"),\n",
    "  mart = mart)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "7386de62-67da-4ca9-a0d4-d5d7ad831a05",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttg <- dplyr::rename(ttg, target_id = ensembl_transcript_id,\n",
    "  ens_gene = ensembl_gene_id, ext_gene = external_gene_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "0521385a-1bd1-402f-a9b0-53b52a1a4aae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>target_id</th><th scope=col>ens_gene</th><th scope=col>ext_gene</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENST00000387314</td><td>ENSG00000210049</td><td>MT-TF  </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENST00000389680</td><td>ENSG00000211459</td><td>MT-RNR1</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENST00000387342</td><td>ENSG00000210077</td><td>MT-TV  </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENST00000387347</td><td>ENSG00000210082</td><td>MT-RNR2</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENST00000386347</td><td>ENSG00000209082</td><td>MT-TL1 </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENST00000361390</td><td>ENSG00000198888</td><td>MT-ND1 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & target\\_id & ens\\_gene & ext\\_gene\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENST00000387314 & ENSG00000210049 & MT-TF  \\\\\n",
       "\t2 & ENST00000389680 & ENSG00000211459 & MT-RNR1\\\\\n",
       "\t3 & ENST00000387342 & ENSG00000210077 & MT-TV  \\\\\n",
       "\t4 & ENST00000387347 & ENSG00000210082 & MT-RNR2\\\\\n",
       "\t5 & ENST00000386347 & ENSG00000209082 & MT-TL1 \\\\\n",
       "\t6 & ENST00000361390 & ENSG00000198888 & MT-ND1 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | target_id &lt;chr&gt; | ens_gene &lt;chr&gt; | ext_gene &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | ENST00000387314 | ENSG00000210049 | MT-TF   |\n",
       "| 2 | ENST00000389680 | ENSG00000211459 | MT-RNR1 |\n",
       "| 3 | ENST00000387342 | ENSG00000210077 | MT-TV   |\n",
       "| 4 | ENST00000387347 | ENSG00000210082 | MT-RNR2 |\n",
       "| 5 | ENST00000386347 | ENSG00000209082 | MT-TL1  |\n",
       "| 6 | ENST00000361390 | ENSG00000198888 | MT-ND1  |\n",
       "\n"
      ],
      "text/plain": [
       "  target_id       ens_gene        ext_gene\n",
       "1 ENST00000387314 ENSG00000210049 MT-TF   \n",
       "2 ENST00000389680 ENSG00000211459 MT-RNR1 \n",
       "3 ENST00000387342 ENSG00000210077 MT-TV   \n",
       "4 ENST00000387347 ENSG00000210082 MT-RNR2 \n",
       "5 ENST00000386347 ENSG00000209082 MT-TL1  \n",
       "6 ENST00000361390 ENSG00000198888 MT-ND1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ttg <- dplyr::select(ttg, c('target_id', 'ens_gene', 'ext_gene'))\n",
    "head(ttg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "7d913a41-9424-440e-8f8d-218c8001fd8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttg <- read.table('/sbgenomics/workspace/pediatric-DS-AML-TAM-Analysis/data/ttg.csv', sep=\",\", header=TRUE, stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "8b09c920-8118-4f34-a303-68e07003a868",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>target_id</th><th scope=col>ens_gene</th><th scope=col>ext_gene</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENST00000387314</td><td>ENSG00000210049</td><td>MT-TF  </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENST00000389680</td><td>ENSG00000211459</td><td>MT-RNR1</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENST00000387342</td><td>ENSG00000210077</td><td>MT-TV  </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENST00000387347</td><td>ENSG00000210082</td><td>MT-RNR2</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENST00000386347</td><td>ENSG00000209082</td><td>MT-TL1 </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENST00000361390</td><td>ENSG00000198888</td><td>MT-ND1 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & target\\_id & ens\\_gene & ext\\_gene\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENST00000387314 & ENSG00000210049 & MT-TF  \\\\\n",
       "\t2 & ENST00000389680 & ENSG00000211459 & MT-RNR1\\\\\n",
       "\t3 & ENST00000387342 & ENSG00000210077 & MT-TV  \\\\\n",
       "\t4 & ENST00000387347 & ENSG00000210082 & MT-RNR2\\\\\n",
       "\t5 & ENST00000386347 & ENSG00000209082 & MT-TL1 \\\\\n",
       "\t6 & ENST00000361390 & ENSG00000198888 & MT-ND1 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | target_id &lt;chr&gt; | ens_gene &lt;chr&gt; | ext_gene &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | ENST00000387314 | ENSG00000210049 | MT-TF   |\n",
       "| 2 | ENST00000389680 | ENSG00000211459 | MT-RNR1 |\n",
       "| 3 | ENST00000387342 | ENSG00000210077 | MT-TV   |\n",
       "| 4 | ENST00000387347 | ENSG00000210082 | MT-RNR2 |\n",
       "| 5 | ENST00000386347 | ENSG00000209082 | MT-TL1  |\n",
       "| 6 | ENST00000361390 | ENSG00000198888 | MT-ND1  |\n",
       "\n"
      ],
      "text/plain": [
       "  target_id       ens_gene        ext_gene\n",
       "1 ENST00000387314 ENSG00000210049 MT-TF   \n",
       "2 ENST00000389680 ENSG00000211459 MT-RNR1 \n",
       "3 ENST00000387342 ENSG00000210077 MT-TV   \n",
       "4 ENST00000387347 ENSG00000210082 MT-RNR2 \n",
       "5 ENST00000386347 ENSG00000209082 MT-TL1  \n",
       "6 ENST00000361390 ENSG00000198888 MT-ND1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(ttg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ad386e6-6c48-4e78-94f2-dc2b76581643",
   "metadata": {},
   "source": [
    "The resulting table contains Ensembl gene names (‘ens_gene’) and the associated transcripts (‘target_id’). Note that the gene-transcript mapping must be compatible with the transcriptome used with kallisto. In other words, to use Ensembl transcript-gene associations kallisto was run using the Ensembl transcriptome.\n",
    "\n",
    "#### Preparing the analysis\n",
    "\n",
    "The next step is to build a sleuth object. The sleuth object contains specification of the experimental design, a map describing grouping of transcripts into genes (or other groups), and a number of user specific parameters. In the example that follows, metadata is the experimental design and target_mapping describes the transcript groupings into genes previously constructed. Furthermore, we provide an aggregation_column, the column name of in ‘target_mapping’ table that is used to aggregate the transcripts. When both ‘target_mapping’ and ‘aggregation_column’ are provided, sleuth will automatically run in gene mode, returning gene differential expression results that came from the aggregation of transcript p-values.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "2f090924-7335-4626-9bd0-a9e3d46db878",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'gene_mode' is TRUE. Sleuth will do counts aggregation at the gene level for downstream normalization, transformation, and modeling steps, as well as for plotting and results.\n",
      "\n",
      "reading in kallisto results\n",
      "\n",
      "dropping unused factor levels\n",
      "\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "\n",
      "\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in if (any(counts_test$total == 0)) {: missing value where TRUE/FALSE needed\n",
     "output_type": "error",
     "traceback": [
      "Error in if (any(counts_test$total == 0)) {: missing value where TRUE/FALSE needed\nTraceback:\n",
      "1. sleuth_prep(metadata, target_mapping = ttg, aggregation_column = \"ens_gene\", \n .     gene_mode = TRUE, extra_bootstrap_summary = TRUE)"
     ]
    }
   ],
   "source": [
    "so <- sleuth_prep(metadata, \n",
    "                  target_mapping = ttg, \n",
    "                  aggregation_column = 'ens_gene',\n",
    "                  gene_mode = TRUE,\n",
    "                  extra_bootstrap_summary = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "344c22b5-cb9b-4ee3-9a04-fdb522635298",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (sample_to_covariates, full_model = NULL, target_mapping = NULL, \n",
       "<span style=white-space:pre-wrap>    aggregation_column = NULL, num_cores = max(1L, parallel::detectCores() - </span>\n",
       "<span style=white-space:pre-wrap>        1L), ...) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    extra_opts &lt;- list(...)</span>\n",
       "<span style=white-space:pre-wrap>    if (\"extra_bootstrap_summary\" %in% names(extra_opts)) {</span>\n",
       "<span style=white-space:pre-wrap>        extra_bootstrap_summary &lt;- extra_opts$extra_bootstrap_summary</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        extra_bootstrap_summary &lt;- FALSE</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (\"read_bootstrap_tpm\" %in% names(extra_opts)) {</span>\n",
       "<span style=white-space:pre-wrap>        read_bootstrap_tpm &lt;- extra_opts$read_bootstrap_tpm</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        read_bootstrap_tpm &lt;- FALSE</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (\"max_bootstrap\" %in% names(extra_opts)) {</span>\n",
       "<span style=white-space:pre-wrap>        max_bootstrap &lt;- extra_opts$max_bootstrap</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        max_bootstrap &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (\"filter_fun\" %in% names(extra_opts)) {</span>\n",
       "<span style=white-space:pre-wrap>        filter_fun &lt;- extra_opts$filter_fun</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        filter_fun &lt;- basic_filter</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (\"filter_target_id\" %in% names(extra_opts)) {</span>\n",
       "<span style=white-space:pre-wrap>        filter_target_id &lt;- extra_opts$filter_target_id</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        filter_target_id &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (\"normalize\" %in% names(extra_opts)) {</span>\n",
       "<span style=white-space:pre-wrap>        normalize &lt;- extra_opts$normalize</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        normalize &lt;- TRUE</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (\"norm_fun_counts\" %in% names(extra_opts)) {</span>\n",
       "<span style=white-space:pre-wrap>        norm_fun_counts &lt;- extra_opts$norm_fun_counts</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        norm_fun_counts &lt;- norm_factors</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (\"norm_fun_tpm\" %in% names(extra_opts)) {</span>\n",
       "<span style=white-space:pre-wrap>        norm_fun_tpm &lt;- extra_opts$norm_fun_tpm</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        norm_fun_tpm &lt;- norm_factors</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (\"transform_fun_counts\" %in% names(extra_opts)) {</span>\n",
       "<span style=white-space:pre-wrap>        transform_fun_counts &lt;- extra_opts$transform_fun_counts</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        transform_fun_counts &lt;- log_transform</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (\"transform_fun_tpm\" %in% names(extra_opts)) {</span>\n",
       "<span style=white-space:pre-wrap>        transform_fun_tpm &lt;- extra_opts$transform_fun_tpm</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        transform_fun_tpm &lt;- identity</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (\"gene_mode\" %in% names(extra_opts)) {</span>\n",
       "<span style=white-space:pre-wrap>        gene_mode &lt;- extra_opts$gene_mode</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        gene_mode &lt;- FALSE</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is(sample_to_covariates, \"data.frame\")) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(paste0(\"'\", substitute(sample_to_covariates), \"' (sample_to_covariates) must be a data.frame\"))</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is(full_model, \"formula\") &amp;&amp; !is(full_model, \"matrix\") &amp;&amp; </span>\n",
       "<span style=white-space:pre-wrap>        !is.null(full_model)) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(paste0(\"'\", substitute(full_model), \"' (full_model) must be a formula or a matrix\"))</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!(\"sample\" %in% colnames(sample_to_covariates))) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(paste0(\"'\", substitute(sample_to_covariates), \"' (sample_to_covariates) must contain a column named 'sample'\"))</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!(\"path\" %in% colnames(sample_to_covariates))) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(paste0(\"'\", substitute(sample_to_covariates)), \"' (sample_to_covariates) must contain a column named 'path'\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(target_mapping) &amp;&amp; !is(target_mapping, \"data.frame\")) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(paste0(\"'\", substitute(target_mapping), \"' (target_mapping) must be a data.frame or NULL\"))</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else if (is(target_mapping, \"data.frame\")) {</span>\n",
       "<span style=white-space:pre-wrap>        if (!(\"target_id\" %in% colnames(target_mapping))) {</span>\n",
       "<span style=white-space:pre-wrap>            stop(paste0(\"'\", substitute(target_mapping), \"' (target_mapping) must contain a column named 'target_id'\"))</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(max_bootstrap) &amp;&amp; max_bootstrap &lt;= 0) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"max_bootstrap must be &gt; 0\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (any(is.na(sample_to_covariates))) {</span>\n",
       "<span style=white-space:pre-wrap>        warning(\"Your 'sample_to_covariates' data.frame contains NA values. This will likely cause issues later.\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (is(full_model, \"matrix\") &amp;&amp; nrow(full_model) != nrow(sample_to_covariates)) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"The design matrix number of rows are not equal to the number of rows in the sample_to_covariates argument.\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is(filter_fun, \"function\")) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"filter_fun must be a function\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(filter_target_id) &amp; !is.character(filter_target_id)) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"if filter_target_id is used, it must be a character vector\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is(norm_fun_counts, \"function\")) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"norm_fun_counts must be a function\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is(norm_fun_tpm, \"function\")) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"norm_fun_tpm must be a function\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is(transform_fun_counts, \"function\")) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"transform_fun_counts must be a function\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is(transform_fun_tpm, \"function\")) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"transform_fun_tpm must be a function\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (is.null(aggregation_column) &amp;&amp; gene_mode) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"You set 'gene_mode' to TRUE, but did not provide an 'aggregation_column' \", </span>\n",
       "<span style=white-space:pre-wrap>            \"to aggregate by. Please provide an 'aggregation_column'.\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else if (gene_mode) {</span>\n",
       "<span style=white-space:pre-wrap>        message(\"'gene_mode' is TRUE. Sleuth will do counts aggregation at the gene level \", </span>\n",
       "<span style=white-space:pre-wrap>            \"for downstream normalization, transformation, and modeling steps, as well as \", </span>\n",
       "<span style=white-space:pre-wrap>            \"for plotting and results.\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(aggregation_column) &amp;&amp; is.null(target_mapping)) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(paste(\"You provided an 'aggregation_column' to aggregate by,\", </span>\n",
       "<span style=white-space:pre-wrap>            \"but not a 'target_mapping'. Please provide a 'target_mapping'.\"))</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    pval_aggregate &lt;- !is.null(aggregation_column) &amp;&amp; !gene_mode</span>\n",
       "<span style=white-space:pre-wrap>    num_cores &lt;- check_num_cores(num_cores)</span>\n",
       "<span style=white-space:pre-wrap>    msg(\"reading in kallisto results\")</span>\n",
       "<span style=white-space:pre-wrap>    sample_to_covariates &lt;- as.data.frame(sample_to_covariates)</span>\n",
       "<span style=white-space:pre-wrap>    sample_to_covariates$sample &lt;- as.character(sample_to_covariates$sample)</span>\n",
       "<span style=white-space:pre-wrap>    if (nrow(sample_to_covariates) == 1 &amp;&amp; !is.null(full_model)) {</span>\n",
       "<span style=white-space:pre-wrap>        warning(\"There is only one sample present, but you also provided a model. \", </span>\n",
       "<span style=white-space:pre-wrap>            \"The model will be set to NULL to prevent downstream errors.\\n\", </span>\n",
       "<span style=white-space:pre-wrap>            \"The sample can be viewed using sleuth_live after preparation, \", </span>\n",
       "<span style=white-space:pre-wrap>            \"but you need more than one sample to run the other aspects of Sleuth.\")</span>\n",
       "<span style=white-space:pre-wrap>        full_model &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    kal_dirs &lt;- sample_to_covariates$path</span>\n",
       "<span style=white-space:pre-wrap>    sample_to_covariates$path &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>    msg(\"dropping unused factor levels\")</span>\n",
       "<span style=white-space:pre-wrap>    sample_to_covariates &lt;- droplevels(sample_to_covariates)</span>\n",
       "<span style=white-space:pre-wrap>    nsamp &lt;- 0</span>\n",
       "<span style=white-space:pre-wrap>    kal_list &lt;- lapply(seq_along(kal_dirs), function(i) {</span>\n",
       "<span style=white-space:pre-wrap>        nsamp &lt;- dot(nsamp)</span>\n",
       "<span style=white-space:pre-wrap>        path &lt;- kal_dirs[i]</span>\n",
       "<span style=white-space:pre-wrap>        suppressMessages({</span>\n",
       "<span style=white-space:pre-wrap>            kal &lt;- read_kallisto(path, read_bootstrap = FALSE, </span>\n",
       "<span style=white-space:pre-wrap>                max_bootstrap = max_bootstrap)</span>\n",
       "<span style=white-space:pre-wrap>        })</span>\n",
       "<span style=white-space:pre-wrap>        kal$abundance &lt;- dplyr::mutate(kal$abundance, sample = sample_to_covariates$sample[i])</span>\n",
       "<span style=white-space:pre-wrap>        kal</span>\n",
       "<span style=white-space:pre-wrap>    })</span>\n",
       "<span style=white-space:pre-wrap>    msg(\"\")</span>\n",
       "<span style=white-space:pre-wrap>    check_result &lt;- check_kal_pack(kal_list)</span>\n",
       "<span style=white-space:pre-wrap>    kal_versions &lt;- check_result$versions</span>\n",
       "<span style=white-space:pre-wrap>    obs_raw &lt;- dplyr::bind_rows(lapply(kal_list, function(k) k$abundance))</span>\n",
       "<span style=white-space:pre-wrap>    counts_test &lt;- data.table::as.data.table(obs_raw)</span>\n",
       "<span style=white-space:pre-wrap>    counts_test &lt;- counts_test[, .(total = sum(est_counts)), </span>\n",
       "<span style=white-space:pre-wrap>        by = \"sample\"]</span>\n",
       "<span style=white-space:pre-wrap>    if (any(counts_test$total == 0)) {</span>\n",
       "<span style=white-space:pre-wrap>        zero_names &lt;- counts_test$sample[which(counts_test$total == </span>\n",
       "<span style=white-space:pre-wrap>            0)]</span>\n",
       "<span style=white-space:pre-wrap>        formatted_names &lt;- paste(zero_names, collapse = \", \")</span>\n",
       "<span style=white-space:pre-wrap>        warning(\"At least one sample have no reads aligned. \", </span>\n",
       "<span style=white-space:pre-wrap>            \"Here are the samples with zero counts:\\n\", formatted_names)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    design_matrix &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>    if (is(full_model, \"formula\")) {</span>\n",
       "<span style=white-space:pre-wrap>        design_matrix &lt;- model.matrix(full_model, sample_to_covariates)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else if (is(full_model, \"matrix\")) {</span>\n",
       "<span style=white-space:pre-wrap>        if (is.null(colnames(full_model))) {</span>\n",
       "<span style=white-space:pre-wrap>            stop(\"If matrix is supplied, column names must also be supplied.\")</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        design_matrix &lt;- full_model</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(full_model)) {</span>\n",
       "<span style=white-space:pre-wrap>        rownames(design_matrix) &lt;- sample_to_covariates$sample</span>\n",
       "<span style=white-space:pre-wrap>        M &lt;- t(design_matrix) %*% design_matrix</span>\n",
       "<span style=white-space:pre-wrap>        det_mod &lt;- determinant(M)$modulus</span>\n",
       "<span style=white-space:pre-wrap>        if (!is.finite(det_mod)) {</span>\n",
       "<span style=white-space:pre-wrap>            stop(\"The full model you provided seems to result in a singular design matrix. \", </span>\n",
       "<span style=white-space:pre-wrap>                \"This frequently happens when one of the covariates is a linear \", </span>\n",
       "<span style=white-space:pre-wrap>                \"combination of one or more other covariates (e.g. one covariate \", </span>\n",
       "<span style=white-space:pre-wrap>                \"yields identical groupings as another covariate). Check your \", </span>\n",
       "<span style=white-space:pre-wrap>                \"sample_to_covariates table and your full model.\")</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        rm(M, det_mod)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    obs_raw &lt;- dplyr::arrange(obs_raw, target_id, sample)</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(target_mapping)) {</span>\n",
       "<span style=white-space:pre-wrap>        tmp_names &lt;- data.frame(target_id = kal_list[[1]]$abundance$target_id, </span>\n",
       "<span style=white-space:pre-wrap>            stringsAsFactors = FALSE)</span>\n",
       "<span style=white-space:pre-wrap>        target_mapping &lt;- check_target_mapping(tmp_names, target_mapping, </span>\n",
       "<span style=white-space:pre-wrap>            !is.null(aggregation_column))</span>\n",
       "<span style=white-space:pre-wrap>        rm(tmp_names)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    ret &lt;- list(kal = kal_list, kal_versions = kal_versions, </span>\n",
       "<span style=white-space:pre-wrap>        obs_raw = obs_raw, sample_to_covariates = sample_to_covariates, </span>\n",
       "<span style=white-space:pre-wrap>        bootstrap_summary = NA, full_formula = full_model, design_matrix = design_matrix, </span>\n",
       "<span style=white-space:pre-wrap>        target_mapping = target_mapping, gene_mode = gene_mode, </span>\n",
       "<span style=white-space:pre-wrap>        gene_column = aggregation_column, norm_fun_counts = norm_fun_counts, </span>\n",
       "<span style=white-space:pre-wrap>        norm_fun_tpm = norm_fun_tpm, transform_fun_counts = transform_fun_counts, </span>\n",
       "<span style=white-space:pre-wrap>        transform_fun_tpm = transform_fun_tpm, pval_aggregate = pval_aggregate)</span>\n",
       "<span style=white-space:pre-wrap>    if (normalize) {</span>\n",
       "<span style=white-space:pre-wrap>        msg(\"normalizing est_counts\")</span>\n",
       "<span style=white-space:pre-wrap>        est_counts_spread &lt;- spread_abundance_by(obs_raw, \"est_counts\", </span>\n",
       "<span style=white-space:pre-wrap>            sample_to_covariates$sample)</span>\n",
       "<span style=white-space:pre-wrap>        if (!is.null(filter_target_id)) {</span>\n",
       "<span style=white-space:pre-wrap>            msg(\"A list of target IDs for filtering was found. Using this for filtering\")</span>\n",
       "<span style=white-space:pre-wrap>            target_ids &lt;- rownames(est_counts_spread)</span>\n",
       "<span style=white-space:pre-wrap>            filter_bool &lt;- target_ids %in% filter_target_id</span>\n",
       "<span style=white-space:pre-wrap>            names(filter_bool) &lt;- target_ids</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            filter_bool &lt;- apply(est_counts_spread, 1, filter_fun)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        filter_true &lt;- filter_bool[filter_bool]</span>\n",
       "<span style=white-space:pre-wrap>        if (sum(filter_bool) == 0) {</span>\n",
       "<span style=white-space:pre-wrap>            stop(\"Zero targets passed the filter you used. Please double check the filter used.\")</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        msg(paste0(sum(filter_bool), \" targets passed the filter\"))</span>\n",
       "<span style=white-space:pre-wrap>        est_counts_sf &lt;- norm_fun_counts(est_counts_spread[filter_bool, </span>\n",
       "<span style=white-space:pre-wrap>            , drop = FALSE])</span>\n",
       "<span style=white-space:pre-wrap>        filter_df &lt;- adf(target_id = names(filter_true))</span>\n",
       "<span style=white-space:pre-wrap>        est_counts_norm &lt;- as_df(t(t(est_counts_spread)/est_counts_sf))</span>\n",
       "<span style=white-space:pre-wrap>        est_counts_norm$target_id &lt;- rownames(est_counts_norm)</span>\n",
       "<span style=white-space:pre-wrap>        est_counts_norm &lt;- tidyr::gather(est_counts_norm, sample, </span>\n",
       "<span style=white-space:pre-wrap>            est_counts, -target_id)</span>\n",
       "<span style=white-space:pre-wrap>        obs_norm &lt;- est_counts_norm</span>\n",
       "<span style=white-space:pre-wrap>        obs_norm$target_id &lt;- as.character(obs_norm$target_id)</span>\n",
       "<span style=white-space:pre-wrap>        obs_norm$sample &lt;- as.character(obs_norm$sample)</span>\n",
       "<span style=white-space:pre-wrap>        rm(est_counts_norm)</span>\n",
       "<span style=white-space:pre-wrap>        msg(\"normalizing tpm\")</span>\n",
       "<span style=white-space:pre-wrap>        tpm_spread &lt;- spread_abundance_by(obs_raw, \"tpm\", sample_to_covariates$sample)</span>\n",
       "<span style=white-space:pre-wrap>        tpm_sf &lt;- norm_fun_tpm(tpm_spread[filter_bool, , drop = FALSE])</span>\n",
       "<span style=white-space:pre-wrap>        tpm_norm &lt;- as_df(t(t(tpm_spread)/tpm_sf))</span>\n",
       "<span style=white-space:pre-wrap>        tpm_norm$target_id &lt;- rownames(tpm_norm)</span>\n",
       "<span style=white-space:pre-wrap>        tpm_norm &lt;- tidyr::gather(tpm_norm, sample, tpm, -target_id)</span>\n",
       "<span style=white-space:pre-wrap>        tpm_norm$sample &lt;- as.character(tpm_norm$sample)</span>\n",
       "<span style=white-space:pre-wrap>        msg(\"merging in metadata\")</span>\n",
       "<span style=white-space:pre-wrap>        obs_norm &lt;- dplyr::arrange(obs_norm, target_id, sample)</span>\n",
       "<span style=white-space:pre-wrap>        tpm_norm &lt;- dplyr::arrange(tpm_norm, target_id, sample)</span>\n",
       "<span style=white-space:pre-wrap>        stopifnot(all.equal(obs_raw$target_id, obs_norm$target_id) &amp;&amp; </span>\n",
       "<span style=white-space:pre-wrap>            all.equal(obs_raw$sample, obs_norm$sample))</span>\n",
       "<span style=white-space:pre-wrap>        suppressWarnings({</span>\n",
       "<span style=white-space:pre-wrap>            if (!all.equal(dplyr::select(obs_norm, target_id, </span>\n",
       "<span style=white-space:pre-wrap>                sample), dplyr::select(tpm_norm, target_id, sample))) {</span>\n",
       "<span style=white-space:pre-wrap>                stop(\"Invalid column rows. In principle, can simply join. Please report error.\")</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            obs_norm &lt;- dplyr::bind_cols(obs_norm, dplyr::select(tpm_norm, </span>\n",
       "<span style=white-space:pre-wrap>                tpm))</span>\n",
       "<span style=white-space:pre-wrap>        })</span>\n",
       "<span style=white-space:pre-wrap>        obs_norm &lt;- dplyr::bind_cols(obs_norm, dplyr::select(obs_raw, </span>\n",
       "<span style=white-space:pre-wrap>            eff_len, len))</span>\n",
       "<span style=white-space:pre-wrap>        obs_norm &lt;- as_df(obs_norm)</span>\n",
       "<span style=white-space:pre-wrap>        ret$obs_norm &lt;- obs_norm</span>\n",
       "<span style=white-space:pre-wrap>        ret$est_counts_sf &lt;- est_counts_sf</span>\n",
       "<span style=white-space:pre-wrap>        ret$filter_bool &lt;- filter_bool</span>\n",
       "<span style=white-space:pre-wrap>        ret$filter_df &lt;- filter_df</span>\n",
       "<span style=white-space:pre-wrap>        ret$obs_norm_filt &lt;- dplyr::semi_join(obs_norm, filter_df, </span>\n",
       "<span style=white-space:pre-wrap>            by = \"target_id\")</span>\n",
       "<span style=white-space:pre-wrap>        ret$tpm_sf &lt;- tpm_sf</span>\n",
       "<span style=white-space:pre-wrap>        path &lt;- kal_dirs[1]</span>\n",
       "<span style=white-space:pre-wrap>        kal_path &lt;- get_kallisto_path(path)</span>\n",
       "<span style=white-space:pre-wrap>        target_id &lt;- as.character(rhdf5::h5read(kal_path$path, </span>\n",
       "<span style=white-space:pre-wrap>            \"aux/ids\"))</span>\n",
       "<span style=white-space:pre-wrap>        num_transcripts &lt;- length(target_id)</span>\n",
       "<span style=white-space:pre-wrap>        ret$bs_quants &lt;- list()</span>\n",
       "<span style=white-space:pre-wrap>        which_target_id &lt;- ret$filter_df$target_id</span>\n",
       "<span style=white-space:pre-wrap>        if (ret$gene_mode) {</span>\n",
       "<span style=white-space:pre-wrap>            msg(paste0(\"aggregating by column: \", aggregation_column))</span>\n",
       "<span style=white-space:pre-wrap>            target_mapping &lt;- data.table::data.table(target_mapping)</span>\n",
       "<span style=white-space:pre-wrap>            target_mapping[target_mapping[[aggregation_column]] == </span>\n",
       "<span style=white-space:pre-wrap>                \"\", aggregation_column] &lt;- NA</span>\n",
       "<span style=white-space:pre-wrap>            agg_id &lt;- unique(target_mapping[, aggregation_column, </span>\n",
       "<span style=white-space:pre-wrap>                with = FALSE])</span>\n",
       "<span style=white-space:pre-wrap>            agg_id &lt;- agg_id[[1]]</span>\n",
       "<span style=white-space:pre-wrap>            agg_id &lt;- agg_id[!is.na(agg_id)]</span>\n",
       "<span style=white-space:pre-wrap>            mappings &lt;- dplyr::select_(target_mapping, \"target_id\", </span>\n",
       "<span style=white-space:pre-wrap>                aggregation_column)</span>\n",
       "<span style=white-space:pre-wrap>            mappings &lt;- data.table::as.data.table(mappings)</span>\n",
       "<span style=white-space:pre-wrap>            which_tms &lt;- which(mappings$target_id %in% which_target_id)</span>\n",
       "<span style=white-space:pre-wrap>            which_agg_id &lt;- unique(mappings[which_tms, aggregation_column, </span>\n",
       "<span style=white-space:pre-wrap>                with = FALSE])</span>\n",
       "<span style=white-space:pre-wrap>            which_agg_id &lt;- which_agg_id[[1]]</span>\n",
       "<span style=white-space:pre-wrap>            which_agg_id &lt;- which_agg_id[!is.na(which_agg_id)]</span>\n",
       "<span style=white-space:pre-wrap>            filter_df &lt;- adf(target_id = which_agg_id)</span>\n",
       "<span style=white-space:pre-wrap>            filter_bool &lt;- agg_id %in% which_agg_id</span>\n",
       "<span style=white-space:pre-wrap>            msg(paste0(length(which_agg_id), \" genes passed the filter\"))</span>\n",
       "<span style=white-space:pre-wrap>            norm_by_length &lt;- TRUE</span>\n",
       "<span style=white-space:pre-wrap>            tmp &lt;- data.table::as.data.table(ret$obs_raw)</span>\n",
       "<span style=white-space:pre-wrap>            tmp &lt;- merge(tmp, mappings, by = \"target_id\", all.x = TRUE)</span>\n",
       "<span style=white-space:pre-wrap>            scale_factor &lt;- tmp[, `:=`(scale_factor, median(eff_len)), </span>\n",
       "<span style=white-space:pre-wrap>                by = list(sample, eval(parse(text = aggregation_column)))]</span>\n",
       "<span style=white-space:pre-wrap>            obs_norm_gene &lt;- reads_per_base_transform(ret$obs_norm, </span>\n",
       "<span style=white-space:pre-wrap>                scale_factor, aggregation_column, mappings, norm_by_length)</span>\n",
       "<span style=white-space:pre-wrap>            tmp &lt;- data.table::as.data.table(tpm_norm)</span>\n",
       "<span style=white-space:pre-wrap>            tmp &lt;- merge(tmp, mappings, by = \"target_id\", all.x = T)</span>\n",
       "<span style=white-space:pre-wrap>            if (any(is.na(tmp[[aggregation_column]]))) {</span>\n",
       "<span style=white-space:pre-wrap>                rows_to_remove &lt;- is.na(tmp[[aggregation_column]])</span>\n",
       "<span style=white-space:pre-wrap>                num_missing &lt;- length(unique(tmp[rows_to_remove, </span>\n",
       "<span style=white-space:pre-wrap>                  target_id]))</span>\n",
       "<span style=white-space:pre-wrap>                warning(num_missing, \" target_ids are missing annotations for the aggregation_column: \", </span>\n",
       "<span style=white-space:pre-wrap>                  aggregation_column, \".\\nThese target_ids will be dropped from the gene-level analysis.\", </span>\n",
       "<span style=white-space:pre-wrap>                  \"\\nIf you did not expect this, check your 'target_mapping' table for missing values.\")</span>\n",
       "<span style=white-space:pre-wrap>                tmp &lt;- tmp[!rows_to_remove]</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            tpm_norm_gene &lt;- tmp[, j = list(tpm = sum(tpm)), </span>\n",
       "<span style=white-space:pre-wrap>                by = list(sample, eval(parse(text = aggregation_column)))]</span>\n",
       "<span style=white-space:pre-wrap>            data.table::setnames(tpm_norm_gene, \"parse\", \"target_id\")</span>\n",
       "<span style=white-space:pre-wrap>            tpm_norm_gene &lt;- as_df(tpm_norm_gene)</span>\n",
       "<span style=white-space:pre-wrap>            obs_norm_gene &lt;- dplyr::arrange(obs_norm_gene, target_id, </span>\n",
       "<span style=white-space:pre-wrap>                sample)</span>\n",
       "<span style=white-space:pre-wrap>            tpm_norm_gene &lt;- dplyr::arrange(tpm_norm_gene, target_id, </span>\n",
       "<span style=white-space:pre-wrap>                sample)</span>\n",
       "<span style=white-space:pre-wrap>            stopifnot(all.equal(dplyr::select(obs_norm_gene, </span>\n",
       "<span style=white-space:pre-wrap>                target_id, sample), dplyr::select(tpm_norm_gene, </span>\n",
       "<span style=white-space:pre-wrap>                target_id, sample)))</span>\n",
       "<span style=white-space:pre-wrap>            suppressWarnings({</span>\n",
       "<span style=white-space:pre-wrap>                if (!all.equal(dplyr::select(obs_norm_gene, target_id, </span>\n",
       "<span style=white-space:pre-wrap>                  sample), dplyr::select(tpm_norm_gene, target_id, </span>\n",
       "<span style=white-space:pre-wrap>                  sample), check.attributes = FALSE)) {</span>\n",
       "<span style=white-space:pre-wrap>                  stop(\"Invalid column rows. In principle, can simply join. Please report error.\")</span>\n",
       "<span style=white-space:pre-wrap>                }</span>\n",
       "<span style=white-space:pre-wrap>                obs_norm_gene &lt;- dplyr::bind_cols(obs_norm_gene, </span>\n",
       "<span style=white-space:pre-wrap>                  dplyr::select(tpm_norm_gene, tpm))</span>\n",
       "<span style=white-space:pre-wrap>            })</span>\n",
       "<span style=white-space:pre-wrap>            ret$filter_df &lt;- adf(target_id = which_agg_id)</span>\n",
       "<span style=white-space:pre-wrap>            ret$filter_bool &lt;- agg_id %in% which_agg_id</span>\n",
       "<span style=white-space:pre-wrap>            ret$obs_norm &lt;- obs_norm_gene</span>\n",
       "<span style=white-space:pre-wrap>            ret$obs_norm_filt &lt;- dplyr::semi_join(obs_norm_gene, </span>\n",
       "<span style=white-space:pre-wrap>                filter_df, by = \"target_id\")</span>\n",
       "<span style=white-space:pre-wrap>            rm(obs_norm, tpm_norm, obs_norm_gene, tpm_norm_gene)</span>\n",
       "<span style=white-space:pre-wrap>            all_sample_bootstrap &lt;- matrix(NA_real_, nrow = length(which_agg_id), </span>\n",
       "<span style=white-space:pre-wrap>                ncol = length(ret$kal))</span>\n",
       "<span style=white-space:pre-wrap>            which_ids &lt;- which_agg_id</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            all_sample_bootstrap &lt;- matrix(NA_real_, nrow = length(which_target_id), </span>\n",
       "<span style=white-space:pre-wrap>                ncol = length(ret$kal))</span>\n",
       "<span style=white-space:pre-wrap>            which_ids &lt;- which_target_id</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        msg(\"summarizing bootstraps\")</span>\n",
       "<span style=white-space:pre-wrap>        apply_function &lt;- if (num_cores == 1) {</span>\n",
       "<span style=white-space:pre-wrap>            lapply</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            function(x, y) parallel::mclapply(x, y, mc.cores = num_cores)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        bs_results &lt;- apply_function(seq_along(kal_dirs), function(i) {</span>\n",
       "<span style=white-space:pre-wrap>            samp_name &lt;- sample_to_covariates$sample[i]</span>\n",
       "<span style=white-space:pre-wrap>            kal_path &lt;- get_kallisto_path(kal_dirs[i])</span>\n",
       "<span style=white-space:pre-wrap>            process_bootstrap(i, samp_name, kal_path, num_transcripts, </span>\n",
       "<span style=white-space:pre-wrap>                est_counts_sf[[i]], read_bootstrap_tpm, ret$gene_mode, </span>\n",
       "<span style=white-space:pre-wrap>                extra_bootstrap_summary, target_id, mappings, </span>\n",
       "<span style=white-space:pre-wrap>                which_ids, ret$gene_column, ret$transform_fun_counts, </span>\n",
       "<span style=white-space:pre-wrap>                ret$transform_fun_tpm, max_bootstrap)</span>\n",
       "<span style=white-space:pre-wrap>        })</span>\n",
       "<span style=white-space:pre-wrap>        error_status &lt;- sapply(bs_results, function(x) is(x, </span>\n",
       "<span style=white-space:pre-wrap>            \"try-error\"))</span>\n",
       "<span style=white-space:pre-wrap>        if (any(error_status)) {</span>\n",
       "<span style=white-space:pre-wrap>            error_msgs &lt;- sapply(which(error_status), function(i) {</span>\n",
       "<span style=white-space:pre-wrap>                bad_run &lt;- bs_results[[i]]</span>\n",
       "<span style=white-space:pre-wrap>                samp_name &lt;- sample_to_covariates$sample[i]</span>\n",
       "<span style=white-space:pre-wrap>                trace &lt;- .traceback(bad_run)</span>\n",
       "<span style=white-space:pre-wrap>                paste0(\"Sample '\", samp_name, \"' had this error message: \", </span>\n",
       "<span style=white-space:pre-wrap>                  trace[1])</span>\n",
       "<span style=white-space:pre-wrap>            })</span>\n",
       "<span style=white-space:pre-wrap>            formatted_error &lt;- paste(error_msgs, collapse = \"\")</span>\n",
       "<span style=white-space:pre-wrap>            message(formatted_error)</span>\n",
       "<span style=white-space:pre-wrap>            stop(\"At least one core from mclapply had an error. See the above error message(s) for more details.\")</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        indices &lt;- sapply(bs_results, function(result) result$index)</span>\n",
       "<span style=white-space:pre-wrap>        stopifnot(identical(indices, order(indices)))</span>\n",
       "<span style=white-space:pre-wrap>        sigma_q_sq_tpm &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>        if (read_bootstrap_tpm) {</span>\n",
       "<span style=white-space:pre-wrap>            all_sample_tpm &lt;- sapply(bs_results, function(result) result$bootstrap_tpm_result)</span>\n",
       "<span style=white-space:pre-wrap>            rownames(all_sample_tpm) &lt;- which_ids</span>\n",
       "<span style=white-space:pre-wrap>            sigma_q_sq_tpm &lt;- rowMeans(all_sample_tpm)</span>\n",
       "<span style=white-space:pre-wrap>            sigma_q_sq_tpm &lt;- sigma_q_sq_tpm[order(names(sigma_q_sq_tpm))]</span>\n",
       "<span style=white-space:pre-wrap>            ret$bs_quants &lt;- lapply(bs_results, function(result) result$bs_quants)</span>\n",
       "<span style=white-space:pre-wrap>            names(ret$bs_quants) &lt;- sample_to_covariates$sample</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else if (extra_bootstrap_summary) {</span>\n",
       "<span style=white-space:pre-wrap>            ret$bs_quants &lt;- lapply(bs_results, function(result) result$bs_quants)</span>\n",
       "<span style=white-space:pre-wrap>            names(ret$bs_quants) &lt;- sample_to_covariates$sample</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        all_sample_bootstrap &lt;- sapply(bs_results, function(result) result$bootstrap_result)</span>\n",
       "<span style=white-space:pre-wrap>        rownames(all_sample_bootstrap) &lt;- which_ids</span>\n",
       "<span style=white-space:pre-wrap>        msg(\"\")</span>\n",
       "<span style=white-space:pre-wrap>        sigma_q_sq &lt;- rowMeans(all_sample_bootstrap)</span>\n",
       "<span style=white-space:pre-wrap>        names(sigma_q_sq) &lt;- which_ids</span>\n",
       "<span style=white-space:pre-wrap>        if (ret$gene_mode) {</span>\n",
       "<span style=white-space:pre-wrap>            obs_counts &lt;- obs_to_matrix(ret, \"scaled_reads_per_base\")[which_agg_id, </span>\n",
       "<span style=white-space:pre-wrap>                , drop = FALSE]</span>\n",
       "<span style=white-space:pre-wrap>            obs_tpm &lt;- obs_to_matrix(ret, \"tpm\")[which_agg_id, </span>\n",
       "<span style=white-space:pre-wrap>                , drop = FALSE]</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            obs_counts &lt;- obs_to_matrix(ret, \"est_counts\")[which_target_id, </span>\n",
       "<span style=white-space:pre-wrap>                , drop = FALSE]</span>\n",
       "<span style=white-space:pre-wrap>            obs_tpm &lt;- obs_to_matrix(ret, \"tpm\")[which_target_id, </span>\n",
       "<span style=white-space:pre-wrap>                , drop = FALSE]</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        sigma_q_sq &lt;- sigma_q_sq[order(names(sigma_q_sq))]</span>\n",
       "<span style=white-space:pre-wrap>        obs_counts &lt;- ret$transform_fun_counts(obs_counts)</span>\n",
       "<span style=white-space:pre-wrap>        obs_counts &lt;- obs_counts[order(rownames(obs_counts)), </span>\n",
       "<span style=white-space:pre-wrap>            ]</span>\n",
       "<span style=white-space:pre-wrap>        obs_tpm &lt;- ret$transform_fun_tpm(obs_tpm)</span>\n",
       "<span style=white-space:pre-wrap>        obs_tpm &lt;- obs_tpm[order(rownames(obs_tpm)), ]</span>\n",
       "<span style=white-space:pre-wrap>        ret$bs_summary &lt;- list(obs_counts = obs_counts, sigma_q_sq = sigma_q_sq)</span>\n",
       "<span style=white-space:pre-wrap>        ret$bs_summary &lt;- list(obs_counts = obs_counts, obs_tpm = obs_tpm, </span>\n",
       "<span style=white-space:pre-wrap>            sigma_q_sq = sigma_q_sq, sigma_q_sq_tpm = sigma_q_sq_tpm)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        ret$norm_fun_counts &lt;- ret$norm_fun_tpm &lt;- NA</span>\n",
       "<span style=white-space:pre-wrap>        ret$transform_fun_counts &lt;- ret$transform_fun_tpm &lt;- NA</span>\n",
       "<span style=white-space:pre-wrap>        ret$obs_norm &lt;- data.frame()</span>\n",
       "<span style=white-space:pre-wrap>        ret$est_counts_sf &lt;- vector()</span>\n",
       "<span style=white-space:pre-wrap>        est_counts_spread &lt;- spread_abundance_by(obs_raw, \"est_counts\", </span>\n",
       "<span style=white-space:pre-wrap>            sample_to_covariates$sample)</span>\n",
       "<span style=white-space:pre-wrap>        if (!is.null(filter_target_id)) {</span>\n",
       "<span style=white-space:pre-wrap>            msg(\"A list of target IDs for filtering was found. Using this for filtering\")</span>\n",
       "<span style=white-space:pre-wrap>            target_ids &lt;- rownames(est_counts_spread)</span>\n",
       "<span style=white-space:pre-wrap>            filter_bool &lt;- target_ids %in% filter_target_id</span>\n",
       "<span style=white-space:pre-wrap>            names(filter_bool) &lt;- target_ids</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            filter_bool &lt;- apply(est_counts_spread, 1, filter_fun)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        filter_true &lt;- filter_bool[filter_bool]</span>\n",
       "<span style=white-space:pre-wrap>        if (sum(filter_bool) == 0) {</span>\n",
       "<span style=white-space:pre-wrap>            stop(\"Zero targets passed the filter you used. Please double check the filter used.\")</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        msg(paste0(sum(filter_bool), \" targets passed the filter\"))</span>\n",
       "<span style=white-space:pre-wrap>        filter_df &lt;- adf(target_id = names(filter_true))</span>\n",
       "<span style=white-space:pre-wrap>        if (ret$gene_mode) {</span>\n",
       "<span style=white-space:pre-wrap>            msg(paste0(\"aggregating by column: \", aggregation_column))</span>\n",
       "<span style=white-space:pre-wrap>            which_target_id &lt;- ret$filter_df$target_id</span>\n",
       "<span style=white-space:pre-wrap>            target_mapping &lt;- data.table::data.table(target_mapping)</span>\n",
       "<span style=white-space:pre-wrap>            target_mapping[target_mapping[[aggregation_column]] == </span>\n",
       "<span style=white-space:pre-wrap>                \"\", aggregation_column] &lt;- NA</span>\n",
       "<span style=white-space:pre-wrap>            agg_id &lt;- unique(target_mapping[, aggregation_column, </span>\n",
       "<span style=white-space:pre-wrap>                with = FALSE])</span>\n",
       "<span style=white-space:pre-wrap>            agg_id &lt;- agg_id[[1]]</span>\n",
       "<span style=white-space:pre-wrap>            agg_id &lt;- agg_id[!is.na(agg_id)]</span>\n",
       "<span style=white-space:pre-wrap>            mappings &lt;- dplyr::select_(target_mapping, \"target_id\", </span>\n",
       "<span style=white-space:pre-wrap>                aggregation_column)</span>\n",
       "<span style=white-space:pre-wrap>            mappings &lt;- data.table::as.data.table(mappings)</span>\n",
       "<span style=white-space:pre-wrap>            which_tms &lt;- which(mappings$target_id %in% which_target_id)</span>\n",
       "<span style=white-space:pre-wrap>            which_agg_id &lt;- unique(mappings[which_tms, aggregation_column, </span>\n",
       "<span style=white-space:pre-wrap>                with = FALSE])</span>\n",
       "<span style=white-space:pre-wrap>            which_agg_id &lt;- which_agg_id[[1]]</span>\n",
       "<span style=white-space:pre-wrap>            which_agg_id &lt;- which_agg_id[!is.na(which_agg_id)]</span>\n",
       "<span style=white-space:pre-wrap>            filter_df &lt;- adf(target_id = which_agg_id)</span>\n",
       "<span style=white-space:pre-wrap>            filter_bool &lt;- agg_id %in% which_agg_id</span>\n",
       "<span style=white-space:pre-wrap>            msg(paste0(length(which_agg_id), \" genes passed the filter\"))</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        ret$filter_bool &lt;- filter_bool</span>\n",
       "<span style=white-space:pre-wrap>        ret$filter_df &lt;- filter_df</span>\n",
       "<span style=white-space:pre-wrap>        ret$obs_norm_filt &lt;- data.frame()</span>\n",
       "<span style=white-space:pre-wrap>        ret$tpm_sf &lt;- vector()</span>\n",
       "<span style=white-space:pre-wrap>        ret$bs_quants &lt;- list()</span>\n",
       "<span style=white-space:pre-wrap>        ret$bs_summary &lt;- list()</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    class(ret) &lt;- \"sleuth\"</span>\n",
       "<span style=white-space:pre-wrap>    ret</span>\n",
       "}</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (sample\\_to\\_covariates, full\\_model = NULL, target\\_mapping = NULL, \n",
       "    aggregation\\_column = NULL, num\\_cores = max(1L, parallel::detectCores() - \n",
       "        1L), ...) \n",
       "\\{\n",
       "    extra\\_opts <- list(...)\n",
       "    if (\"extra\\_bootstrap\\_summary\" \\%in\\% names(extra\\_opts)) \\{\n",
       "        extra\\_bootstrap\\_summary <- extra\\_opts\\$extra\\_bootstrap\\_summary\n",
       "    \\}\n",
       "    else \\{\n",
       "        extra\\_bootstrap\\_summary <- FALSE\n",
       "    \\}\n",
       "    if (\"read\\_bootstrap\\_tpm\" \\%in\\% names(extra\\_opts)) \\{\n",
       "        read\\_bootstrap\\_tpm <- extra\\_opts\\$read\\_bootstrap\\_tpm\n",
       "    \\}\n",
       "    else \\{\n",
       "        read\\_bootstrap\\_tpm <- FALSE\n",
       "    \\}\n",
       "    if (\"max\\_bootstrap\" \\%in\\% names(extra\\_opts)) \\{\n",
       "        max\\_bootstrap <- extra\\_opts\\$max\\_bootstrap\n",
       "    \\}\n",
       "    else \\{\n",
       "        max\\_bootstrap <- NULL\n",
       "    \\}\n",
       "    if (\"filter\\_fun\" \\%in\\% names(extra\\_opts)) \\{\n",
       "        filter\\_fun <- extra\\_opts\\$filter\\_fun\n",
       "    \\}\n",
       "    else \\{\n",
       "        filter\\_fun <- basic\\_filter\n",
       "    \\}\n",
       "    if (\"filter\\_target\\_id\" \\%in\\% names(extra\\_opts)) \\{\n",
       "        filter\\_target\\_id <- extra\\_opts\\$filter\\_target\\_id\n",
       "    \\}\n",
       "    else \\{\n",
       "        filter\\_target\\_id <- NULL\n",
       "    \\}\n",
       "    if (\"normalize\" \\%in\\% names(extra\\_opts)) \\{\n",
       "        normalize <- extra\\_opts\\$normalize\n",
       "    \\}\n",
       "    else \\{\n",
       "        normalize <- TRUE\n",
       "    \\}\n",
       "    if (\"norm\\_fun\\_counts\" \\%in\\% names(extra\\_opts)) \\{\n",
       "        norm\\_fun\\_counts <- extra\\_opts\\$norm\\_fun\\_counts\n",
       "    \\}\n",
       "    else \\{\n",
       "        norm\\_fun\\_counts <- norm\\_factors\n",
       "    \\}\n",
       "    if (\"norm\\_fun\\_tpm\" \\%in\\% names(extra\\_opts)) \\{\n",
       "        norm\\_fun\\_tpm <- extra\\_opts\\$norm\\_fun\\_tpm\n",
       "    \\}\n",
       "    else \\{\n",
       "        norm\\_fun\\_tpm <- norm\\_factors\n",
       "    \\}\n",
       "    if (\"transform\\_fun\\_counts\" \\%in\\% names(extra\\_opts)) \\{\n",
       "        transform\\_fun\\_counts <- extra\\_opts\\$transform\\_fun\\_counts\n",
       "    \\}\n",
       "    else \\{\n",
       "        transform\\_fun\\_counts <- log\\_transform\n",
       "    \\}\n",
       "    if (\"transform\\_fun\\_tpm\" \\%in\\% names(extra\\_opts)) \\{\n",
       "        transform\\_fun\\_tpm <- extra\\_opts\\$transform\\_fun\\_tpm\n",
       "    \\}\n",
       "    else \\{\n",
       "        transform\\_fun\\_tpm <- identity\n",
       "    \\}\n",
       "    if (\"gene\\_mode\" \\%in\\% names(extra\\_opts)) \\{\n",
       "        gene\\_mode <- extra\\_opts\\$gene\\_mode\n",
       "    \\}\n",
       "    else \\{\n",
       "        gene\\_mode <- FALSE\n",
       "    \\}\n",
       "    if (!is(sample\\_to\\_covariates, \"data.frame\")) \\{\n",
       "        stop(paste0(\"'\", substitute(sample\\_to\\_covariates), \"' (sample\\_to\\_covariates) must be a data.frame\"))\n",
       "    \\}\n",
       "    if (!is(full\\_model, \"formula\") \\&\\& !is(full\\_model, \"matrix\") \\&\\& \n",
       "        !is.null(full\\_model)) \\{\n",
       "        stop(paste0(\"'\", substitute(full\\_model), \"' (full\\_model) must be a formula or a matrix\"))\n",
       "    \\}\n",
       "    if (!(\"sample\" \\%in\\% colnames(sample\\_to\\_covariates))) \\{\n",
       "        stop(paste0(\"'\", substitute(sample\\_to\\_covariates), \"' (sample\\_to\\_covariates) must contain a column named 'sample'\"))\n",
       "    \\}\n",
       "    if (!(\"path\" \\%in\\% colnames(sample\\_to\\_covariates))) \\{\n",
       "        stop(paste0(\"'\", substitute(sample\\_to\\_covariates)), \"' (sample\\_to\\_covariates) must contain a column named 'path'\")\n",
       "    \\}\n",
       "    if (!is.null(target\\_mapping) \\&\\& !is(target\\_mapping, \"data.frame\")) \\{\n",
       "        stop(paste0(\"'\", substitute(target\\_mapping), \"' (target\\_mapping) must be a data.frame or NULL\"))\n",
       "    \\}\n",
       "    else if (is(target\\_mapping, \"data.frame\")) \\{\n",
       "        if (!(\"target\\_id\" \\%in\\% colnames(target\\_mapping))) \\{\n",
       "            stop(paste0(\"'\", substitute(target\\_mapping), \"' (target\\_mapping) must contain a column named 'target\\_id'\"))\n",
       "        \\}\n",
       "    \\}\n",
       "    if (!is.null(max\\_bootstrap) \\&\\& max\\_bootstrap <= 0) \\{\n",
       "        stop(\"max\\_bootstrap must be > 0\")\n",
       "    \\}\n",
       "    if (any(is.na(sample\\_to\\_covariates))) \\{\n",
       "        warning(\"Your 'sample\\_to\\_covariates' data.frame contains NA values. This will likely cause issues later.\")\n",
       "    \\}\n",
       "    if (is(full\\_model, \"matrix\") \\&\\& nrow(full\\_model) != nrow(sample\\_to\\_covariates)) \\{\n",
       "        stop(\"The design matrix number of rows are not equal to the number of rows in the sample\\_to\\_covariates argument.\")\n",
       "    \\}\n",
       "    if (!is(filter\\_fun, \"function\")) \\{\n",
       "        stop(\"filter\\_fun must be a function\")\n",
       "    \\}\n",
       "    if (!is.null(filter\\_target\\_id) \\& !is.character(filter\\_target\\_id)) \\{\n",
       "        stop(\"if filter\\_target\\_id is used, it must be a character vector\")\n",
       "    \\}\n",
       "    if (!is(norm\\_fun\\_counts, \"function\")) \\{\n",
       "        stop(\"norm\\_fun\\_counts must be a function\")\n",
       "    \\}\n",
       "    if (!is(norm\\_fun\\_tpm, \"function\")) \\{\n",
       "        stop(\"norm\\_fun\\_tpm must be a function\")\n",
       "    \\}\n",
       "    if (!is(transform\\_fun\\_counts, \"function\")) \\{\n",
       "        stop(\"transform\\_fun\\_counts must be a function\")\n",
       "    \\}\n",
       "    if (!is(transform\\_fun\\_tpm, \"function\")) \\{\n",
       "        stop(\"transform\\_fun\\_tpm must be a function\")\n",
       "    \\}\n",
       "    if (is.null(aggregation\\_column) \\&\\& gene\\_mode) \\{\n",
       "        stop(\"You set 'gene\\_mode' to TRUE, but did not provide an 'aggregation\\_column' \", \n",
       "            \"to aggregate by. Please provide an 'aggregation\\_column'.\")\n",
       "    \\}\n",
       "    else if (gene\\_mode) \\{\n",
       "        message(\"'gene\\_mode' is TRUE. Sleuth will do counts aggregation at the gene level \", \n",
       "            \"for downstream normalization, transformation, and modeling steps, as well as \", \n",
       "            \"for plotting and results.\")\n",
       "    \\}\n",
       "    if (!is.null(aggregation\\_column) \\&\\& is.null(target\\_mapping)) \\{\n",
       "        stop(paste(\"You provided an 'aggregation\\_column' to aggregate by,\", \n",
       "            \"but not a 'target\\_mapping'. Please provide a 'target\\_mapping'.\"))\n",
       "    \\}\n",
       "    pval\\_aggregate <- !is.null(aggregation\\_column) \\&\\& !gene\\_mode\n",
       "    num\\_cores <- check\\_num\\_cores(num\\_cores)\n",
       "    msg(\"reading in kallisto results\")\n",
       "    sample\\_to\\_covariates <- as.data.frame(sample\\_to\\_covariates)\n",
       "    sample\\_to\\_covariates\\$sample <- as.character(sample\\_to\\_covariates\\$sample)\n",
       "    if (nrow(sample\\_to\\_covariates) == 1 \\&\\& !is.null(full\\_model)) \\{\n",
       "        warning(\"There is only one sample present, but you also provided a model. \", \n",
       "            \"The model will be set to NULL to prevent downstream errors.\\textbackslash{}n\", \n",
       "            \"The sample can be viewed using sleuth\\_live after preparation, \", \n",
       "            \"but you need more than one sample to run the other aspects of Sleuth.\")\n",
       "        full\\_model <- NULL\n",
       "    \\}\n",
       "    kal\\_dirs <- sample\\_to\\_covariates\\$path\n",
       "    sample\\_to\\_covariates\\$path <- NULL\n",
       "    msg(\"dropping unused factor levels\")\n",
       "    sample\\_to\\_covariates <- droplevels(sample\\_to\\_covariates)\n",
       "    nsamp <- 0\n",
       "    kal\\_list <- lapply(seq\\_along(kal\\_dirs), function(i) \\{\n",
       "        nsamp <- dot(nsamp)\n",
       "        path <- kal\\_dirs{[}i{]}\n",
       "        suppressMessages(\\{\n",
       "            kal <- read\\_kallisto(path, read\\_bootstrap = FALSE, \n",
       "                max\\_bootstrap = max\\_bootstrap)\n",
       "        \\})\n",
       "        kal\\$abundance <- dplyr::mutate(kal\\$abundance, sample = sample\\_to\\_covariates\\$sample{[}i{]})\n",
       "        kal\n",
       "    \\})\n",
       "    msg(\"\")\n",
       "    check\\_result <- check\\_kal\\_pack(kal\\_list)\n",
       "    kal\\_versions <- check\\_result\\$versions\n",
       "    obs\\_raw <- dplyr::bind\\_rows(lapply(kal\\_list, function(k) k\\$abundance))\n",
       "    counts\\_test <- data.table::as.data.table(obs\\_raw)\n",
       "    counts\\_test <- counts\\_test{[}, .(total = sum(est\\_counts)), \n",
       "        by = \"sample\"{]}\n",
       "    if (any(counts\\_test\\$total == 0)) \\{\n",
       "        zero\\_names <- counts\\_test\\$sample{[}which(counts\\_test\\$total == \n",
       "            0){]}\n",
       "        formatted\\_names <- paste(zero\\_names, collapse = \", \")\n",
       "        warning(\"At least one sample have no reads aligned. \", \n",
       "            \"Here are the samples with zero counts:\\textbackslash{}n\", formatted\\_names)\n",
       "    \\}\n",
       "    design\\_matrix <- NULL\n",
       "    if (is(full\\_model, \"formula\")) \\{\n",
       "        design\\_matrix <- model.matrix(full\\_model, sample\\_to\\_covariates)\n",
       "    \\}\n",
       "    else if (is(full\\_model, \"matrix\")) \\{\n",
       "        if (is.null(colnames(full\\_model))) \\{\n",
       "            stop(\"If matrix is supplied, column names must also be supplied.\")\n",
       "        \\}\n",
       "        design\\_matrix <- full\\_model\n",
       "    \\}\n",
       "    if (!is.null(full\\_model)) \\{\n",
       "        rownames(design\\_matrix) <- sample\\_to\\_covariates\\$sample\n",
       "        M <- t(design\\_matrix) \\%*\\% design\\_matrix\n",
       "        det\\_mod <- determinant(M)\\$modulus\n",
       "        if (!is.finite(det\\_mod)) \\{\n",
       "            stop(\"The full model you provided seems to result in a singular design matrix. \", \n",
       "                \"This frequently happens when one of the covariates is a linear \", \n",
       "                \"combination of one or more other covariates (e.g. one covariate \", \n",
       "                \"yields identical groupings as another covariate). Check your \", \n",
       "                \"sample\\_to\\_covariates table and your full model.\")\n",
       "        \\}\n",
       "        rm(M, det\\_mod)\n",
       "    \\}\n",
       "    obs\\_raw <- dplyr::arrange(obs\\_raw, target\\_id, sample)\n",
       "    if (!is.null(target\\_mapping)) \\{\n",
       "        tmp\\_names <- data.frame(target\\_id = kal\\_list{[}{[}1{]}{]}\\$abundance\\$target\\_id, \n",
       "            stringsAsFactors = FALSE)\n",
       "        target\\_mapping <- check\\_target\\_mapping(tmp\\_names, target\\_mapping, \n",
       "            !is.null(aggregation\\_column))\n",
       "        rm(tmp\\_names)\n",
       "    \\}\n",
       "    ret <- list(kal = kal\\_list, kal\\_versions = kal\\_versions, \n",
       "        obs\\_raw = obs\\_raw, sample\\_to\\_covariates = sample\\_to\\_covariates, \n",
       "        bootstrap\\_summary = NA, full\\_formula = full\\_model, design\\_matrix = design\\_matrix, \n",
       "        target\\_mapping = target\\_mapping, gene\\_mode = gene\\_mode, \n",
       "        gene\\_column = aggregation\\_column, norm\\_fun\\_counts = norm\\_fun\\_counts, \n",
       "        norm\\_fun\\_tpm = norm\\_fun\\_tpm, transform\\_fun\\_counts = transform\\_fun\\_counts, \n",
       "        transform\\_fun\\_tpm = transform\\_fun\\_tpm, pval\\_aggregate = pval\\_aggregate)\n",
       "    if (normalize) \\{\n",
       "        msg(\"normalizing est\\_counts\")\n",
       "        est\\_counts\\_spread <- spread\\_abundance\\_by(obs\\_raw, \"est\\_counts\", \n",
       "            sample\\_to\\_covariates\\$sample)\n",
       "        if (!is.null(filter\\_target\\_id)) \\{\n",
       "            msg(\"A list of target IDs for filtering was found. Using this for filtering\")\n",
       "            target\\_ids <- rownames(est\\_counts\\_spread)\n",
       "            filter\\_bool <- target\\_ids \\%in\\% filter\\_target\\_id\n",
       "            names(filter\\_bool) <- target\\_ids\n",
       "        \\}\n",
       "        else \\{\n",
       "            filter\\_bool <- apply(est\\_counts\\_spread, 1, filter\\_fun)\n",
       "        \\}\n",
       "        filter\\_true <- filter\\_bool{[}filter\\_bool{]}\n",
       "        if (sum(filter\\_bool) == 0) \\{\n",
       "            stop(\"Zero targets passed the filter you used. Please double check the filter used.\")\n",
       "        \\}\n",
       "        msg(paste0(sum(filter\\_bool), \" targets passed the filter\"))\n",
       "        est\\_counts\\_sf <- norm\\_fun\\_counts(est\\_counts\\_spread{[}filter\\_bool, \n",
       "            , drop = FALSE{]})\n",
       "        filter\\_df <- adf(target\\_id = names(filter\\_true))\n",
       "        est\\_counts\\_norm <- as\\_df(t(t(est\\_counts\\_spread)/est\\_counts\\_sf))\n",
       "        est\\_counts\\_norm\\$target\\_id <- rownames(est\\_counts\\_norm)\n",
       "        est\\_counts\\_norm <- tidyr::gather(est\\_counts\\_norm, sample, \n",
       "            est\\_counts, -target\\_id)\n",
       "        obs\\_norm <- est\\_counts\\_norm\n",
       "        obs\\_norm\\$target\\_id <- as.character(obs\\_norm\\$target\\_id)\n",
       "        obs\\_norm\\$sample <- as.character(obs\\_norm\\$sample)\n",
       "        rm(est\\_counts\\_norm)\n",
       "        msg(\"normalizing tpm\")\n",
       "        tpm\\_spread <- spread\\_abundance\\_by(obs\\_raw, \"tpm\", sample\\_to\\_covariates\\$sample)\n",
       "        tpm\\_sf <- norm\\_fun\\_tpm(tpm\\_spread{[}filter\\_bool, , drop = FALSE{]})\n",
       "        tpm\\_norm <- as\\_df(t(t(tpm\\_spread)/tpm\\_sf))\n",
       "        tpm\\_norm\\$target\\_id <- rownames(tpm\\_norm)\n",
       "        tpm\\_norm <- tidyr::gather(tpm\\_norm, sample, tpm, -target\\_id)\n",
       "        tpm\\_norm\\$sample <- as.character(tpm\\_norm\\$sample)\n",
       "        msg(\"merging in metadata\")\n",
       "        obs\\_norm <- dplyr::arrange(obs\\_norm, target\\_id, sample)\n",
       "        tpm\\_norm <- dplyr::arrange(tpm\\_norm, target\\_id, sample)\n",
       "        stopifnot(all.equal(obs\\_raw\\$target\\_id, obs\\_norm\\$target\\_id) \\&\\& \n",
       "            all.equal(obs\\_raw\\$sample, obs\\_norm\\$sample))\n",
       "        suppressWarnings(\\{\n",
       "            if (!all.equal(dplyr::select(obs\\_norm, target\\_id, \n",
       "                sample), dplyr::select(tpm\\_norm, target\\_id, sample))) \\{\n",
       "                stop(\"Invalid column rows. In principle, can simply join. Please report error.\")\n",
       "            \\}\n",
       "            obs\\_norm <- dplyr::bind\\_cols(obs\\_norm, dplyr::select(tpm\\_norm, \n",
       "                tpm))\n",
       "        \\})\n",
       "        obs\\_norm <- dplyr::bind\\_cols(obs\\_norm, dplyr::select(obs\\_raw, \n",
       "            eff\\_len, len))\n",
       "        obs\\_norm <- as\\_df(obs\\_norm)\n",
       "        ret\\$obs\\_norm <- obs\\_norm\n",
       "        ret\\$est\\_counts\\_sf <- est\\_counts\\_sf\n",
       "        ret\\$filter\\_bool <- filter\\_bool\n",
       "        ret\\$filter\\_df <- filter\\_df\n",
       "        ret\\$obs\\_norm\\_filt <- dplyr::semi\\_join(obs\\_norm, filter\\_df, \n",
       "            by = \"target\\_id\")\n",
       "        ret\\$tpm\\_sf <- tpm\\_sf\n",
       "        path <- kal\\_dirs{[}1{]}\n",
       "        kal\\_path <- get\\_kallisto\\_path(path)\n",
       "        target\\_id <- as.character(rhdf5::h5read(kal\\_path\\$path, \n",
       "            \"aux/ids\"))\n",
       "        num\\_transcripts <- length(target\\_id)\n",
       "        ret\\$bs\\_quants <- list()\n",
       "        which\\_target\\_id <- ret\\$filter\\_df\\$target\\_id\n",
       "        if (ret\\$gene\\_mode) \\{\n",
       "            msg(paste0(\"aggregating by column: \", aggregation\\_column))\n",
       "            target\\_mapping <- data.table::data.table(target\\_mapping)\n",
       "            target\\_mapping{[}target\\_mapping{[}{[}aggregation\\_column{]}{]} == \n",
       "                \"\", aggregation\\_column{]} <- NA\n",
       "            agg\\_id <- unique(target\\_mapping{[}, aggregation\\_column, \n",
       "                with = FALSE{]})\n",
       "            agg\\_id <- agg\\_id{[}{[}1{]}{]}\n",
       "            agg\\_id <- agg\\_id{[}!is.na(agg\\_id){]}\n",
       "            mappings <- dplyr::select\\_(target\\_mapping, \"target\\_id\", \n",
       "                aggregation\\_column)\n",
       "            mappings <- data.table::as.data.table(mappings)\n",
       "            which\\_tms <- which(mappings\\$target\\_id \\%in\\% which\\_target\\_id)\n",
       "            which\\_agg\\_id <- unique(mappings{[}which\\_tms, aggregation\\_column, \n",
       "                with = FALSE{]})\n",
       "            which\\_agg\\_id <- which\\_agg\\_id{[}{[}1{]}{]}\n",
       "            which\\_agg\\_id <- which\\_agg\\_id{[}!is.na(which\\_agg\\_id){]}\n",
       "            filter\\_df <- adf(target\\_id = which\\_agg\\_id)\n",
       "            filter\\_bool <- agg\\_id \\%in\\% which\\_agg\\_id\n",
       "            msg(paste0(length(which\\_agg\\_id), \" genes passed the filter\"))\n",
       "            norm\\_by\\_length <- TRUE\n",
       "            tmp <- data.table::as.data.table(ret\\$obs\\_raw)\n",
       "            tmp <- merge(tmp, mappings, by = \"target\\_id\", all.x = TRUE)\n",
       "            scale\\_factor <- tmp{[}, `:=`(scale\\_factor, median(eff\\_len)), \n",
       "                by = list(sample, eval(parse(text = aggregation\\_column))){]}\n",
       "            obs\\_norm\\_gene <- reads\\_per\\_base\\_transform(ret\\$obs\\_norm, \n",
       "                scale\\_factor, aggregation\\_column, mappings, norm\\_by\\_length)\n",
       "            tmp <- data.table::as.data.table(tpm\\_norm)\n",
       "            tmp <- merge(tmp, mappings, by = \"target\\_id\", all.x = T)\n",
       "            if (any(is.na(tmp{[}{[}aggregation\\_column{]}{]}))) \\{\n",
       "                rows\\_to\\_remove <- is.na(tmp{[}{[}aggregation\\_column{]}{]})\n",
       "                num\\_missing <- length(unique(tmp{[}rows\\_to\\_remove, \n",
       "                  target\\_id{]}))\n",
       "                warning(num\\_missing, \" target\\_ids are missing annotations for the aggregation\\_column: \", \n",
       "                  aggregation\\_column, \".\\textbackslash{}nThese target\\_ids will be dropped from the gene-level analysis.\", \n",
       "                  \"\\textbackslash{}nIf you did not expect this, check your 'target\\_mapping' table for missing values.\")\n",
       "                tmp <- tmp{[}!rows\\_to\\_remove{]}\n",
       "            \\}\n",
       "            tpm\\_norm\\_gene <- tmp{[}, j = list(tpm = sum(tpm)), \n",
       "                by = list(sample, eval(parse(text = aggregation\\_column))){]}\n",
       "            data.table::setnames(tpm\\_norm\\_gene, \"parse\", \"target\\_id\")\n",
       "            tpm\\_norm\\_gene <- as\\_df(tpm\\_norm\\_gene)\n",
       "            obs\\_norm\\_gene <- dplyr::arrange(obs\\_norm\\_gene, target\\_id, \n",
       "                sample)\n",
       "            tpm\\_norm\\_gene <- dplyr::arrange(tpm\\_norm\\_gene, target\\_id, \n",
       "                sample)\n",
       "            stopifnot(all.equal(dplyr::select(obs\\_norm\\_gene, \n",
       "                target\\_id, sample), dplyr::select(tpm\\_norm\\_gene, \n",
       "                target\\_id, sample)))\n",
       "            suppressWarnings(\\{\n",
       "                if (!all.equal(dplyr::select(obs\\_norm\\_gene, target\\_id, \n",
       "                  sample), dplyr::select(tpm\\_norm\\_gene, target\\_id, \n",
       "                  sample), check.attributes = FALSE)) \\{\n",
       "                  stop(\"Invalid column rows. In principle, can simply join. Please report error.\")\n",
       "                \\}\n",
       "                obs\\_norm\\_gene <- dplyr::bind\\_cols(obs\\_norm\\_gene, \n",
       "                  dplyr::select(tpm\\_norm\\_gene, tpm))\n",
       "            \\})\n",
       "            ret\\$filter\\_df <- adf(target\\_id = which\\_agg\\_id)\n",
       "            ret\\$filter\\_bool <- agg\\_id \\%in\\% which\\_agg\\_id\n",
       "            ret\\$obs\\_norm <- obs\\_norm\\_gene\n",
       "            ret\\$obs\\_norm\\_filt <- dplyr::semi\\_join(obs\\_norm\\_gene, \n",
       "                filter\\_df, by = \"target\\_id\")\n",
       "            rm(obs\\_norm, tpm\\_norm, obs\\_norm\\_gene, tpm\\_norm\\_gene)\n",
       "            all\\_sample\\_bootstrap <- matrix(NA\\_real\\_, nrow = length(which\\_agg\\_id), \n",
       "                ncol = length(ret\\$kal))\n",
       "            which\\_ids <- which\\_agg\\_id\n",
       "        \\}\n",
       "        else \\{\n",
       "            all\\_sample\\_bootstrap <- matrix(NA\\_real\\_, nrow = length(which\\_target\\_id), \n",
       "                ncol = length(ret\\$kal))\n",
       "            which\\_ids <- which\\_target\\_id\n",
       "        \\}\n",
       "        msg(\"summarizing bootstraps\")\n",
       "        apply\\_function <- if (num\\_cores == 1) \\{\n",
       "            lapply\n",
       "        \\}\n",
       "        else \\{\n",
       "            function(x, y) parallel::mclapply(x, y, mc.cores = num\\_cores)\n",
       "        \\}\n",
       "        bs\\_results <- apply\\_function(seq\\_along(kal\\_dirs), function(i) \\{\n",
       "            samp\\_name <- sample\\_to\\_covariates\\$sample{[}i{]}\n",
       "            kal\\_path <- get\\_kallisto\\_path(kal\\_dirs{[}i{]})\n",
       "            process\\_bootstrap(i, samp\\_name, kal\\_path, num\\_transcripts, \n",
       "                est\\_counts\\_sf{[}{[}i{]}{]}, read\\_bootstrap\\_tpm, ret\\$gene\\_mode, \n",
       "                extra\\_bootstrap\\_summary, target\\_id, mappings, \n",
       "                which\\_ids, ret\\$gene\\_column, ret\\$transform\\_fun\\_counts, \n",
       "                ret\\$transform\\_fun\\_tpm, max\\_bootstrap)\n",
       "        \\})\n",
       "        error\\_status <- sapply(bs\\_results, function(x) is(x, \n",
       "            \"try-error\"))\n",
       "        if (any(error\\_status)) \\{\n",
       "            error\\_msgs <- sapply(which(error\\_status), function(i) \\{\n",
       "                bad\\_run <- bs\\_results{[}{[}i{]}{]}\n",
       "                samp\\_name <- sample\\_to\\_covariates\\$sample{[}i{]}\n",
       "                trace <- .traceback(bad\\_run)\n",
       "                paste0(\"Sample '\", samp\\_name, \"' had this error message: \", \n",
       "                  trace{[}1{]})\n",
       "            \\})\n",
       "            formatted\\_error <- paste(error\\_msgs, collapse = \"\")\n",
       "            message(formatted\\_error)\n",
       "            stop(\"At least one core from mclapply had an error. See the above error message(s) for more details.\")\n",
       "        \\}\n",
       "        indices <- sapply(bs\\_results, function(result) result\\$index)\n",
       "        stopifnot(identical(indices, order(indices)))\n",
       "        sigma\\_q\\_sq\\_tpm <- NULL\n",
       "        if (read\\_bootstrap\\_tpm) \\{\n",
       "            all\\_sample\\_tpm <- sapply(bs\\_results, function(result) result\\$bootstrap\\_tpm\\_result)\n",
       "            rownames(all\\_sample\\_tpm) <- which\\_ids\n",
       "            sigma\\_q\\_sq\\_tpm <- rowMeans(all\\_sample\\_tpm)\n",
       "            sigma\\_q\\_sq\\_tpm <- sigma\\_q\\_sq\\_tpm{[}order(names(sigma\\_q\\_sq\\_tpm)){]}\n",
       "            ret\\$bs\\_quants <- lapply(bs\\_results, function(result) result\\$bs\\_quants)\n",
       "            names(ret\\$bs\\_quants) <- sample\\_to\\_covariates\\$sample\n",
       "        \\}\n",
       "        else if (extra\\_bootstrap\\_summary) \\{\n",
       "            ret\\$bs\\_quants <- lapply(bs\\_results, function(result) result\\$bs\\_quants)\n",
       "            names(ret\\$bs\\_quants) <- sample\\_to\\_covariates\\$sample\n",
       "        \\}\n",
       "        all\\_sample\\_bootstrap <- sapply(bs\\_results, function(result) result\\$bootstrap\\_result)\n",
       "        rownames(all\\_sample\\_bootstrap) <- which\\_ids\n",
       "        msg(\"\")\n",
       "        sigma\\_q\\_sq <- rowMeans(all\\_sample\\_bootstrap)\n",
       "        names(sigma\\_q\\_sq) <- which\\_ids\n",
       "        if (ret\\$gene\\_mode) \\{\n",
       "            obs\\_counts <- obs\\_to\\_matrix(ret, \"scaled\\_reads\\_per\\_base\"){[}which\\_agg\\_id, \n",
       "                , drop = FALSE{]}\n",
       "            obs\\_tpm <- obs\\_to\\_matrix(ret, \"tpm\"){[}which\\_agg\\_id, \n",
       "                , drop = FALSE{]}\n",
       "        \\}\n",
       "        else \\{\n",
       "            obs\\_counts <- obs\\_to\\_matrix(ret, \"est\\_counts\"){[}which\\_target\\_id, \n",
       "                , drop = FALSE{]}\n",
       "            obs\\_tpm <- obs\\_to\\_matrix(ret, \"tpm\"){[}which\\_target\\_id, \n",
       "                , drop = FALSE{]}\n",
       "        \\}\n",
       "        sigma\\_q\\_sq <- sigma\\_q\\_sq{[}order(names(sigma\\_q\\_sq)){]}\n",
       "        obs\\_counts <- ret\\$transform\\_fun\\_counts(obs\\_counts)\n",
       "        obs\\_counts <- obs\\_counts{[}order(rownames(obs\\_counts)), \n",
       "            {]}\n",
       "        obs\\_tpm <- ret\\$transform\\_fun\\_tpm(obs\\_tpm)\n",
       "        obs\\_tpm <- obs\\_tpm{[}order(rownames(obs\\_tpm)), {]}\n",
       "        ret\\$bs\\_summary <- list(obs\\_counts = obs\\_counts, sigma\\_q\\_sq = sigma\\_q\\_sq)\n",
       "        ret\\$bs\\_summary <- list(obs\\_counts = obs\\_counts, obs\\_tpm = obs\\_tpm, \n",
       "            sigma\\_q\\_sq = sigma\\_q\\_sq, sigma\\_q\\_sq\\_tpm = sigma\\_q\\_sq\\_tpm)\n",
       "    \\}\n",
       "    else \\{\n",
       "        ret\\$norm\\_fun\\_counts <- ret\\$norm\\_fun\\_tpm <- NA\n",
       "        ret\\$transform\\_fun\\_counts <- ret\\$transform\\_fun\\_tpm <- NA\n",
       "        ret\\$obs\\_norm <- data.frame()\n",
       "        ret\\$est\\_counts\\_sf <- vector()\n",
       "        est\\_counts\\_spread <- spread\\_abundance\\_by(obs\\_raw, \"est\\_counts\", \n",
       "            sample\\_to\\_covariates\\$sample)\n",
       "        if (!is.null(filter\\_target\\_id)) \\{\n",
       "            msg(\"A list of target IDs for filtering was found. Using this for filtering\")\n",
       "            target\\_ids <- rownames(est\\_counts\\_spread)\n",
       "            filter\\_bool <- target\\_ids \\%in\\% filter\\_target\\_id\n",
       "            names(filter\\_bool) <- target\\_ids\n",
       "        \\}\n",
       "        else \\{\n",
       "            filter\\_bool <- apply(est\\_counts\\_spread, 1, filter\\_fun)\n",
       "        \\}\n",
       "        filter\\_true <- filter\\_bool{[}filter\\_bool{]}\n",
       "        if (sum(filter\\_bool) == 0) \\{\n",
       "            stop(\"Zero targets passed the filter you used. Please double check the filter used.\")\n",
       "        \\}\n",
       "        msg(paste0(sum(filter\\_bool), \" targets passed the filter\"))\n",
       "        filter\\_df <- adf(target\\_id = names(filter\\_true))\n",
       "        if (ret\\$gene\\_mode) \\{\n",
       "            msg(paste0(\"aggregating by column: \", aggregation\\_column))\n",
       "            which\\_target\\_id <- ret\\$filter\\_df\\$target\\_id\n",
       "            target\\_mapping <- data.table::data.table(target\\_mapping)\n",
       "            target\\_mapping{[}target\\_mapping{[}{[}aggregation\\_column{]}{]} == \n",
       "                \"\", aggregation\\_column{]} <- NA\n",
       "            agg\\_id <- unique(target\\_mapping{[}, aggregation\\_column, \n",
       "                with = FALSE{]})\n",
       "            agg\\_id <- agg\\_id{[}{[}1{]}{]}\n",
       "            agg\\_id <- agg\\_id{[}!is.na(agg\\_id){]}\n",
       "            mappings <- dplyr::select\\_(target\\_mapping, \"target\\_id\", \n",
       "                aggregation\\_column)\n",
       "            mappings <- data.table::as.data.table(mappings)\n",
       "            which\\_tms <- which(mappings\\$target\\_id \\%in\\% which\\_target\\_id)\n",
       "            which\\_agg\\_id <- unique(mappings{[}which\\_tms, aggregation\\_column, \n",
       "                with = FALSE{]})\n",
       "            which\\_agg\\_id <- which\\_agg\\_id{[}{[}1{]}{]}\n",
       "            which\\_agg\\_id <- which\\_agg\\_id{[}!is.na(which\\_agg\\_id){]}\n",
       "            filter\\_df <- adf(target\\_id = which\\_agg\\_id)\n",
       "            filter\\_bool <- agg\\_id \\%in\\% which\\_agg\\_id\n",
       "            msg(paste0(length(which\\_agg\\_id), \" genes passed the filter\"))\n",
       "        \\}\n",
       "        ret\\$filter\\_bool <- filter\\_bool\n",
       "        ret\\$filter\\_df <- filter\\_df\n",
       "        ret\\$obs\\_norm\\_filt <- data.frame()\n",
       "        ret\\$tpm\\_sf <- vector()\n",
       "        ret\\$bs\\_quants <- list()\n",
       "        ret\\$bs\\_summary <- list()\n",
       "    \\}\n",
       "    class(ret) <- \"sleuth\"\n",
       "    ret\n",
       "\\}\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (sample_to_covariates, full_model = NULL, target_mapping = NULL, \n",
       "    aggregation_column = NULL, num_cores = max(1L, parallel::detectCores() - \n",
       "        1L), ...) \n",
       "{\n",
       "    extra_opts <- list(...)\n",
       "    if (\"extra_bootstrap_summary\" %in% names(extra_opts)) {\n",
       "        extra_bootstrap_summary <- extra_opts$extra_bootstrap_summary\n",
       "    }\n",
       "    else {\n",
       "        extra_bootstrap_summary <- FALSE\n",
       "    }\n",
       "    if (\"read_bootstrap_tpm\" %in% names(extra_opts)) {\n",
       "        read_bootstrap_tpm <- extra_opts$read_bootstrap_tpm\n",
       "    }\n",
       "    else {\n",
       "        read_bootstrap_tpm <- FALSE\n",
       "    }\n",
       "    if (\"max_bootstrap\" %in% names(extra_opts)) {\n",
       "        max_bootstrap <- extra_opts$max_bootstrap\n",
       "    }\n",
       "    else {\n",
       "        max_bootstrap <- NULL\n",
       "    }\n",
       "    if (\"filter_fun\" %in% names(extra_opts)) {\n",
       "        filter_fun <- extra_opts$filter_fun\n",
       "    }\n",
       "    else {\n",
       "        filter_fun <- basic_filter\n",
       "    }\n",
       "    if (\"filter_target_id\" %in% names(extra_opts)) {\n",
       "        filter_target_id <- extra_opts$filter_target_id\n",
       "    }\n",
       "    else {\n",
       "        filter_target_id <- NULL\n",
       "    }\n",
       "    if (\"normalize\" %in% names(extra_opts)) {\n",
       "        normalize <- extra_opts$normalize\n",
       "    }\n",
       "    else {\n",
       "        normalize <- TRUE\n",
       "    }\n",
       "    if (\"norm_fun_counts\" %in% names(extra_opts)) {\n",
       "        norm_fun_counts <- extra_opts$norm_fun_counts\n",
       "    }\n",
       "    else {\n",
       "        norm_fun_counts <- norm_factors\n",
       "    }\n",
       "    if (\"norm_fun_tpm\" %in% names(extra_opts)) {\n",
       "        norm_fun_tpm <- extra_opts$norm_fun_tpm\n",
       "    }\n",
       "    else {\n",
       "        norm_fun_tpm <- norm_factors\n",
       "    }\n",
       "    if (\"transform_fun_counts\" %in% names(extra_opts)) {\n",
       "        transform_fun_counts <- extra_opts$transform_fun_counts\n",
       "    }\n",
       "    else {\n",
       "        transform_fun_counts <- log_transform\n",
       "    }\n",
       "    if (\"transform_fun_tpm\" %in% names(extra_opts)) {\n",
       "        transform_fun_tpm <- extra_opts$transform_fun_tpm\n",
       "    }\n",
       "    else {\n",
       "        transform_fun_tpm <- identity\n",
       "    }\n",
       "    if (\"gene_mode\" %in% names(extra_opts)) {\n",
       "        gene_mode <- extra_opts$gene_mode\n",
       "    }\n",
       "    else {\n",
       "        gene_mode <- FALSE\n",
       "    }\n",
       "    if (!is(sample_to_covariates, \"data.frame\")) {\n",
       "        stop(paste0(\"'\", substitute(sample_to_covariates), \"' (sample_to_covariates) must be a data.frame\"))\n",
       "    }\n",
       "    if (!is(full_model, \"formula\") && !is(full_model, \"matrix\") && \n",
       "        !is.null(full_model)) {\n",
       "        stop(paste0(\"'\", substitute(full_model), \"' (full_model) must be a formula or a matrix\"))\n",
       "    }\n",
       "    if (!(\"sample\" %in% colnames(sample_to_covariates))) {\n",
       "        stop(paste0(\"'\", substitute(sample_to_covariates), \"' (sample_to_covariates) must contain a column named 'sample'\"))\n",
       "    }\n",
       "    if (!(\"path\" %in% colnames(sample_to_covariates))) {\n",
       "        stop(paste0(\"'\", substitute(sample_to_covariates)), \"' (sample_to_covariates) must contain a column named 'path'\")\n",
       "    }\n",
       "    if (!is.null(target_mapping) && !is(target_mapping, \"data.frame\")) {\n",
       "        stop(paste0(\"'\", substitute(target_mapping), \"' (target_mapping) must be a data.frame or NULL\"))\n",
       "    }\n",
       "    else if (is(target_mapping, \"data.frame\")) {\n",
       "        if (!(\"target_id\" %in% colnames(target_mapping))) {\n",
       "            stop(paste0(\"'\", substitute(target_mapping), \"' (target_mapping) must contain a column named 'target_id'\"))\n",
       "        }\n",
       "    }\n",
       "    if (!is.null(max_bootstrap) && max_bootstrap <= 0) {\n",
       "        stop(\"max_bootstrap must be > 0\")\n",
       "    }\n",
       "    if (any(is.na(sample_to_covariates))) {\n",
       "        warning(\"Your 'sample_to_covariates' data.frame contains NA values. This will likely cause issues later.\")\n",
       "    }\n",
       "    if (is(full_model, \"matrix\") && nrow(full_model) != nrow(sample_to_covariates)) {\n",
       "        stop(\"The design matrix number of rows are not equal to the number of rows in the sample_to_covariates argument.\")\n",
       "    }\n",
       "    if (!is(filter_fun, \"function\")) {\n",
       "        stop(\"filter_fun must be a function\")\n",
       "    }\n",
       "    if (!is.null(filter_target_id) & !is.character(filter_target_id)) {\n",
       "        stop(\"if filter_target_id is used, it must be a character vector\")\n",
       "    }\n",
       "    if (!is(norm_fun_counts, \"function\")) {\n",
       "        stop(\"norm_fun_counts must be a function\")\n",
       "    }\n",
       "    if (!is(norm_fun_tpm, \"function\")) {\n",
       "        stop(\"norm_fun_tpm must be a function\")\n",
       "    }\n",
       "    if (!is(transform_fun_counts, \"function\")) {\n",
       "        stop(\"transform_fun_counts must be a function\")\n",
       "    }\n",
       "    if (!is(transform_fun_tpm, \"function\")) {\n",
       "        stop(\"transform_fun_tpm must be a function\")\n",
       "    }\n",
       "    if (is.null(aggregation_column) && gene_mode) {\n",
       "        stop(\"You set 'gene_mode' to TRUE, but did not provide an 'aggregation_column' \", \n",
       "            \"to aggregate by. Please provide an 'aggregation_column'.\")\n",
       "    }\n",
       "    else if (gene_mode) {\n",
       "        message(\"'gene_mode' is TRUE. Sleuth will do counts aggregation at the gene level \", \n",
       "            \"for downstream normalization, transformation, and modeling steps, as well as \", \n",
       "            \"for plotting and results.\")\n",
       "    }\n",
       "    if (!is.null(aggregation_column) && is.null(target_mapping)) {\n",
       "        stop(paste(\"You provided an 'aggregation_column' to aggregate by,\", \n",
       "            \"but not a 'target_mapping'. Please provide a 'target_mapping'.\"))\n",
       "    }\n",
       "    pval_aggregate <- !is.null(aggregation_column) && !gene_mode\n",
       "    num_cores <- check_num_cores(num_cores)\n",
       "    msg(\"reading in kallisto results\")\n",
       "    sample_to_covariates <- as.data.frame(sample_to_covariates)\n",
       "    sample_to_covariates$sample <- as.character(sample_to_covariates$sample)\n",
       "    if (nrow(sample_to_covariates) == 1 && !is.null(full_model)) {\n",
       "        warning(\"There is only one sample present, but you also provided a model. \", \n",
       "            \"The model will be set to NULL to prevent downstream errors.\\n\", \n",
       "            \"The sample can be viewed using sleuth_live after preparation, \", \n",
       "            \"but you need more than one sample to run the other aspects of Sleuth.\")\n",
       "        full_model <- NULL\n",
       "    }\n",
       "    kal_dirs <- sample_to_covariates$path\n",
       "    sample_to_covariates$path <- NULL\n",
       "    msg(\"dropping unused factor levels\")\n",
       "    sample_to_covariates <- droplevels(sample_to_covariates)\n",
       "    nsamp <- 0\n",
       "    kal_list <- lapply(seq_along(kal_dirs), function(i) {\n",
       "        nsamp <- dot(nsamp)\n",
       "        path <- kal_dirs[i]\n",
       "        suppressMessages({\n",
       "            kal <- read_kallisto(path, read_bootstrap = FALSE, \n",
       "                max_bootstrap = max_bootstrap)\n",
       "        })\n",
       "        kal$abundance <- dplyr::mutate(kal$abundance, sample = sample_to_covariates$sample[i])\n",
       "        kal\n",
       "    })\n",
       "    msg(\"\")\n",
       "    check_result <- check_kal_pack(kal_list)\n",
       "    kal_versions <- check_result$versions\n",
       "    obs_raw <- dplyr::bind_rows(lapply(kal_list, function(k) k$abundance))\n",
       "    counts_test <- data.table::as.data.table(obs_raw)\n",
       "    counts_test <- counts_test[, .(total = sum(est_counts)), \n",
       "        by = \"sample\"]\n",
       "    if (any(counts_test$total == 0)) {\n",
       "        zero_names <- counts_test$sample[which(counts_test$total == \n",
       "            0)]\n",
       "        formatted_names <- paste(zero_names, collapse = \", \")\n",
       "        warning(\"At least one sample have no reads aligned. \", \n",
       "            \"Here are the samples with zero counts:\\n\", formatted_names)\n",
       "    }\n",
       "    design_matrix <- NULL\n",
       "    if (is(full_model, \"formula\")) {\n",
       "        design_matrix <- model.matrix(full_model, sample_to_covariates)\n",
       "    }\n",
       "    else if (is(full_model, \"matrix\")) {\n",
       "        if (is.null(colnames(full_model))) {\n",
       "            stop(\"If matrix is supplied, column names must also be supplied.\")\n",
       "        }\n",
       "        design_matrix <- full_model\n",
       "    }\n",
       "    if (!is.null(full_model)) {\n",
       "        rownames(design_matrix) <- sample_to_covariates$sample\n",
       "        M <- t(design_matrix) %*% design_matrix\n",
       "        det_mod <- determinant(M)$modulus\n",
       "        if (!is.finite(det_mod)) {\n",
       "            stop(\"The full model you provided seems to result in a singular design matrix. \", \n",
       "                \"This frequently happens when one of the covariates is a linear \", \n",
       "                \"combination of one or more other covariates (e.g. one covariate \", \n",
       "                \"yields identical groupings as another covariate). Check your \", \n",
       "                \"sample_to_covariates table and your full model.\")\n",
       "        }\n",
       "        rm(M, det_mod)\n",
       "    }\n",
       "    obs_raw <- dplyr::arrange(obs_raw, target_id, sample)\n",
       "    if (!is.null(target_mapping)) {\n",
       "        tmp_names <- data.frame(target_id = kal_list[[1]]$abundance$target_id, \n",
       "            stringsAsFactors = FALSE)\n",
       "        target_mapping <- check_target_mapping(tmp_names, target_mapping, \n",
       "            !is.null(aggregation_column))\n",
       "        rm(tmp_names)\n",
       "    }\n",
       "    ret <- list(kal = kal_list, kal_versions = kal_versions, \n",
       "        obs_raw = obs_raw, sample_to_covariates = sample_to_covariates, \n",
       "        bootstrap_summary = NA, full_formula = full_model, design_matrix = design_matrix, \n",
       "        target_mapping = target_mapping, gene_mode = gene_mode, \n",
       "        gene_column = aggregation_column, norm_fun_counts = norm_fun_counts, \n",
       "        norm_fun_tpm = norm_fun_tpm, transform_fun_counts = transform_fun_counts, \n",
       "        transform_fun_tpm = transform_fun_tpm, pval_aggregate = pval_aggregate)\n",
       "    if (normalize) {\n",
       "        msg(\"normalizing est_counts\")\n",
       "        est_counts_spread <- spread_abundance_by(obs_raw, \"est_counts\", \n",
       "            sample_to_covariates$sample)\n",
       "        if (!is.null(filter_target_id)) {\n",
       "            msg(\"A list of target IDs for filtering was found. Using this for filtering\")\n",
       "            target_ids <- rownames(est_counts_spread)\n",
       "            filter_bool <- target_ids %in% filter_target_id\n",
       "            names(filter_bool) <- target_ids\n",
       "        }\n",
       "        else {\n",
       "            filter_bool <- apply(est_counts_spread, 1, filter_fun)\n",
       "        }\n",
       "        filter_true <- filter_bool[filter_bool]\n",
       "        if (sum(filter_bool) == 0) {\n",
       "            stop(\"Zero targets passed the filter you used. Please double check the filter used.\")\n",
       "        }\n",
       "        msg(paste0(sum(filter_bool), \" targets passed the filter\"))\n",
       "        est_counts_sf <- norm_fun_counts(est_counts_spread[filter_bool, \n",
       "            , drop = FALSE])\n",
       "        filter_df <- adf(target_id = names(filter_true))\n",
       "        est_counts_norm <- as_df(t(t(est_counts_spread)/est_counts_sf))\n",
       "        est_counts_norm$target_id <- rownames(est_counts_norm)\n",
       "        est_counts_norm <- tidyr::gather(est_counts_norm, sample, \n",
       "            est_counts, -target_id)\n",
       "        obs_norm <- est_counts_norm\n",
       "        obs_norm$target_id <- as.character(obs_norm$target_id)\n",
       "        obs_norm$sample <- as.character(obs_norm$sample)\n",
       "        rm(est_counts_norm)\n",
       "        msg(\"normalizing tpm\")\n",
       "        tpm_spread <- spread_abundance_by(obs_raw, \"tpm\", sample_to_covariates$sample)\n",
       "        tpm_sf <- norm_fun_tpm(tpm_spread[filter_bool, , drop = FALSE])\n",
       "        tpm_norm <- as_df(t(t(tpm_spread)/tpm_sf))\n",
       "        tpm_norm$target_id <- rownames(tpm_norm)\n",
       "        tpm_norm <- tidyr::gather(tpm_norm, sample, tpm, -target_id)\n",
       "        tpm_norm$sample <- as.character(tpm_norm$sample)\n",
       "        msg(\"merging in metadata\")\n",
       "        obs_norm <- dplyr::arrange(obs_norm, target_id, sample)\n",
       "        tpm_norm <- dplyr::arrange(tpm_norm, target_id, sample)\n",
       "        stopifnot(all.equal(obs_raw$target_id, obs_norm$target_id) && \n",
       "            all.equal(obs_raw$sample, obs_norm$sample))\n",
       "        suppressWarnings({\n",
       "            if (!all.equal(dplyr::select(obs_norm, target_id, \n",
       "                sample), dplyr::select(tpm_norm, target_id, sample))) {\n",
       "                stop(\"Invalid column rows. In principle, can simply join. Please report error.\")\n",
       "            }\n",
       "            obs_norm <- dplyr::bind_cols(obs_norm, dplyr::select(tpm_norm, \n",
       "                tpm))\n",
       "        })\n",
       "        obs_norm <- dplyr::bind_cols(obs_norm, dplyr::select(obs_raw, \n",
       "            eff_len, len))\n",
       "        obs_norm <- as_df(obs_norm)\n",
       "        ret$obs_norm <- obs_norm\n",
       "        ret$est_counts_sf <- est_counts_sf\n",
       "        ret$filter_bool <- filter_bool\n",
       "        ret$filter_df <- filter_df\n",
       "        ret$obs_norm_filt <- dplyr::semi_join(obs_norm, filter_df, \n",
       "            by = \"target_id\")\n",
       "        ret$tpm_sf <- tpm_sf\n",
       "        path <- kal_dirs[1]\n",
       "        kal_path <- get_kallisto_path(path)\n",
       "        target_id <- as.character(rhdf5::h5read(kal_path$path, \n",
       "            \"aux/ids\"))\n",
       "        num_transcripts <- length(target_id)\n",
       "        ret$bs_quants <- list()\n",
       "        which_target_id <- ret$filter_df$target_id\n",
       "        if (ret$gene_mode) {\n",
       "            msg(paste0(\"aggregating by column: \", aggregation_column))\n",
       "            target_mapping <- data.table::data.table(target_mapping)\n",
       "            target_mapping[target_mapping[[aggregation_column]] == \n",
       "                \"\", aggregation_column] <- NA\n",
       "            agg_id <- unique(target_mapping[, aggregation_column, \n",
       "                with = FALSE])\n",
       "            agg_id <- agg_id[[1]]\n",
       "            agg_id <- agg_id[!is.na(agg_id)]\n",
       "            mappings <- dplyr::select_(target_mapping, \"target_id\", \n",
       "                aggregation_column)\n",
       "            mappings <- data.table::as.data.table(mappings)\n",
       "            which_tms <- which(mappings$target_id %in% which_target_id)\n",
       "            which_agg_id <- unique(mappings[which_tms, aggregation_column, \n",
       "                with = FALSE])\n",
       "            which_agg_id <- which_agg_id[[1]]\n",
       "            which_agg_id <- which_agg_id[!is.na(which_agg_id)]\n",
       "            filter_df <- adf(target_id = which_agg_id)\n",
       "            filter_bool <- agg_id %in% which_agg_id\n",
       "            msg(paste0(length(which_agg_id), \" genes passed the filter\"))\n",
       "            norm_by_length <- TRUE\n",
       "            tmp <- data.table::as.data.table(ret$obs_raw)\n",
       "            tmp <- merge(tmp, mappings, by = \"target_id\", all.x = TRUE)\n",
       "            scale_factor <- tmp[, `:=`(scale_factor, median(eff_len)), \n",
       "                by = list(sample, eval(parse(text = aggregation_column)))]\n",
       "            obs_norm_gene <- reads_per_base_transform(ret$obs_norm, \n",
       "                scale_factor, aggregation_column, mappings, norm_by_length)\n",
       "            tmp <- data.table::as.data.table(tpm_norm)\n",
       "            tmp <- merge(tmp, mappings, by = \"target_id\", all.x = T)\n",
       "            if (any(is.na(tmp[[aggregation_column]]))) {\n",
       "                rows_to_remove <- is.na(tmp[[aggregation_column]])\n",
       "                num_missing <- length(unique(tmp[rows_to_remove, \n",
       "                  target_id]))\n",
       "                warning(num_missing, \" target_ids are missing annotations for the aggregation_column: \", \n",
       "                  aggregation_column, \".\\nThese target_ids will be dropped from the gene-level analysis.\", \n",
       "                  \"\\nIf you did not expect this, check your 'target_mapping' table for missing values.\")\n",
       "                tmp <- tmp[!rows_to_remove]\n",
       "            }\n",
       "            tpm_norm_gene <- tmp[, j = list(tpm = sum(tpm)), \n",
       "                by = list(sample, eval(parse(text = aggregation_column)))]\n",
       "            data.table::setnames(tpm_norm_gene, \"parse\", \"target_id\")\n",
       "            tpm_norm_gene <- as_df(tpm_norm_gene)\n",
       "            obs_norm_gene <- dplyr::arrange(obs_norm_gene, target_id, \n",
       "                sample)\n",
       "            tpm_norm_gene <- dplyr::arrange(tpm_norm_gene, target_id, \n",
       "                sample)\n",
       "            stopifnot(all.equal(dplyr::select(obs_norm_gene, \n",
       "                target_id, sample), dplyr::select(tpm_norm_gene, \n",
       "                target_id, sample)))\n",
       "            suppressWarnings({\n",
       "                if (!all.equal(dplyr::select(obs_norm_gene, target_id, \n",
       "                  sample), dplyr::select(tpm_norm_gene, target_id, \n",
       "                  sample), check.attributes = FALSE)) {\n",
       "                  stop(\"Invalid column rows. In principle, can simply join. Please report error.\")\n",
       "                }\n",
       "                obs_norm_gene <- dplyr::bind_cols(obs_norm_gene, \n",
       "                  dplyr::select(tpm_norm_gene, tpm))\n",
       "            })\n",
       "            ret$filter_df <- adf(target_id = which_agg_id)\n",
       "            ret$filter_bool <- agg_id %in% which_agg_id\n",
       "            ret$obs_norm <- obs_norm_gene\n",
       "            ret$obs_norm_filt <- dplyr::semi_join(obs_norm_gene, \n",
       "                filter_df, by = \"target_id\")\n",
       "            rm(obs_norm, tpm_norm, obs_norm_gene, tpm_norm_gene)\n",
       "            all_sample_bootstrap <- matrix(NA_real_, nrow = length(which_agg_id), \n",
       "                ncol = length(ret$kal))\n",
       "            which_ids <- which_agg_id\n",
       "        }\n",
       "        else {\n",
       "            all_sample_bootstrap <- matrix(NA_real_, nrow = length(which_target_id), \n",
       "                ncol = length(ret$kal))\n",
       "            which_ids <- which_target_id\n",
       "        }\n",
       "        msg(\"summarizing bootstraps\")\n",
       "        apply_function <- if (num_cores == 1) {\n",
       "            lapply\n",
       "        }\n",
       "        else {\n",
       "            function(x, y) parallel::mclapply(x, y, mc.cores = num_cores)\n",
       "        }\n",
       "        bs_results <- apply_function(seq_along(kal_dirs), function(i) {\n",
       "            samp_name <- sample_to_covariates$sample[i]\n",
       "            kal_path <- get_kallisto_path(kal_dirs[i])\n",
       "            process_bootstrap(i, samp_name, kal_path, num_transcripts, \n",
       "                est_counts_sf[[i]], read_bootstrap_tpm, ret$gene_mode, \n",
       "                extra_bootstrap_summary, target_id, mappings, \n",
       "                which_ids, ret$gene_column, ret$transform_fun_counts, \n",
       "                ret$transform_fun_tpm, max_bootstrap)\n",
       "        })\n",
       "        error_status <- sapply(bs_results, function(x) is(x, \n",
       "            \"try-error\"))\n",
       "        if (any(error_status)) {\n",
       "            error_msgs <- sapply(which(error_status), function(i) {\n",
       "                bad_run <- bs_results[[i]]\n",
       "                samp_name <- sample_to_covariates$sample[i]\n",
       "                trace <- .traceback(bad_run)\n",
       "                paste0(\"Sample '\", samp_name, \"' had this error message: \", \n",
       "                  trace[1])\n",
       "            })\n",
       "            formatted_error <- paste(error_msgs, collapse = \"\")\n",
       "            message(formatted_error)\n",
       "            stop(\"At least one core from mclapply had an error. See the above error message(s) for more details.\")\n",
       "        }\n",
       "        indices <- sapply(bs_results, function(result) result$index)\n",
       "        stopifnot(identical(indices, order(indices)))\n",
       "        sigma_q_sq_tpm <- NULL\n",
       "        if (read_bootstrap_tpm) {\n",
       "            all_sample_tpm <- sapply(bs_results, function(result) result$bootstrap_tpm_result)\n",
       "            rownames(all_sample_tpm) <- which_ids\n",
       "            sigma_q_sq_tpm <- rowMeans(all_sample_tpm)\n",
       "            sigma_q_sq_tpm <- sigma_q_sq_tpm[order(names(sigma_q_sq_tpm))]\n",
       "            ret$bs_quants <- lapply(bs_results, function(result) result$bs_quants)\n",
       "            names(ret$bs_quants) <- sample_to_covariates$sample\n",
       "        }\n",
       "        else if (extra_bootstrap_summary) {\n",
       "            ret$bs_quants <- lapply(bs_results, function(result) result$bs_quants)\n",
       "            names(ret$bs_quants) <- sample_to_covariates$sample\n",
       "        }\n",
       "        all_sample_bootstrap <- sapply(bs_results, function(result) result$bootstrap_result)\n",
       "        rownames(all_sample_bootstrap) <- which_ids\n",
       "        msg(\"\")\n",
       "        sigma_q_sq <- rowMeans(all_sample_bootstrap)\n",
       "        names(sigma_q_sq) <- which_ids\n",
       "        if (ret$gene_mode) {\n",
       "            obs_counts <- obs_to_matrix(ret, \"scaled_reads_per_base\")[which_agg_id, \n",
       "                , drop = FALSE]\n",
       "            obs_tpm <- obs_to_matrix(ret, \"tpm\")[which_agg_id, \n",
       "                , drop = FALSE]\n",
       "        }\n",
       "        else {\n",
       "            obs_counts <- obs_to_matrix(ret, \"est_counts\")[which_target_id, \n",
       "                , drop = FALSE]\n",
       "            obs_tpm <- obs_to_matrix(ret, \"tpm\")[which_target_id, \n",
       "                , drop = FALSE]\n",
       "        }\n",
       "        sigma_q_sq <- sigma_q_sq[order(names(sigma_q_sq))]\n",
       "        obs_counts <- ret$transform_fun_counts(obs_counts)\n",
       "        obs_counts <- obs_counts[order(rownames(obs_counts)), \n",
       "            ]\n",
       "        obs_tpm <- ret$transform_fun_tpm(obs_tpm)\n",
       "        obs_tpm <- obs_tpm[order(rownames(obs_tpm)), ]\n",
       "        ret$bs_summary <- list(obs_counts = obs_counts, sigma_q_sq = sigma_q_sq)\n",
       "        ret$bs_summary <- list(obs_counts = obs_counts, obs_tpm = obs_tpm, \n",
       "            sigma_q_sq = sigma_q_sq, sigma_q_sq_tpm = sigma_q_sq_tpm)\n",
       "    }\n",
       "    else {\n",
       "        ret$norm_fun_counts <- ret$norm_fun_tpm <- NA\n",
       "        ret$transform_fun_counts <- ret$transform_fun_tpm <- NA\n",
       "        ret$obs_norm <- data.frame()\n",
       "        ret$est_counts_sf <- vector()\n",
       "        est_counts_spread <- spread_abundance_by(obs_raw, \"est_counts\", \n",
       "            sample_to_covariates$sample)\n",
       "        if (!is.null(filter_target_id)) {\n",
       "            msg(\"A list of target IDs for filtering was found. Using this for filtering\")\n",
       "            target_ids <- rownames(est_counts_spread)\n",
       "            filter_bool <- target_ids %in% filter_target_id\n",
       "            names(filter_bool) <- target_ids\n",
       "        }\n",
       "        else {\n",
       "            filter_bool <- apply(est_counts_spread, 1, filter_fun)\n",
       "        }\n",
       "        filter_true <- filter_bool[filter_bool]\n",
       "        if (sum(filter_bool) == 0) {\n",
       "            stop(\"Zero targets passed the filter you used. Please double check the filter used.\")\n",
       "        }\n",
       "        msg(paste0(sum(filter_bool), \" targets passed the filter\"))\n",
       "        filter_df <- adf(target_id = names(filter_true))\n",
       "        if (ret$gene_mode) {\n",
       "            msg(paste0(\"aggregating by column: \", aggregation_column))\n",
       "            which_target_id <- ret$filter_df$target_id\n",
       "            target_mapping <- data.table::data.table(target_mapping)\n",
       "            target_mapping[target_mapping[[aggregation_column]] == \n",
       "                \"\", aggregation_column] <- NA\n",
       "            agg_id <- unique(target_mapping[, aggregation_column, \n",
       "                with = FALSE])\n",
       "            agg_id <- agg_id[[1]]\n",
       "            agg_id <- agg_id[!is.na(agg_id)]\n",
       "            mappings <- dplyr::select_(target_mapping, \"target_id\", \n",
       "                aggregation_column)\n",
       "            mappings <- data.table::as.data.table(mappings)\n",
       "            which_tms <- which(mappings$target_id %in% which_target_id)\n",
       "            which_agg_id <- unique(mappings[which_tms, aggregation_column, \n",
       "                with = FALSE])\n",
       "            which_agg_id <- which_agg_id[[1]]\n",
       "            which_agg_id <- which_agg_id[!is.na(which_agg_id)]\n",
       "            filter_df <- adf(target_id = which_agg_id)\n",
       "            filter_bool <- agg_id %in% which_agg_id\n",
       "            msg(paste0(length(which_agg_id), \" genes passed the filter\"))\n",
       "        }\n",
       "        ret$filter_bool <- filter_bool\n",
       "        ret$filter_df <- filter_df\n",
       "        ret$obs_norm_filt <- data.frame()\n",
       "        ret$tpm_sf <- vector()\n",
       "        ret$bs_quants <- list()\n",
       "        ret$bs_summary <- list()\n",
       "    }\n",
       "    class(ret) <- \"sleuth\"\n",
       "    ret\n",
       "}\n",
       "```"
      ],
      "text/plain": [
       "function (sample_to_covariates, full_model = NULL, target_mapping = NULL, \n",
       "    aggregation_column = NULL, num_cores = max(1L, parallel::detectCores() - \n",
       "        1L), ...) \n",
       "{\n",
       "    extra_opts <- list(...)\n",
       "    if (\"extra_bootstrap_summary\" %in% names(extra_opts)) {\n",
       "        extra_bootstrap_summary <- extra_opts$extra_bootstrap_summary\n",
       "    }\n",
       "    else {\n",
       "        extra_bootstrap_summary <- FALSE\n",
       "    }\n",
       "    if (\"read_bootstrap_tpm\" %in% names(extra_opts)) {\n",
       "        read_bootstrap_tpm <- extra_opts$read_bootstrap_tpm\n",
       "    }\n",
       "    else {\n",
       "        read_bootstrap_tpm <- FALSE\n",
       "    }\n",
       "    if (\"max_bootstrap\" %in% names(extra_opts)) {\n",
       "        max_bootstrap <- extra_opts$max_bootstrap\n",
       "    }\n",
       "    else {\n",
       "        max_bootstrap <- NULL\n",
       "    }\n",
       "    if (\"filter_fun\" %in% names(extra_opts)) {\n",
       "        filter_fun <- extra_opts$filter_fun\n",
       "    }\n",
       "    else {\n",
       "        filter_fun <- basic_filter\n",
       "    }\n",
       "    if (\"filter_target_id\" %in% names(extra_opts)) {\n",
       "        filter_target_id <- extra_opts$filter_target_id\n",
       "    }\n",
       "    else {\n",
       "        filter_target_id <- NULL\n",
       "    }\n",
       "    if (\"normalize\" %in% names(extra_opts)) {\n",
       "        normalize <- extra_opts$normalize\n",
       "    }\n",
       "    else {\n",
       "        normalize <- TRUE\n",
       "    }\n",
       "    if (\"norm_fun_counts\" %in% names(extra_opts)) {\n",
       "        norm_fun_counts <- extra_opts$norm_fun_counts\n",
       "    }\n",
       "    else {\n",
       "        norm_fun_counts <- norm_factors\n",
       "    }\n",
       "    if (\"norm_fun_tpm\" %in% names(extra_opts)) {\n",
       "        norm_fun_tpm <- extra_opts$norm_fun_tpm\n",
       "    }\n",
       "    else {\n",
       "        norm_fun_tpm <- norm_factors\n",
       "    }\n",
       "    if (\"transform_fun_counts\" %in% names(extra_opts)) {\n",
       "        transform_fun_counts <- extra_opts$transform_fun_counts\n",
       "    }\n",
       "    else {\n",
       "        transform_fun_counts <- log_transform\n",
       "    }\n",
       "    if (\"transform_fun_tpm\" %in% names(extra_opts)) {\n",
       "        transform_fun_tpm <- extra_opts$transform_fun_tpm\n",
       "    }\n",
       "    else {\n",
       "        transform_fun_tpm <- identity\n",
       "    }\n",
       "    if (\"gene_mode\" %in% names(extra_opts)) {\n",
       "        gene_mode <- extra_opts$gene_mode\n",
       "    }\n",
       "    else {\n",
       "        gene_mode <- FALSE\n",
       "    }\n",
       "    if (!is(sample_to_covariates, \"data.frame\")) {\n",
       "        stop(paste0(\"'\", substitute(sample_to_covariates), \"' (sample_to_covariates) must be a data.frame\"))\n",
       "    }\n",
       "    if (!is(full_model, \"formula\") && !is(full_model, \"matrix\") && \n",
       "        !is.null(full_model)) {\n",
       "        stop(paste0(\"'\", substitute(full_model), \"' (full_model) must be a formula or a matrix\"))\n",
       "    }\n",
       "    if (!(\"sample\" %in% colnames(sample_to_covariates))) {\n",
       "        stop(paste0(\"'\", substitute(sample_to_covariates), \"' (sample_to_covariates) must contain a column named 'sample'\"))\n",
       "    }\n",
       "    if (!(\"path\" %in% colnames(sample_to_covariates))) {\n",
       "        stop(paste0(\"'\", substitute(sample_to_covariates)), \"' (sample_to_covariates) must contain a column named 'path'\")\n",
       "    }\n",
       "    if (!is.null(target_mapping) && !is(target_mapping, \"data.frame\")) {\n",
       "        stop(paste0(\"'\", substitute(target_mapping), \"' (target_mapping) must be a data.frame or NULL\"))\n",
       "    }\n",
       "    else if (is(target_mapping, \"data.frame\")) {\n",
       "        if (!(\"target_id\" %in% colnames(target_mapping))) {\n",
       "            stop(paste0(\"'\", substitute(target_mapping), \"' (target_mapping) must contain a column named 'target_id'\"))\n",
       "        }\n",
       "    }\n",
       "    if (!is.null(max_bootstrap) && max_bootstrap <= 0) {\n",
       "        stop(\"max_bootstrap must be > 0\")\n",
       "    }\n",
       "    if (any(is.na(sample_to_covariates))) {\n",
       "        warning(\"Your 'sample_to_covariates' data.frame contains NA values. This will likely cause issues later.\")\n",
       "    }\n",
       "    if (is(full_model, \"matrix\") && nrow(full_model) != nrow(sample_to_covariates)) {\n",
       "        stop(\"The design matrix number of rows are not equal to the number of rows in the sample_to_covariates argument.\")\n",
       "    }\n",
       "    if (!is(filter_fun, \"function\")) {\n",
       "        stop(\"filter_fun must be a function\")\n",
       "    }\n",
       "    if (!is.null(filter_target_id) & !is.character(filter_target_id)) {\n",
       "        stop(\"if filter_target_id is used, it must be a character vector\")\n",
       "    }\n",
       "    if (!is(norm_fun_counts, \"function\")) {\n",
       "        stop(\"norm_fun_counts must be a function\")\n",
       "    }\n",
       "    if (!is(norm_fun_tpm, \"function\")) {\n",
       "        stop(\"norm_fun_tpm must be a function\")\n",
       "    }\n",
       "    if (!is(transform_fun_counts, \"function\")) {\n",
       "        stop(\"transform_fun_counts must be a function\")\n",
       "    }\n",
       "    if (!is(transform_fun_tpm, \"function\")) {\n",
       "        stop(\"transform_fun_tpm must be a function\")\n",
       "    }\n",
       "    if (is.null(aggregation_column) && gene_mode) {\n",
       "        stop(\"You set 'gene_mode' to TRUE, but did not provide an 'aggregation_column' \", \n",
       "            \"to aggregate by. Please provide an 'aggregation_column'.\")\n",
       "    }\n",
       "    else if (gene_mode) {\n",
       "        message(\"'gene_mode' is TRUE. Sleuth will do counts aggregation at the gene level \", \n",
       "            \"for downstream normalization, transformation, and modeling steps, as well as \", \n",
       "            \"for plotting and results.\")\n",
       "    }\n",
       "    if (!is.null(aggregation_column) && is.null(target_mapping)) {\n",
       "        stop(paste(\"You provided an 'aggregation_column' to aggregate by,\", \n",
       "            \"but not a 'target_mapping'. Please provide a 'target_mapping'.\"))\n",
       "    }\n",
       "    pval_aggregate <- !is.null(aggregation_column) && !gene_mode\n",
       "    num_cores <- check_num_cores(num_cores)\n",
       "    msg(\"reading in kallisto results\")\n",
       "    sample_to_covariates <- as.data.frame(sample_to_covariates)\n",
       "    sample_to_covariates$sample <- as.character(sample_to_covariates$sample)\n",
       "    if (nrow(sample_to_covariates) == 1 && !is.null(full_model)) {\n",
       "        warning(\"There is only one sample present, but you also provided a model. \", \n",
       "            \"The model will be set to NULL to prevent downstream errors.\\n\", \n",
       "            \"The sample can be viewed using sleuth_live after preparation, \", \n",
       "            \"but you need more than one sample to run the other aspects of Sleuth.\")\n",
       "        full_model <- NULL\n",
       "    }\n",
       "    kal_dirs <- sample_to_covariates$path\n",
       "    sample_to_covariates$path <- NULL\n",
       "    msg(\"dropping unused factor levels\")\n",
       "    sample_to_covariates <- droplevels(sample_to_covariates)\n",
       "    nsamp <- 0\n",
       "    kal_list <- lapply(seq_along(kal_dirs), function(i) {\n",
       "        nsamp <- dot(nsamp)\n",
       "        path <- kal_dirs[i]\n",
       "        suppressMessages({\n",
       "            kal <- read_kallisto(path, read_bootstrap = FALSE, \n",
       "                max_bootstrap = max_bootstrap)\n",
       "        })\n",
       "        kal$abundance <- dplyr::mutate(kal$abundance, sample = sample_to_covariates$sample[i])\n",
       "        kal\n",
       "    })\n",
       "    msg(\"\")\n",
       "    check_result <- check_kal_pack(kal_list)\n",
       "    kal_versions <- check_result$versions\n",
       "    obs_raw <- dplyr::bind_rows(lapply(kal_list, function(k) k$abundance))\n",
       "    counts_test <- data.table::as.data.table(obs_raw)\n",
       "    counts_test <- counts_test[, .(total = sum(est_counts)), \n",
       "        by = \"sample\"]\n",
       "    if (any(counts_test$total == 0)) {\n",
       "        zero_names <- counts_test$sample[which(counts_test$total == \n",
       "            0)]\n",
       "        formatted_names <- paste(zero_names, collapse = \", \")\n",
       "        warning(\"At least one sample have no reads aligned. \", \n",
       "            \"Here are the samples with zero counts:\\n\", formatted_names)\n",
       "    }\n",
       "    design_matrix <- NULL\n",
       "    if (is(full_model, \"formula\")) {\n",
       "        design_matrix <- model.matrix(full_model, sample_to_covariates)\n",
       "    }\n",
       "    else if (is(full_model, \"matrix\")) {\n",
       "        if (is.null(colnames(full_model))) {\n",
       "            stop(\"If matrix is supplied, column names must also be supplied.\")\n",
       "        }\n",
       "        design_matrix <- full_model\n",
       "    }\n",
       "    if (!is.null(full_model)) {\n",
       "        rownames(design_matrix) <- sample_to_covariates$sample\n",
       "        M <- t(design_matrix) %*% design_matrix\n",
       "        det_mod <- determinant(M)$modulus\n",
       "        if (!is.finite(det_mod)) {\n",
       "            stop(\"The full model you provided seems to result in a singular design matrix. \", \n",
       "                \"This frequently happens when one of the covariates is a linear \", \n",
       "                \"combination of one or more other covariates (e.g. one covariate \", \n",
       "                \"yields identical groupings as another covariate). Check your \", \n",
       "                \"sample_to_covariates table and your full model.\")\n",
       "        }\n",
       "        rm(M, det_mod)\n",
       "    }\n",
       "    obs_raw <- dplyr::arrange(obs_raw, target_id, sample)\n",
       "    if (!is.null(target_mapping)) {\n",
       "        tmp_names <- data.frame(target_id = kal_list[[1]]$abundance$target_id, \n",
       "            stringsAsFactors = FALSE)\n",
       "        target_mapping <- check_target_mapping(tmp_names, target_mapping, \n",
       "            !is.null(aggregation_column))\n",
       "        rm(tmp_names)\n",
       "    }\n",
       "    ret <- list(kal = kal_list, kal_versions = kal_versions, \n",
       "        obs_raw = obs_raw, sample_to_covariates = sample_to_covariates, \n",
       "        bootstrap_summary = NA, full_formula = full_model, design_matrix = design_matrix, \n",
       "        target_mapping = target_mapping, gene_mode = gene_mode, \n",
       "        gene_column = aggregation_column, norm_fun_counts = norm_fun_counts, \n",
       "        norm_fun_tpm = norm_fun_tpm, transform_fun_counts = transform_fun_counts, \n",
       "        transform_fun_tpm = transform_fun_tpm, pval_aggregate = pval_aggregate)\n",
       "    if (normalize) {\n",
       "        msg(\"normalizing est_counts\")\n",
       "        est_counts_spread <- spread_abundance_by(obs_raw, \"est_counts\", \n",
       "            sample_to_covariates$sample)\n",
       "        if (!is.null(filter_target_id)) {\n",
       "            msg(\"A list of target IDs for filtering was found. Using this for filtering\")\n",
       "            target_ids <- rownames(est_counts_spread)\n",
       "            filter_bool <- target_ids %in% filter_target_id\n",
       "            names(filter_bool) <- target_ids\n",
       "        }\n",
       "        else {\n",
       "            filter_bool <- apply(est_counts_spread, 1, filter_fun)\n",
       "        }\n",
       "        filter_true <- filter_bool[filter_bool]\n",
       "        if (sum(filter_bool) == 0) {\n",
       "            stop(\"Zero targets passed the filter you used. Please double check the filter used.\")\n",
       "        }\n",
       "        msg(paste0(sum(filter_bool), \" targets passed the filter\"))\n",
       "        est_counts_sf <- norm_fun_counts(est_counts_spread[filter_bool, \n",
       "            , drop = FALSE])\n",
       "        filter_df <- adf(target_id = names(filter_true))\n",
       "        est_counts_norm <- as_df(t(t(est_counts_spread)/est_counts_sf))\n",
       "        est_counts_norm$target_id <- rownames(est_counts_norm)\n",
       "        est_counts_norm <- tidyr::gather(est_counts_norm, sample, \n",
       "            est_counts, -target_id)\n",
       "        obs_norm <- est_counts_norm\n",
       "        obs_norm$target_id <- as.character(obs_norm$target_id)\n",
       "        obs_norm$sample <- as.character(obs_norm$sample)\n",
       "        rm(est_counts_norm)\n",
       "        msg(\"normalizing tpm\")\n",
       "        tpm_spread <- spread_abundance_by(obs_raw, \"tpm\", sample_to_covariates$sample)\n",
       "        tpm_sf <- norm_fun_tpm(tpm_spread[filter_bool, , drop = FALSE])\n",
       "        tpm_norm <- as_df(t(t(tpm_spread)/tpm_sf))\n",
       "        tpm_norm$target_id <- rownames(tpm_norm)\n",
       "        tpm_norm <- tidyr::gather(tpm_norm, sample, tpm, -target_id)\n",
       "        tpm_norm$sample <- as.character(tpm_norm$sample)\n",
       "        msg(\"merging in metadata\")\n",
       "        obs_norm <- dplyr::arrange(obs_norm, target_id, sample)\n",
       "        tpm_norm <- dplyr::arrange(tpm_norm, target_id, sample)\n",
       "        stopifnot(all.equal(obs_raw$target_id, obs_norm$target_id) && \n",
       "            all.equal(obs_raw$sample, obs_norm$sample))\n",
       "        suppressWarnings({\n",
       "            if (!all.equal(dplyr::select(obs_norm, target_id, \n",
       "                sample), dplyr::select(tpm_norm, target_id, sample))) {\n",
       "                stop(\"Invalid column rows. In principle, can simply join. Please report error.\")\n",
       "            }\n",
       "            obs_norm <- dplyr::bind_cols(obs_norm, dplyr::select(tpm_norm, \n",
       "                tpm))\n",
       "        })\n",
       "        obs_norm <- dplyr::bind_cols(obs_norm, dplyr::select(obs_raw, \n",
       "            eff_len, len))\n",
       "        obs_norm <- as_df(obs_norm)\n",
       "        ret$obs_norm <- obs_norm\n",
       "        ret$est_counts_sf <- est_counts_sf\n",
       "        ret$filter_bool <- filter_bool\n",
       "        ret$filter_df <- filter_df\n",
       "        ret$obs_norm_filt <- dplyr::semi_join(obs_norm, filter_df, \n",
       "            by = \"target_id\")\n",
       "        ret$tpm_sf <- tpm_sf\n",
       "        path <- kal_dirs[1]\n",
       "        kal_path <- get_kallisto_path(path)\n",
       "        target_id <- as.character(rhdf5::h5read(kal_path$path, \n",
       "            \"aux/ids\"))\n",
       "        num_transcripts <- length(target_id)\n",
       "        ret$bs_quants <- list()\n",
       "        which_target_id <- ret$filter_df$target_id\n",
       "        if (ret$gene_mode) {\n",
       "            msg(paste0(\"aggregating by column: \", aggregation_column))\n",
       "            target_mapping <- data.table::data.table(target_mapping)\n",
       "            target_mapping[target_mapping[[aggregation_column]] == \n",
       "                \"\", aggregation_column] <- NA\n",
       "            agg_id <- unique(target_mapping[, aggregation_column, \n",
       "                with = FALSE])\n",
       "            agg_id <- agg_id[[1]]\n",
       "            agg_id <- agg_id[!is.na(agg_id)]\n",
       "            mappings <- dplyr::select_(target_mapping, \"target_id\", \n",
       "                aggregation_column)\n",
       "            mappings <- data.table::as.data.table(mappings)\n",
       "            which_tms <- which(mappings$target_id %in% which_target_id)\n",
       "            which_agg_id <- unique(mappings[which_tms, aggregation_column, \n",
       "                with = FALSE])\n",
       "            which_agg_id <- which_agg_id[[1]]\n",
       "            which_agg_id <- which_agg_id[!is.na(which_agg_id)]\n",
       "            filter_df <- adf(target_id = which_agg_id)\n",
       "            filter_bool <- agg_id %in% which_agg_id\n",
       "            msg(paste0(length(which_agg_id), \" genes passed the filter\"))\n",
       "            norm_by_length <- TRUE\n",
       "            tmp <- data.table::as.data.table(ret$obs_raw)\n",
       "            tmp <- merge(tmp, mappings, by = \"target_id\", all.x = TRUE)\n",
       "            scale_factor <- tmp[, `:=`(scale_factor, median(eff_len)), \n",
       "                by = list(sample, eval(parse(text = aggregation_column)))]\n",
       "            obs_norm_gene <- reads_per_base_transform(ret$obs_norm, \n",
       "                scale_factor, aggregation_column, mappings, norm_by_length)\n",
       "            tmp <- data.table::as.data.table(tpm_norm)\n",
       "            tmp <- merge(tmp, mappings, by = \"target_id\", all.x = T)\n",
       "            if (any(is.na(tmp[[aggregation_column]]))) {\n",
       "                rows_to_remove <- is.na(tmp[[aggregation_column]])\n",
       "                num_missing <- length(unique(tmp[rows_to_remove, \n",
       "                  target_id]))\n",
       "                warning(num_missing, \" target_ids are missing annotations for the aggregation_column: \", \n",
       "                  aggregation_column, \".\\nThese target_ids will be dropped from the gene-level analysis.\", \n",
       "                  \"\\nIf you did not expect this, check your 'target_mapping' table for missing values.\")\n",
       "                tmp <- tmp[!rows_to_remove]\n",
       "            }\n",
       "            tpm_norm_gene <- tmp[, j = list(tpm = sum(tpm)), \n",
       "                by = list(sample, eval(parse(text = aggregation_column)))]\n",
       "            data.table::setnames(tpm_norm_gene, \"parse\", \"target_id\")\n",
       "            tpm_norm_gene <- as_df(tpm_norm_gene)\n",
       "            obs_norm_gene <- dplyr::arrange(obs_norm_gene, target_id, \n",
       "                sample)\n",
       "            tpm_norm_gene <- dplyr::arrange(tpm_norm_gene, target_id, \n",
       "                sample)\n",
       "            stopifnot(all.equal(dplyr::select(obs_norm_gene, \n",
       "                target_id, sample), dplyr::select(tpm_norm_gene, \n",
       "                target_id, sample)))\n",
       "            suppressWarnings({\n",
       "                if (!all.equal(dplyr::select(obs_norm_gene, target_id, \n",
       "                  sample), dplyr::select(tpm_norm_gene, target_id, \n",
       "                  sample), check.attributes = FALSE)) {\n",
       "                  stop(\"Invalid column rows. In principle, can simply join. Please report error.\")\n",
       "                }\n",
       "                obs_norm_gene <- dplyr::bind_cols(obs_norm_gene, \n",
       "                  dplyr::select(tpm_norm_gene, tpm))\n",
       "            })\n",
       "            ret$filter_df <- adf(target_id = which_agg_id)\n",
       "            ret$filter_bool <- agg_id %in% which_agg_id\n",
       "            ret$obs_norm <- obs_norm_gene\n",
       "            ret$obs_norm_filt <- dplyr::semi_join(obs_norm_gene, \n",
       "                filter_df, by = \"target_id\")\n",
       "            rm(obs_norm, tpm_norm, obs_norm_gene, tpm_norm_gene)\n",
       "            all_sample_bootstrap <- matrix(NA_real_, nrow = length(which_agg_id), \n",
       "                ncol = length(ret$kal))\n",
       "            which_ids <- which_agg_id\n",
       "        }\n",
       "        else {\n",
       "            all_sample_bootstrap <- matrix(NA_real_, nrow = length(which_target_id), \n",
       "                ncol = length(ret$kal))\n",
       "            which_ids <- which_target_id\n",
       "        }\n",
       "        msg(\"summarizing bootstraps\")\n",
       "        apply_function <- if (num_cores == 1) {\n",
       "            lapply\n",
       "        }\n",
       "        else {\n",
       "            function(x, y) parallel::mclapply(x, y, mc.cores = num_cores)\n",
       "        }\n",
       "        bs_results <- apply_function(seq_along(kal_dirs), function(i) {\n",
       "            samp_name <- sample_to_covariates$sample[i]\n",
       "            kal_path <- get_kallisto_path(kal_dirs[i])\n",
       "            process_bootstrap(i, samp_name, kal_path, num_transcripts, \n",
       "                est_counts_sf[[i]], read_bootstrap_tpm, ret$gene_mode, \n",
       "                extra_bootstrap_summary, target_id, mappings, \n",
       "                which_ids, ret$gene_column, ret$transform_fun_counts, \n",
       "                ret$transform_fun_tpm, max_bootstrap)\n",
       "        })\n",
       "        error_status <- sapply(bs_results, function(x) is(x, \n",
       "            \"try-error\"))\n",
       "        if (any(error_status)) {\n",
       "            error_msgs <- sapply(which(error_status), function(i) {\n",
       "                bad_run <- bs_results[[i]]\n",
       "                samp_name <- sample_to_covariates$sample[i]\n",
       "                trace <- .traceback(bad_run)\n",
       "                paste0(\"Sample '\", samp_name, \"' had this error message: \", \n",
       "                  trace[1])\n",
       "            })\n",
       "            formatted_error <- paste(error_msgs, collapse = \"\")\n",
       "            message(formatted_error)\n",
       "            stop(\"At least one core from mclapply had an error. See the above error message(s) for more details.\")\n",
       "        }\n",
       "        indices <- sapply(bs_results, function(result) result$index)\n",
       "        stopifnot(identical(indices, order(indices)))\n",
       "        sigma_q_sq_tpm <- NULL\n",
       "        if (read_bootstrap_tpm) {\n",
       "            all_sample_tpm <- sapply(bs_results, function(result) result$bootstrap_tpm_result)\n",
       "            rownames(all_sample_tpm) <- which_ids\n",
       "            sigma_q_sq_tpm <- rowMeans(all_sample_tpm)\n",
       "            sigma_q_sq_tpm <- sigma_q_sq_tpm[order(names(sigma_q_sq_tpm))]\n",
       "            ret$bs_quants <- lapply(bs_results, function(result) result$bs_quants)\n",
       "            names(ret$bs_quants) <- sample_to_covariates$sample\n",
       "        }\n",
       "        else if (extra_bootstrap_summary) {\n",
       "            ret$bs_quants <- lapply(bs_results, function(result) result$bs_quants)\n",
       "            names(ret$bs_quants) <- sample_to_covariates$sample\n",
       "        }\n",
       "        all_sample_bootstrap <- sapply(bs_results, function(result) result$bootstrap_result)\n",
       "        rownames(all_sample_bootstrap) <- which_ids\n",
       "        msg(\"\")\n",
       "        sigma_q_sq <- rowMeans(all_sample_bootstrap)\n",
       "        names(sigma_q_sq) <- which_ids\n",
       "        if (ret$gene_mode) {\n",
       "            obs_counts <- obs_to_matrix(ret, \"scaled_reads_per_base\")[which_agg_id, \n",
       "                , drop = FALSE]\n",
       "            obs_tpm <- obs_to_matrix(ret, \"tpm\")[which_agg_id, \n",
       "                , drop = FALSE]\n",
       "        }\n",
       "        else {\n",
       "            obs_counts <- obs_to_matrix(ret, \"est_counts\")[which_target_id, \n",
       "                , drop = FALSE]\n",
       "            obs_tpm <- obs_to_matrix(ret, \"tpm\")[which_target_id, \n",
       "                , drop = FALSE]\n",
       "        }\n",
       "        sigma_q_sq <- sigma_q_sq[order(names(sigma_q_sq))]\n",
       "        obs_counts <- ret$transform_fun_counts(obs_counts)\n",
       "        obs_counts <- obs_counts[order(rownames(obs_counts)), \n",
       "            ]\n",
       "        obs_tpm <- ret$transform_fun_tpm(obs_tpm)\n",
       "        obs_tpm <- obs_tpm[order(rownames(obs_tpm)), ]\n",
       "        ret$bs_summary <- list(obs_counts = obs_counts, sigma_q_sq = sigma_q_sq)\n",
       "        ret$bs_summary <- list(obs_counts = obs_counts, obs_tpm = obs_tpm, \n",
       "            sigma_q_sq = sigma_q_sq, sigma_q_sq_tpm = sigma_q_sq_tpm)\n",
       "    }\n",
       "    else {\n",
       "        ret$norm_fun_counts <- ret$norm_fun_tpm <- NA\n",
       "        ret$transform_fun_counts <- ret$transform_fun_tpm <- NA\n",
       "        ret$obs_norm <- data.frame()\n",
       "        ret$est_counts_sf <- vector()\n",
       "        est_counts_spread <- spread_abundance_by(obs_raw, \"est_counts\", \n",
       "            sample_to_covariates$sample)\n",
       "        if (!is.null(filter_target_id)) {\n",
       "            msg(\"A list of target IDs for filtering was found. Using this for filtering\")\n",
       "            target_ids <- rownames(est_counts_spread)\n",
       "            filter_bool <- target_ids %in% filter_target_id\n",
       "            names(filter_bool) <- target_ids\n",
       "        }\n",
       "        else {\n",
       "            filter_bool <- apply(est_counts_spread, 1, filter_fun)\n",
       "        }\n",
       "        filter_true <- filter_bool[filter_bool]\n",
       "        if (sum(filter_bool) == 0) {\n",
       "            stop(\"Zero targets passed the filter you used. Please double check the filter used.\")\n",
       "        }\n",
       "        msg(paste0(sum(filter_bool), \" targets passed the filter\"))\n",
       "        filter_df <- adf(target_id = names(filter_true))\n",
       "        if (ret$gene_mode) {\n",
       "            msg(paste0(\"aggregating by column: \", aggregation_column))\n",
       "            which_target_id <- ret$filter_df$target_id\n",
       "            target_mapping <- data.table::data.table(target_mapping)\n",
       "            target_mapping[target_mapping[[aggregation_column]] == \n",
       "                \"\", aggregation_column] <- NA\n",
       "            agg_id <- unique(target_mapping[, aggregation_column, \n",
       "                with = FALSE])\n",
       "            agg_id <- agg_id[[1]]\n",
       "            agg_id <- agg_id[!is.na(agg_id)]\n",
       "            mappings <- dplyr::select_(target_mapping, \"target_id\", \n",
       "                aggregation_column)\n",
       "            mappings <- data.table::as.data.table(mappings)\n",
       "            which_tms <- which(mappings$target_id %in% which_target_id)\n",
       "            which_agg_id <- unique(mappings[which_tms, aggregation_column, \n",
       "                with = FALSE])\n",
       "            which_agg_id <- which_agg_id[[1]]\n",
       "            which_agg_id <- which_agg_id[!is.na(which_agg_id)]\n",
       "            filter_df <- adf(target_id = which_agg_id)\n",
       "            filter_bool <- agg_id %in% which_agg_id\n",
       "            msg(paste0(length(which_agg_id), \" genes passed the filter\"))\n",
       "        }\n",
       "        ret$filter_bool <- filter_bool\n",
       "        ret$filter_df <- filter_df\n",
       "        ret$obs_norm_filt <- data.frame()\n",
       "        ret$tpm_sf <- vector()\n",
       "        ret$bs_quants <- list()\n",
       "        ret$bs_summary <- list()\n",
       "    }\n",
       "    class(ret) <- \"sleuth\"\n",
       "    ret\n",
       "}\n",
       "<bytecode: 0x555dbc4f9e68>\n",
       "<environment: namespace:sleuth>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sleuth_prep\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "936b2528-36a5-47da-a298-efab33f94241",
   "metadata": {},
   "source": [
    " \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "4442bf52-6431-47e9-877f-7987ad4055b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttg.df <- data.frame (ttg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "dfc64998-e3ff-4295-9cd7-21121236ef12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
